// @sanity/base/lib/components/previews/general/MediaPreview.styled.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TooltipContentStack = exports.RootBox = exports.ProgressFlex = exports.MediaSkeleton = exports.MediaFlex = void 0;
var _ui = require("@sanity/ui");
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var RootBox = (0, _styledComponents.default)(_ui.Box)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  position: relative;\n"])));
exports.RootBox = RootBox;
var MediaFlex = (0, _styledComponents.default)(_ui.Flex).attrs({
  align: 'center',
  justify: 'center'
})(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  position: absolute;\n  left: 0;\n  top: 0;\n  right: 0;\n  bottom: 0;\n"])));
exports.MediaFlex = MediaFlex;
var MediaSkeleton = (0, _styledComponents.default)(_ui.Skeleton).attrs({
  animated: true,
  radius: 2
})(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  width: 100%;\n  height: 100%;\n"])));
exports.MediaSkeleton = MediaSkeleton;
var ProgressFlex = (0, _styledComponents.default)(_ui.Flex).attrs({
  align: 'center',
  justify: 'center'
})(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n  position: absolute;\n  left: 0;\n  top: 0;\n  right: 0;\n  bottom: 0;\n\n  &:before {\n    background-color: var(--card-bg-color);\n    opacity: 0.75;\n    content: '';\n    display: block;\n    position: absolute;\n    left: 0;\n    top: 0;\n    right: 0;\n    bottom: 0;\n  }\n\n  > svg {\n    position: relative;\n    z-index: 2;\n  }\n"])));
exports.ProgressFlex = ProgressFlex;
var TooltipContentStack = (0, _styledComponents.default)(_ui.Stack).attrs({
  padding: 2,
  space: 2
})(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["\n  max-width: ", ";\n"])), (0, _ui.rem)(200));
exports.TooltipContentStack = TooltipContentStack;
