// @sanity/default-layout/lib/navbar/search/components/NoResults.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NoResults = NoResults;
var _ui = require("@sanity/ui");
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _search = require("../contexts/search");
var _TypePills = require("./TypePills");
var _templateObject;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var Semibold = _styledComponents.default.span(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  font-weight: ", ";\n"])), _ref => {
  var theme = _ref.theme;
  return theme.sanity.fonts.text.weights.semibold;
});
function NoResults() {
  var _useSearchState = (0, _search.useSearchState)(),
    state = _useSearchState.state;
  var typesSelected = state.terms.types.length > 0;
  return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    "aria-live": "assertive",
    direction: "column",
    flex: 1,
    gap: 4,
    paddingX: 4,
    paddingY: 5
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true
  }, "No results for ", /*#__PURE__*/_react.default.createElement(Semibold, null, state.terms.query)), typesSelected && /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Inline, {
    space: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true,
    size: 1
  }, "in"), /*#__PURE__*/_react.default.createElement(_TypePills.TypePills, {
    types: state.terms.types
  }))));
}
