// @sanity/default-layout/lib/navbar/search/components/PlaceholderSearchInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PlaceholderSearchInput = void 0;
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _constants = require("../constants");
var _search = require("../contexts/search");
var _templateObject, _templateObject2, _templateObject3;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var KeyboardShortcutFlex = (0, _styledComponents.default)(_ui.Flex)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  position: absolute;\n  right: 0;\n  top: 0;\n\n  & > :first-child {\n    margin-right: 1px;\n  }\n"])));
var PlaceholderSearchInputFlex = (0, _styledComponents.default)(_ui.Flex)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  min-width: 253px;\n  max-width: 350px;\n  position: relative;\n"])));
var PlaceholderTextInput = (0, _styledComponents.default)(_ui.TextInput)(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  padding-right: 60px;\n"])));
var PlaceholderSearchInput = /*#__PURE__*/(0, _react.forwardRef)(function DummyInput(_ref, ref) {
  var onOpen = _ref.onOpen;
  var _useSearchState = (0, _search.useSearchState)(),
    dispatch = _useSearchState.dispatch,
    terms = _useSearchState.state.terms;
  var handleChange = (0, _react.useCallback)(event => {
    dispatch({
      type: 'TERMS_QUERY_SET',
      query: event.currentTarget.value
    });
    onOpen();
  }, [dispatch, onOpen]);
  var handleKeyDown = (0, _react.useCallback)(event => {
    if (event.key === 'Enter') {
      onOpen();
    }
  }, [onOpen]);
  return /*#__PURE__*/_react.default.createElement(PlaceholderSearchInputFlex, {
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(PlaceholderTextInput, {
    radius: 2,
    "aria-autocomplete": "list",
    "aria-expanded": "false",
    autoComplete: "off",
    icon: _icons.SearchIcon,
    "data-testid": "studio-search",
    onChange: handleChange,
    onClick: onOpen,
    onKeyDown: handleKeyDown,
    placeholder: "Search",
    ref: ref,
    role: "combobox",
    value: terms.query
  }), /*#__PURE__*/_react.default.createElement(KeyboardShortcutFlex, {
    align: "center",
    height: "fill",
    marginRight: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.KBD, null, _constants.GLOBAL_SEARCH_KEY_MODIFIER.toUpperCase()), /*#__PURE__*/_react.default.createElement(_ui.KBD, null, _constants.GLOBAL_SEARCH_KEY.toUpperCase())));
});
exports.PlaceholderSearchInput = PlaceholderSearchInput;
