// @sanity/default-layout/lib/navbar/search/components/RecentSearchItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RecentSearchItem = RecentSearchItem;
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireWildcard(require("styled-components"));
var _commandList = require("../contexts/commandList");
var _TypePills = require("./TypePills");
var _templateObject, _templateObject2, _templateObject3, _templateObject4;
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var DEFAULT_COMBINED_TYPE_COUNT = 40;
var RecentSearchItemButton = (0, _styledComponents.default)(_ui.Button)(_ref => {
  var $level = _ref.$level;
  return (0, _styledComponents.css)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n    [data-focused='true'][data-level='", "'] &,\n    [data-hovered='true'][data-level='", "'] & {\n      &[data-active='true'] {\n        background: ", ";\n        // Disable box-shadow to hide the halo effect when we have keyboard focus over a selected <Button>\n        box-shadow: none;\n      }\n    }\n  "])), $level, $level, _ref2 => {
    var theme = _ref2.theme;
    return theme.sanity.color.button.bleed.default.hovered.bg;
  });
});
var SearchItemPillsBox = (0, _styledComponents.default)(_ui.Box)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  flex-shrink: 3;\n"])));
var SearchItemQueryBox = (0, _styledComponents.default)(_ui.Box)(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  flex-shrink: 2;\n"])));
var CloseButtonDiv = _styledComponents.default.div(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n  opacity: 0.8;\n  visibility: hidden;\n\n  @media (hover: hover) {\n    ", ":hover & {\n      visibility: visible;\n    }\n    &:hover {\n      opacity: 0.4;\n    }\n  }\n"])), RecentSearchItemButton);
function RecentSearchItem(props) {
  var index = props.index,
    _props$maxVisibleType = props.maxVisibleTypePillChars,
    maxVisibleTypePillChars = _props$maxVisibleType === void 0 ? DEFAULT_COMBINED_TYPE_COUNT : _props$maxVisibleType,
    value = props.value,
    onClick = props.onClick,
    onDelete = props.onDelete;
  var _useCommandList = (0, _commandList.useCommandList)(),
    level = _useCommandList.level,
    onChildClick = _useCommandList.onChildClick,
    onChildMouseDown = _useCommandList.onChildMouseDown,
    onChildMouseEnter = _useCommandList.onChildMouseEnter;
  var handleRecentSearchClick = (0, _react.useCallback)(() => {
    onChildClick === null || onChildClick === void 0 ? void 0 : onChildClick();
    onClick(value);
  }, [onChildClick, onClick, value]);
  var typesSelected = value.types.length > 0;

  // Determine how many characters are left to render type pills
  var availableCharacters = maxVisibleTypePillChars - value.query.length;
  return /*#__PURE__*/_react.default.createElement(RecentSearchItemButton, {
    $level: level,
    "data-index": index,
    mode: "bleed",
    onClick: handleRecentSearchClick,
    onMouseDown: onChildMouseDown,
    onMouseEnter: onChildMouseEnter(index),
    paddingLeft: 3,
    paddingRight: 1,
    paddingY: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    paddingY: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, /*#__PURE__*/_react.default.createElement(_icons.ClockIcon, null))), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    flex: 1,
    gap: 3,
    justify: "flex-start",
    marginLeft: 3
  }, value.query && /*#__PURE__*/_react.default.createElement(SearchItemQueryBox, {
    marginLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    textOverflow: "ellipsis"
  }, value.query)), typesSelected && /*#__PURE__*/_react.default.createElement(SearchItemPillsBox, null, /*#__PURE__*/_react.default.createElement(_TypePills.TypePills, {
    availableCharacters: availableCharacters,
    types: value.types
  }))), /*#__PURE__*/_react.default.createElement(CloseButtonDiv, {
    onClick: onDelete
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    padding: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, /*#__PURE__*/_react.default.createElement(_icons.CloseIcon, null))))));
}
