// @sanity/default-layout/lib/navbar/search/components/SearchError.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SearchError = SearchError;
var _components = require("@sanity/base/components");
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _react = _interopRequireDefault(require("react"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function SearchError() {
  return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    "aria-live": "assertive",
    direction: "column",
    flex: 1,
    gap: 3,
    marginY: 2,
    padding: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginBottom: 1
  }, /*#__PURE__*/_react.default.createElement(_components.TextWithTone, {
    tone: "critical"
  }, /*#__PURE__*/_react.default.createElement(_icons.WarningOutlineIcon, null))), /*#__PURE__*/_react.default.createElement(_components.TextWithTone, {
    size: 2,
    tone: "critical",
    weight: "semibold"
  }, "Something went wrong while searching"), /*#__PURE__*/_react.default.createElement(_components.TextWithTone, {
    size: 1,
    tone: "critical"
  }, "Please try again or check your connection"));
}
