// @sanity/base/lib/preview/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.INVALID_PREVIEW_FALLBACK = exports.INCLUDE_FIELDS_QUERY = exports.INCLUDE_FIELDS = exports.AVAILABILITY_READABLE = exports.AVAILABILITY_PERMISSION_DENIED = exports.AVAILABILITY_NOT_FOUND = void 0;
var _icons = require("@sanity/icons");
var _react = _interopRequireDefault(require("react"));
var _types = require("./types");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var INCLUDE_FIELDS_QUERY = ['_id', '_rev', '_type'];
exports.INCLUDE_FIELDS_QUERY = INCLUDE_FIELDS_QUERY;
var INCLUDE_FIELDS = [...INCLUDE_FIELDS_QUERY, '_key'];
exports.INCLUDE_FIELDS = INCLUDE_FIELDS;
var AVAILABILITY_READABLE = {
  available: true,
  reason: _types.AvailabilityReason.READABLE
};
exports.AVAILABILITY_READABLE = AVAILABILITY_READABLE;
var AVAILABILITY_PERMISSION_DENIED = {
  available: false,
  reason: _types.AvailabilityReason.PERMISSION_DENIED
};
exports.AVAILABILITY_PERMISSION_DENIED = AVAILABILITY_PERMISSION_DENIED;
var AVAILABILITY_NOT_FOUND = {
  available: false,
  reason: _types.AvailabilityReason.NOT_FOUND
};
exports.AVAILABILITY_NOT_FOUND = AVAILABILITY_NOT_FOUND;
var INVALID_PREVIEW_FALLBACK = {
  title: 'Invalid preview config',
  subtitle: 'Check the error log in the console',
  media: /*#__PURE__*/_react.default.createElement(_icons.WarningOutlineIcon, null)
};
exports.INVALID_PREVIEW_FALLBACK = INVALID_PREVIEW_FALLBACK;
