// @sanity/base/lib/preview/documentPair.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.observePathsDocumentPair = observePathsDocumentPair;
var _rxjs = require("rxjs");
var _operators = require("rxjs/operators");
var _draftUtils = require("../util/draftUtils");
var _availability = require("./availability");
var _ = require("./");
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
var ALWAYS_INCLUDED_SNAPSHOT_PATHS = [['_updatedAt'], ['_createdAt'], ['_type']];
function observePathsDocumentPair(id, paths) {
  var _getIdPair = (0, _draftUtils.getIdPair)(id),
    draftId = _getIdPair.draftId,
    publishedId = _getIdPair.publishedId;
  return (0, _availability.observeDocumentPairAvailability)(id).pipe((0, _operators.switchMap)(availability => {
    if (!availability.draft.available && !availability.published.available) {
      // short circuit, neither draft nor published is available so no point in trying to get a snapshot
      return (0, _rxjs.of)({
        id,
        type: null,
        draft: {
          availability: availability.draft,
          snapshot: undefined
        },
        published: {
          availability: availability.published,
          snapshot: undefined
        }
      });
    }
    var snapshotPaths = [...paths, ...ALWAYS_INCLUDED_SNAPSHOT_PATHS];
    return (0, _rxjs.combineLatest)([(0, _.observePaths)(draftId, snapshotPaths), (0, _.observePaths)(publishedId, snapshotPaths)]).pipe((0, _operators.map)(_ref => {
      var _ref2 = _slicedToArray(_ref, 2),
        draftSnapshot = _ref2[0],
        publishedSnapshot = _ref2[1];
      // note: assume type is always the same
      var type = (draftSnapshot === null || draftSnapshot === void 0 ? void 0 : draftSnapshot._type) || (publishedSnapshot === null || publishedSnapshot === void 0 ? void 0 : publishedSnapshot._type) || null;
      return {
        id: publishedId,
        type,
        draft: {
          availability: availability.draft,
          snapshot: draftSnapshot
        },
        published: {
          availability: availability.published,
          snapshot: publishedSnapshot
        }
      };
    }));
  }));
}
