// @sanity/desk-tool/lib/components/paneItem/ReferencedDocTooltip.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReferencedDocTooltip = ReferencedDocTooltip;
var _react = _interopRequireDefault(require("react"));
var _ui = require("@sanity/ui");
var _icons = require("@sanity/icons");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function ReferencedDocTooltip(_ref) {
  var totalReferenceCount = _ref.totalReferenceCount;
  return /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 2,
    marginRight: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
    content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1
    }, "Referenced by ", totalReferenceCount, totalReferenceCount === 1 ? ' document' : ' documents')),
    placement: "right",
    fallbackPlacements: ['bottom'],
    portal: true
  }, /*#__PURE__*/_react.default.createElement(_icons.LinkIcon, null))));
}
