// @sanity/desk-tool/lib/components/pane/__workshop__/SplitPanesStory/Navbar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Navbar = Navbar;
var _components = require("@sanity/base/components");
var _ui = require("@sanity/ui");
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _templateObject;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var RootCard = (0, _styledComponents.default)(_ui.Card)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  padding-left: env(safe-area-inset-left);\n  padding-right: env(safe-area-inset-right);\n"])));
var features = {
  inverted: true
};
function Navbar(props) {
  var path = props.path,
    setPath = props.setPath;
  var theme = (0, _ui.useTheme)();
  var dark = theme.sanity.color.dark;
  var handleHomeClick = (0, _react.useCallback)(() => {
    setPath(path.slice(0, 1));
  }, [path, setPath]);
  return /*#__PURE__*/_react.default.createElement(_components.LegacyLayerProvider, {
    zOffset: "navbar"
  }, /*#__PURE__*/_react.default.createElement(_ui.Layer, null, /*#__PURE__*/_react.default.createElement(RootCard, {
    scheme: features.inverted ? 'dark' : undefined,
    shadow: features.inverted ? dark ? 1 : 0 : 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, null, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    as: "button",
    onClick: handleHomeClick,
    padding: 3,
    radius: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    weight: "bold"
  }, "Studio")))))));
}
