// @sanity/desk-tool/lib/components/pane/__workshop__/SplitPanesStory/SplitPanesStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SplitPanesStory = SplitPanesStory;
var _ui = require("@sanity/ui");
var _react = _interopRequireWildcard(require("react"));
var _uiWorkshop = require("@sanity/ui-workshop");
var _PaneLayout = require("../../PaneLayout");
var _DocumentPane = require("./DocumentPane");
var _ListPane = require("./ListPane");
var _Navbar = require("./Navbar");
var _config = require("./config");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function SplitPanesStory() {
  var debug = (0, _uiWorkshop.useBoolean)('Debug', false) || false;
  var _useState = (0, _react.useState)(null),
    _useState2 = _slicedToArray(_useState, 2),
    portalElement = _useState2[0],
    setPortalElement = _useState2[1];
  var _useState3 = (0, _react.useState)(false),
    _useState4 = _slicedToArray(_useState3, 2),
    layoutCollapsed = _useState4[0],
    setLayoutCollapsed = _useState4[1];
  var _useState5 = (0, _react.useState)(['root']),
    _useState6 = _slicedToArray(_useState5, 2),
    path = _useState6[0],
    setPath = _useState6[1];
  var handleCollapse = (0, _react.useCallback)(() => setLayoutCollapsed(true), []);
  var handleExpand = (0, _react.useCallback)(() => setLayoutCollapsed(false), []);
  return /*#__PURE__*/_react.default.createElement(_ui.ToastProvider, {
    paddingY: 7,
    zOffset: [100, 11000]
  }, /*#__PURE__*/_react.default.createElement(_ui.PortalProvider, {
    element: portalElement
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    direction: "column",
    height: layoutCollapsed ? undefined : 'fill',
    style: {
      minHeight: '100%'
    }
  }, /*#__PURE__*/_react.default.createElement(_Navbar.Navbar, {
    path: path,
    setPath: setPath
  }), /*#__PURE__*/_react.default.createElement(DeskTool, {
    collapsed: layoutCollapsed,
    onExpand: handleExpand,
    onCollapse: handleCollapse,
    path: path,
    setPath: setPath
  }))), /*#__PURE__*/_react.default.createElement("div", {
    "data-portal": "",
    ref: setPortalElement,
    style: {
      outline: '1px solid red'
    }
  }));
}
function DeskTool(props) {
  var collapsed = props.collapsed,
    onCollapse = props.onCollapse,
    onExpand = props.onExpand,
    path = props.path,
    setPath = props.setPath;
  return /*#__PURE__*/_react.default.createElement(_PaneLayout.PaneLayout, {
    flex: 1,
    height: collapsed ? undefined : 'fill',
    minWidth: 512,
    onCollapse: onCollapse,
    onExpand: onExpand
  }, path.map((s, i) => {
    var key = "".concat(s, "-").concat(i);
    var pane = _config.panes.find(p => p.id === s);
    if (!pane) {
      return /*#__PURE__*/_react.default.createElement("div", {
        key: key
      }, "not found: ", s);
    }
    if (pane.type === 'list') {
      return /*#__PURE__*/_react.default.createElement(_ListPane.ListPane, {
        active: i === path.length - 2,
        childId: path[i + 1],
        index: i,
        key: key,
        node: pane,
        setPath: setPath
      });
    }
    return /*#__PURE__*/_react.default.createElement(_DocumentPane.DocumentPane, {
      index: i,
      key: key,
      node: pane,
      setPath: setPath
    });
  }));
}
