// @sanity/base/lib/components/_deprecated/CreateDocumentPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreateDocumentPreview = CreateDocumentPreview;
var _react = _interopRequireDefault(require("react"));
var _IntentButton = require("../IntentButton");
var _previews = require("../previews");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var DEFAULT_MEDIA_DIMENSION = {
  width: 80,
  height: 80,
  aspect: 1,
  fit: 'crop'
};
var BLOCK_STYLE = {
  display: 'flex',
  height: '100%',
  width: '100%',
  alignItems: 'flex-start'
};

/**
 * @deprecated
 */
function CreateDocumentPreview(props) {
  var _props$title = props.title,
    title = _props$title === void 0 ? 'Untitled' : _props$title,
    subtitle = props.subtitle,
    _props$media = props.media,
    media = _props$media === void 0 ? props.icon : _props$media,
    isPlaceholder = props.isPlaceholder,
    _props$mediaDimension = props.mediaDimensions,
    mediaDimensions = _props$mediaDimension === void 0 ? DEFAULT_MEDIA_DIMENSION : _props$mediaDimension,
    description = props.description,
    params = props.params,
    templateParams = props.templateParams;
  return /*#__PURE__*/_react.default.createElement(_IntentButton.IntentButton, {
    intent: "create",
    params: [params, templateParams],
    title: subtitle ? "Create new ".concat(title, " (").concat(subtitle, ")") : "Create new ".concat(title),
    onClick: props.onClick,
    style: BLOCK_STYLE,
    mode: "ghost",
    fontSize: 2
  }, /*#__PURE__*/_react.default.createElement(_previews.TemplatePreview, {
    description: description,
    isPlaceholder: isPlaceholder,
    media: media,
    mediaDimensions: mediaDimensions
  }));
}
