// @sanity/base/lib/components/previews/template/TemplatePreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TemplatePreview = TemplatePreview;
var _react = _interopRequireDefault(require("react"));
var _ui = require("@sanity/ui");
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _useDevicePixelRatio = require("use-device-pixel-ratio");
var _Media = require("../_common/Media");
var _constants = require("../constants");
var _templateObject, _templateObject2, _templateObject3, _templateObject4;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
var DEFAULT_MEDIA_DIMENSION = _objectSpread(_objectSpread({}, _constants.PREVIEW_MEDIA_SIZE.default), {}, {
  aspect: 1,
  fit: 'crop',
  dpr: (0, _useDevicePixelRatio.getDevicePixelRatio)()
});
var Root = (0, _styledComponents.default)(_ui.Box)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  height: 100%;\n\n  a {\n    color: currentColor;\n    text-decoration: none;\n  }\n\n  svg[data-sanity-icon] {\n    margin: 0;\n  }\n"])));
var HeaderFlex = (0, _styledComponents.default)(_ui.Flex).attrs({
  align: 'center'
})(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  height: ", ";\n"])), (0, _ui.rem)(_constants.PREVIEW_MEDIA_SIZE.default.height));
var TitleSkeleton = (0, _styledComponents.default)(_ui.TextSkeleton).attrs({
  animated: true,
  radius: 1
})(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  max-width: ", ";\n  width: 80%;\n"])), (0, _ui.rem)(160));
var SubtitleSkeleton = (0, _styledComponents.default)(_ui.TextSkeleton).attrs({
  animated: true,
  radius: 1,
  size: 1
})(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n  max-width: ", ";\n  width: 60%;\n"])), (0, _ui.rem)(120));
function TemplatePreview(props) {
  var description = props.description,
    isPlaceholder = props.isPlaceholder,
    media = props.media,
    _props$mediaDimension = props.mediaDimensions,
    mediaDimensions = _props$mediaDimension === void 0 ? DEFAULT_MEDIA_DIMENSION : _props$mediaDimension,
    subtitle = props.subtitle,
    _props$title = props.title,
    title = _props$title === void 0 ? 'Untitled' : _props$title;
  if (isPlaceholder) {
    return /*#__PURE__*/_react.default.createElement(Root, null, /*#__PURE__*/_react.default.createElement(HeaderFlex, null, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      flex: 1,
      space: 2
    }, /*#__PURE__*/_react.default.createElement(TitleSkeleton, null), /*#__PURE__*/_react.default.createElement(SubtitleSkeleton, null))));
  }
  return /*#__PURE__*/_react.default.createElement(Root, null, /*#__PURE__*/_react.default.createElement(HeaderFlex, null, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    flex: 1,
    space: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    textOverflow: "ellipsis"
  }, typeof title !== 'function' && title, typeof title === 'function' && title({
    layout: 'default'
  })), subtitle && /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true,
    size: 1,
    textOverflow: "ellipsis"
  }, typeof subtitle === 'function' && subtitle({
    layout: 'default'
  }) || subtitle)), media && /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "flex-start",
    paddingLeft: 2
  }, /*#__PURE__*/_react.default.createElement(_Media.Media, {
    dimensions: mediaDimensions,
    layout: "default",
    media: media
  }))), description && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginTop: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true,
    size: 1,
    style: {
      whiteSpace: 'break-spaces'
    }
  }, description)));
}
