// @sanity/form-builder/lib/inputs/files/FileInput/FileInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _uniqueId2 = _interopRequireDefault(require("lodash/uniqueId"));
var _partition2 = _interopRequireDefault(require("lodash/partition"));
var _get2 = _interopRequireDefault(require("lodash/get"));
var _react = _interopRequireDefault(require("react"));
var _components = require("@sanity/base/components");
var _changeIndicators = require("@sanity/base/change-indicators");
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _presence = require("@sanity/base/presence");
var _assetUtils = require("@sanity/asset-utils");
var _WithReferencedAsset = require("../../../utils/WithReferencedAsset");
var _PatchEvent = _interopRequireWildcard(require("../../../PatchEvent"));
var _styles = require("../common/styles");
var _UploadProgress = require("../common/UploadProgress");
var _assetSource = require("../common/assetSource");
var _resolveUploader = _interopRequireDefault(require("../../../sanity/uploads/resolveUploader"));
var _ActionsMenu = require("../common/ActionsMenu");
var _PlaceholderText = require("../common/PlaceholderText");
var _UploadPlaceholder = _interopRequireDefault(require("../common/UploadPlaceholder"));
var _UploadWarning = require("../common/UploadWarning");
var _empty = require("../../../utils/empty");
var _styles2 = require("./styles");
var _FileInputField = require("./FileInputField");
var _FileDetails = require("./FileDetails");
var _FileSkeleton = require("./FileSkeleton");
var _InvalidFileWarning = require("./InvalidFileWarning");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
var HIDDEN_FIELDS = ['asset'];
class FileInput extends _react.default.PureComponent {
  constructor() {
    var _this;
    super(...arguments);
    _this = this;
    _defineProperty(this, "_inputId", (0, _uniqueId2.default)('FileInput'));
    _defineProperty(this, "dialogId", (0, _uniqueId2.default)('fileinput-dialog'));
    _defineProperty(this, "_focusRef", null);
    _defineProperty(this, "uploadSubscription", null);
    _defineProperty(this, "state", {
      isUploading: false,
      selectedAssetSource: null,
      hoveringFiles: [],
      isStale: false,
      isMenuOpen: false
    });
    _defineProperty(this, "toast", null);
    _defineProperty(this, "handleRemoveButtonClick", () => {
      var _this$props = this.props,
        value = _this$props.value,
        getValuePath = _this$props.getValuePath;
      var parentPathSegment = getValuePath().slice(-1)[0];

      // String path segment mean an object path, while a number or a
      // keyed segment means we're a direct child of an array
      var isArrayElement = typeof parentPathSegment !== 'string';

      // When removing the file, _type and _key are "meta"-properties and
      // are not significant unless other properties are present. Thus, we
      // want to remove the entire "container" object if these are the only
      // properties present, BUT only if we're not an array element, as
      // removing the array element will close the selection dialog. Instead,
      // when closing the dialog, the array logic will check for an "empty"
      // value and remove it for us
      var allKeys = Object.keys(value || {});
      var remainingKeys = allKeys.filter(key => !['_type', '_key', '_upload', 'asset'].includes(key));
      var isEmpty = remainingKeys.length === 0;
      var removeKeys = ['asset'].concat(allKeys.filter(key => ['_upload'].includes(key))).map(key => (0, _PatchEvent.unset)([key]));
      this.props.onChange(_PatchEvent.default.from(isEmpty && !isArrayElement ? (0, _PatchEvent.unset)() : removeKeys));
    });
    _defineProperty(this, "handleCancelUpload", () => {
      this.cancelUpload();
    });
    _defineProperty(this, "handleClearUploadState", () => {
      this.setState({
        isStale: false
      });
      this.clearUploadStatus();
    });
    _defineProperty(this, "handleStaleUpload", () => {
      this.setState({
        isStale: true
      });
    });
    _defineProperty(this, "handleClearField", () => {
      this.props.onChange(_PatchEvent.default.from((0, _PatchEvent.unset)(['asset'])));
    });
    _defineProperty(this, "handleSelectFiles", files => {
      var _this$props2 = this.props,
        directUploads = _this$props2.directUploads,
        readOnly = _this$props2.readOnly,
        value = _this$props2.value;
      var hoveringFiles = this.state.hoveringFiles;
      if (directUploads && !readOnly) {
        this.uploadFirstAccepted(files);
      } else if (hoveringFiles.length > 0) {
        this.handleFilesOut();
      }
    });
    _defineProperty(this, "handleSelectFileFromAssetSource", source => {
      this.setState({
        selectedAssetSource: source
      });
    });
    _defineProperty(this, "handleAssetSourceClosed", () => {
      this.setState({
        selectedAssetSource: null
      });
    });
    _defineProperty(this, "uploadWith", function (uploader, file) {
      var assetDocumentProps = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      var _this$props3 = _this.props,
        type = _this$props3.type,
        onChange = _this$props3.onChange;
      var source = assetDocumentProps.source;
      var options = {
        metadata: (0, _get2.default)(type, 'options.metadata'),
        storeOriginalFilename: (0, _get2.default)(type, 'options.storeOriginalFilename'),
        source
      };
      _this.cancelUpload();
      _this.setState({
        isUploading: true
      });
      onChange(_PatchEvent.default.from([(0, _PatchEvent.setIfMissing)({
        _type: type.name
      })]));
      _this.uploadSubscription = uploader.upload(file, type, options).subscribe({
        next: uploadEvent => {
          if (uploadEvent.patches) {
            onChange(_PatchEvent.default.from(uploadEvent.patches));
          }
        },
        error: err => {
          var _this$toast;
          // eslint-disable-next-line no-console
          console.error(err);
          (_this$toast = _this.toast) === null || _this$toast === void 0 ? void 0 : _this$toast.push({
            status: 'error',
            description: 'The upload could not be completed at this time.',
            title: 'Upload failed'
          });
          _this.clearUploadStatus();
        },
        complete: () => {
          _this.setState({
            isUploading: false
          });
        }
      });
    });
    _defineProperty(this, "handleFieldChange", event => {
      var _this$props4 = this.props,
        onChange = _this$props4.onChange,
        type = _this$props4.type;

      // When editing a metadata field for a file (eg `description`), and no asset
      // is currently selected, we want to unset the entire file field if the
      // field we are currently editing goes blank and gets unset.
      //
      // For instance:
      // A file field with a `description` and a `title` subfield, where the file `asset`
      // and the `title` field is empty, and we are erasing the `description` field.
      // We do _not_ however want to clear the field if any content is present in
      // the other fields, eg if `title` _has_ a value and we erase `description`
      //
      // Also, we don't want to use this logic for array items, since the parent will
      // take care of it when closing the array dialog
      if (!this.valueIsArrayElement() && this.eventIsUnsettingLastFilledField(event)) {
        onChange(_PatchEvent.default.from((0, _PatchEvent.unset)()));
        return;
      }
      onChange(event.prepend((0, _PatchEvent.setIfMissing)({
        _type: type.name
      })));
    });
    _defineProperty(this, "eventIsUnsettingLastFilledField", event => {
      var patch = event.patches[0];
      if (event.patches.length !== 1 || patch.type !== 'unset') {
        return false;
      }
      var allKeys = Object.keys(this.props.value || {});
      var remainingKeys = allKeys.filter(key => !['_type', '_key'].includes(key));
      var isEmpty = event.patches[0].path.length === 1 && remainingKeys.length === 1 && remainingKeys[0] === event.patches[0].path[0];
      return isEmpty;
    });
    _defineProperty(this, "valueIsArrayElement", () => {
      var getValuePath = this.props.getValuePath;
      var parentPathSegment = getValuePath().slice(-1)[0];

      // String path segment mean an object path, while a number or a
      // keyed segment means we're a direct child of an array
      return typeof parentPathSegment !== 'string';
    });
    _defineProperty(this, "handleOpenDialog", () => {
      var _this$props5 = this.props,
        type = _this$props5.type,
        onFocus = _this$props5.onFocus;
      var otherFields = type.fields.filter(field => {
        var _field$type, _field$type$options;
        return !HIDDEN_FIELDS.includes(field.name) && !((_field$type = field.type) !== null && _field$type !== void 0 && (_field$type$options = _field$type.options) !== null && _field$type$options !== void 0 && _field$type$options.isHighlighted);
      });

      /* for context: this code will only ever run if there are fields which are not highlighted.
      in the FileDetails component (used in the renderAsset method) there is a button which is active if there are
      fields that are not highlighted, opening this dialog */
      onFocus([otherFields[0].name]);
    });
    _defineProperty(this, "handleStopAdvancedEdit", () => {
      this.props.onFocus([]);
    });
    _defineProperty(this, "handleSelectAssetFromSource", assetFromSource => {
      var _this$props6 = this.props,
        onChange = _this$props6.onChange,
        type = _this$props6.type;
      (0, _assetSource.handleSelectAssetFromSource)({
        assetFromSource,
        onChange,
        type,
        resolveUploader: _resolveUploader.default,
        uploadWith: this.uploadWith
      });
      this.setState({
        selectedAssetSource: null
      });
    });
    _defineProperty(this, "handleFileTargetFocus", event => {
      // We want to handle focus when the file target element *itself* receives
      // focus, not when an interactive child element receives focus. Since React has decided
      // to let focus bubble, so this workaround is needed
      // Background: https://github.com/facebook/react/issues/6410#issuecomment-671915381
      if (event.currentTarget === event.target && event.currentTarget === this._focusRef) {
        this.props.onFocus(['asset']);
      }
    });
    _defineProperty(this, "handleFileTargetBlur", () => {
      this.props.onBlur();
    });
    _defineProperty(this, "handleFilesOver", fileInfo => {
      this.setState({
        hoveringFiles: fileInfo
      });
    });
    _defineProperty(this, "handleFilesOut", () => {
      this.setState({
        hoveringFiles: []
      });
    });
    _defineProperty(this, "setFocusInput", ref => {
      this._focusRef = ref;
    });
    _defineProperty(this, "handleUpload", _ref => {
      var file = _ref.file,
        uploader = _ref.uploader;
      this.uploadWith(uploader, file);
    });
    _defineProperty(this, "setToast", toast => {
      this.toast = toast;
    });
  }
  clearUploadStatus() {
    var _this$props$value;
    // todo: this is kind of hackish
    if ((_this$props$value = this.props.value) !== null && _this$props$value !== void 0 && _this$props$value._upload) {
      this.props.onChange(_PatchEvent.default.from([(0, _PatchEvent.unset)(['_upload'])]));
    }
  }
  cancelUpload() {
    if (this.uploadSubscription) {
      this.uploadSubscription.unsubscribe();
      this.clearUploadStatus();
    }
  }
  uploadFirstAccepted(files) {
    var type = this.props.type;
    var match = files.map(file => ({
      file,
      uploader: (0, _resolveUploader.default)(type, file)
    })).find(result => result.uploader);
    if (match) {
      this.uploadWith(match.uploader, match.file);
    }
    this.setState({
      isMenuOpen: false
    });
  }
  renderUploadState(uploadState) {
    var isUploading = this.state.isUploading;
    return /*#__PURE__*/_react.default.createElement(_UploadProgress.UploadProgress, {
      uploadState: uploadState,
      onCancel: isUploading ? this.handleCancelUpload : undefined,
      onStale: this.handleStaleUpload
    });
  }
  renderAssetSource() {
    var selectedAssetSource = this.state.selectedAssetSource;
    var _this$props7 = this.props,
      value = _this$props7.value,
      observeAsset = _this$props7.observeAsset;
    if (!selectedAssetSource) {
      return null;
    }
    var Component = selectedAssetSource.component;
    if (value && value.asset) {
      return /*#__PURE__*/_react.default.createElement(_WithReferencedAsset.WithReferencedAsset, {
        observeAsset: observeAsset,
        reference: value.asset,
        waitPlaceholder: /*#__PURE__*/_react.default.createElement(_FileSkeleton.FileSkeleton, null)
      }, fileAsset => /*#__PURE__*/_react.default.createElement(Component, {
        selectedAssets: [fileAsset],
        selectionType: "single",
        assetType: "file",
        dialogHeaderTitle: "Select file",
        onClose: this.handleAssetSourceClosed,
        onSelect: this.handleSelectAssetFromSource
      }));
    }
    return /*#__PURE__*/_react.default.createElement(Component, {
      selectedAssets: [],
      selectionType: "single",
      assetType: "file",
      dialogHeaderTitle: "Select file",
      onClose: this.handleAssetSourceClosed,
      onSelect: this.handleSelectAssetFromSource
    });
  }
  renderAdvancedEdit(fields) {
    return /*#__PURE__*/_react.default.createElement(_ui.Dialog, {
      header: "Edit details",
      id: this.dialogId,
      onClose: this.handleStopAdvancedEdit,
      width: 1,
      __unstable_autoFocus: false
    }, /*#__PURE__*/_react.default.createElement(_presence.PresenceOverlay, {
      margins: [0, 0, 1, 0]
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 4
    }, this.renderFields(fields))));
  }
  renderFields(fields) {
    return fields.map(field => this.renderField(field));
  }
  hasFileTargetFocus() {
    var _this$props$focusPath;
    return ((_this$props$focusPath = this.props.focusPath) === null || _this$props$focusPath === void 0 ? void 0 : _this$props$focusPath[0]) === 'asset';
  }
  renderField(field) {
    var _this$props8 = this.props,
      value = _this$props8.value,
      level = _this$props8.level,
      focusPath = _this$props8.focusPath,
      onFocus = _this$props8.onFocus,
      readOnly = _this$props8.readOnly,
      onBlur = _this$props8.onBlur,
      compareValue = _this$props8.compareValue,
      presence = _this$props8.presence,
      markers = _this$props8.markers;
    var fieldValue = value === null || value === void 0 ? void 0 : value[field.name];
    var fieldMarkers = markers.filter(marker => marker.path[0] === field.name);
    return /*#__PURE__*/_react.default.createElement(_FileInputField.FileInputField, {
      key: field.name,
      field: field,
      parentValue: value,
      value: fieldValue,
      onChange: this.handleFieldChange,
      onFocus: onFocus,
      compareValue: compareValue,
      onBlur: onBlur,
      readOnly: readOnly || field.type.readOnly,
      focusPath: focusPath,
      level: level,
      presence: presence,
      markers: fieldMarkers
    });
  }
  renderAsset(hasAdvancedFields) {
    var _this$props9 = this.props,
      value = _this$props9.value,
      readOnly = _this$props9.readOnly,
      assetSources = _this$props9.assetSources,
      type = _this$props9.type,
      directUploads = _this$props9.directUploads,
      observeAsset = _this$props9.observeAsset;
    var isMenuOpen = this.state.isMenuOpen;
    var asset = value === null || value === void 0 ? void 0 : value.asset;
    if (!asset) {
      return null;
    }
    var accept = (0, _get2.default)(type, 'options.accept', 'image/*');
    var browseMenuItem = assetSources && (assetSources === null || assetSources === void 0 ? void 0 : assetSources.length) === 0 ? null : /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      icon: _icons.SearchIcon,
      text: "Browse",
      onClick: () => {
        this.setState({
          isMenuOpen: false
        });
        this.handleSelectFileFromAssetSource(assetSources[0]);
      },
      disabled: readOnly,
      "data-testid": "file-input-browse-button"
    });
    if (assetSources.length > 1) {
      browseMenuItem = assetSources.map(assetSource => {
        return /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
          key: assetSource.name,
          text: assetSource.title,
          onClick: () => {
            this.setState({
              isMenuOpen: false
            });
            this.handleSelectFileFromAssetSource(assetSource);
          },
          icon: assetSource.icon || _icons.ImageIcon,
          disabled: readOnly,
          "data-testid": "file-input-browse-button-".concat(assetSource.name)
        });
      });
    }
    return /*#__PURE__*/_react.default.createElement(_WithReferencedAsset.WithReferencedAsset, {
      reference: asset,
      observeAsset: observeAsset,
      waitPlaceholder: /*#__PURE__*/_react.default.createElement(_FileSkeleton.FileSkeleton, null)
    }, assetDocument => /*#__PURE__*/_react.default.createElement(_FileDetails.FileDetails, {
      size: assetDocument.size,
      originalFilename: (assetDocument === null || assetDocument === void 0 ? void 0 : assetDocument.originalFilename) || "download.".concat(assetDocument.extension),
      onClick: hasAdvancedFields ? this.handleOpenDialog : undefined,
      muted: !hasAdvancedFields && readOnly,
      disabled: !hasAdvancedFields,
      onMenuOpen: isOpen => this.setState({
        isMenuOpen: isOpen
      }),
      isMenuOpen: isMenuOpen
    }, /*#__PURE__*/_react.default.createElement(_ActionsMenu.ActionsMenu, {
      onUpload: this.handleSelectFiles,
      browse: browseMenuItem,
      onReset: this.handleRemoveButtonClick,
      downloadUrl: "".concat(assetDocument.url, "?dl"),
      copyUrl: "".concat(assetDocument.url),
      readOnly: readOnly,
      accept: accept,
      directUploads: directUploads
    })));
  }
  renderAssetMenu(tone) {
    var _this$props10 = this.props,
      type = _this$props10.type,
      readOnly = _this$props10.readOnly,
      directUploads = _this$props10.directUploads;
    var hoveringFiles = this.state.hoveringFiles;
    var acceptedFiles = hoveringFiles.filter(file => (0, _resolveUploader.default)(type, file));
    var rejectedFilesCount = hoveringFiles.length - acceptedFiles.length;
    return /*#__PURE__*/_react.default.createElement(_styles2.CardOverlay, {
      tone: tone
    }, /*#__PURE__*/_react.default.createElement(_styles2.FlexContainer, {
      align: "center",
      justify: "center",
      gap: 2,
      flex: 1
    }, /*#__PURE__*/_react.default.createElement(_PlaceholderText.PlaceholderText, {
      readOnly: readOnly,
      hoveringFiles: hoveringFiles,
      acceptedFiles: acceptedFiles,
      rejectedFilesCount: rejectedFilesCount,
      directUploads: directUploads,
      type: 'file'
    })));
  }
  renderBrowser() {
    var _this$props11 = this.props,
      assetSources = _this$props11.assetSources,
      readOnly = _this$props11.readOnly,
      directUploads = _this$props11.directUploads;
    if (assetSources.length === 0) return null;
    if (assetSources.length > 1 && !readOnly && directUploads) {
      return /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
        id: "".concat(this._inputId, "_assetFileButton"),
        button: /*#__PURE__*/_react.default.createElement(_ui.Button, {
          mode: "ghost",
          text: "Select",
          "data-testid": "file-input-multi-browse-button",
          icon: _icons.SearchIcon
        }),
        "data-testid": "input-select-button",
        menu: /*#__PURE__*/_react.default.createElement(_ui.Menu, null, assetSources.map(assetSource => {
          return /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
            key: assetSource.name,
            text: assetSource.title,
            onClick: () => {
              this.setState({
                isMenuOpen: false
              });
              this.handleSelectFileFromAssetSource(assetSource);
            },
            icon: assetSource.icon || _icons.ImageIcon,
            disabled: readOnly,
            "data-testid": "file-input-browse-button-".concat(assetSource.name)
          });
        }))
      });
    }
    return /*#__PURE__*/_react.default.createElement(_ui.Button, {
      fontSize: 2,
      text: "Browse",
      icon: _icons.SearchIcon,
      mode: "ghost",
      onClick: () => {
        this.setState({
          isMenuOpen: false
        });
        this.handleSelectFileFromAssetSource(assetSources[0]);
      },
      "data-testid": "file-input-browse-button",
      disabled: readOnly
    });
  }
  renderUploadPlaceholder() {
    var _this$props12 = this.props,
      readOnly = _this$props12.readOnly,
      type = _this$props12.type,
      directUploads = _this$props12.directUploads;
    var hoveringFiles = this.state.hoveringFiles;
    var acceptedFiles = hoveringFiles.filter(file => (0, _resolveUploader.default)(type, file));
    var rejectedFilesCount = hoveringFiles.length - acceptedFiles.length;
    var accept = (0, _get2.default)(type, 'options.accept', '');
    return /*#__PURE__*/_react.default.createElement("div", {
      style: {
        padding: 1
      }
    }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
      tone: readOnly ? 'transparent' : 'inherit',
      border: true,
      padding: 3,
      style: hoveringFiles.length === 0 ? {
        borderStyle: 'dashed'
      } : {
        borderStyle: 'dashed',
        borderColor: 'transparent'
      }
    }, /*#__PURE__*/_react.default.createElement(_UploadPlaceholder.default, {
      browse: this.renderBrowser(),
      onUpload: this.handleSelectFiles,
      readOnly: readOnly,
      hoveringFiles: hoveringFiles,
      acceptedFiles: acceptedFiles,
      rejectedFilesCount: rejectedFilesCount,
      type: "file",
      accept: accept,
      directUploads: directUploads
    })));
  }
  focus() {
    if (this._focusRef) {
      this._focusRef.focus();
    }
  }
  render() {
    var _value$asset, _compareValue$asset, _value$asset2;
    var _this$props13 = this.props,
      directUploads = _this$props13.directUploads,
      type = _this$props13.type,
      value = _this$props13.value,
      compareValue = _this$props13.compareValue,
      level = _this$props13.level,
      markers = _this$props13.markers,
      readOnly = _this$props13.readOnly,
      presence = _this$props13.presence,
      _this$props13$focusPa = _this$props13.focusPath,
      focusPath = _this$props13$focusPa === void 0 ? _empty.EMPTY_ARRAY : _this$props13$focusPa;
    var _this$state = this.state,
      hoveringFiles = _this$state.hoveringFiles,
      selectedAssetSource = _this$state.selectedAssetSource,
      isStale = _this$state.isStale;
    var fieldGroups = (0, _partition2.default)(type.fields.filter(field => !HIDDEN_FIELDS.includes(field.name)), 'type.options.isHighlighted');
    var _fieldGroups = _slicedToArray(fieldGroups, 2),
      highlightedFields = _fieldGroups[0],
      otherFields = _fieldGroups[1];

    // Whoever is present at the asset field is who we show on the field itself
    var assetFieldPresence = presence.filter(item => item.path[0] === 'asset');
    var isDialogOpen = focusPath.length > 0 && otherFields.some(field => focusPath[0] === field.name);
    function getFileTone() {
      var acceptedFiles = hoveringFiles.filter(file => (0, _resolveUploader.default)(type, file));
      var rejectedFilesCount = hoveringFiles.length - acceptedFiles.length;
      if (hoveringFiles.length > 0) {
        if (rejectedFilesCount > 0 || !directUploads) {
          return 'critical';
        }
      }
      if (!(value !== null && value !== void 0 && value._upload) && !readOnly && hoveringFiles.length > 0) {
        return 'primary';
      }
      return value !== null && value !== void 0 && value._upload && value !== null && value !== void 0 && value.asset && readOnly ? 'transparent' : 'default';
    }
    var hasValueOrUpload = Boolean((value === null || value === void 0 ? void 0 : value._upload) || (value === null || value === void 0 ? void 0 : value.asset));
    var hasInvalidFile = value && typeof value.asset !== 'undefined' && !(value !== null && value !== void 0 && value._upload) && !(0, _assetUtils.isFileSource)(value);
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_components.ImperativeToast, {
      ref: this.setToast
    }), /*#__PURE__*/_react.default.createElement(_components.FormFieldSet, {
      __unstable_markers: markers,
      title: type.title,
      description: type.description,
      level: highlightedFields.length > 0 ? level : 0,
      __unstable_presence: isDialogOpen ? _empty.EMPTY_ARRAY : assetFieldPresence,
      __unstable_changeIndicator: false
    }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_changeIndicators.ChangeIndicatorCompareValueProvider, {
      value: value === null || value === void 0 ? void 0 : (_value$asset = value.asset) === null || _value$asset === void 0 ? void 0 : _value$asset._ref,
      compareValue: compareValue === null || compareValue === void 0 ? void 0 : (_compareValue$asset = compareValue.asset) === null || _compareValue$asset === void 0 ? void 0 : _compareValue$asset._ref
    }, isStale && /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginBottom: 2
    }, /*#__PURE__*/_react.default.createElement(_UploadWarning.UploadWarning, {
      onClearStale: this.handleClearUploadState
    })), /*#__PURE__*/_react.default.createElement(_changeIndicators.ChangeIndicatorWithProvidedFullPath, {
      path: [],
      hasFocus: this.hasFileTargetFocus(),
      value: value === null || value === void 0 ? void 0 : (_value$asset2 = value.asset) === null || _value$asset2 === void 0 ? void 0 : _value$asset2._ref
    }, !(value !== null && value !== void 0 && value._upload) && !hasInvalidFile && /*#__PURE__*/_react.default.createElement(_styles.FileTarget, {
      tabIndex: 0,
      disabled: Boolean(readOnly),
      ref: this.setFocusInput,
      onFiles: this.handleSelectFiles,
      onFilesOver: this.handleFilesOver,
      onFilesOut: this.handleFilesOut,
      onFocus: this.handleFileTargetFocus,
      onBlur: this.handleFileTargetBlur,
      tone: getFileTone(),
      $border: hasValueOrUpload || hoveringFiles.length > 0,
      padding: hasValueOrUpload ? 1 : 0,
      radius: 2
    }, (value === null || value === void 0 ? void 0 : value.asset) && this.renderAsset(otherFields.length > 0), !(value !== null && value !== void 0 && value.asset) && this.renderUploadPlaceholder(), value !== null && value !== void 0 && value.asset && hoveringFiles.length > 0 ? this.renderAssetMenu(getFileTone()) : null), hasInvalidFile && /*#__PURE__*/_react.default.createElement(_InvalidFileWarning.InvalidFileWarning, {
      onClearValue: this.handleClearField
    }), (value === null || value === void 0 ? void 0 : value._upload) && this.renderUploadState(value._upload)))), highlightedFields.length > 0 && this.renderFields(highlightedFields), isDialogOpen && this.renderAdvancedEdit(otherFields), selectedAssetSource && this.renderAssetSource()));
  }
}
exports.default = FileInput;
