// @sanity/form-builder/lib/inputs/files/FileInput/InvalidFileWarning.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InvalidFileWarning = InvalidFileWarning;
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _templateObject;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var ButtonWrapper = (0, _styledComponents.default)(_ui.Button)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  width: 100%;\n"])));
function InvalidFileWarning(_ref) {
  var onClearValue = _ref.onClearValue;
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    tone: "caution",
    padding: 4,
    border: true,
    radius: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    gap: 4,
    marginBottom: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, /*#__PURE__*/_react.default.createElement(_icons.WarningOutlineIcon, null))), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    weight: "semibold"
  }, "Invalid file value"), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, "The value of this field is not a valid file. Resetting this field will let you choose a new file."))), /*#__PURE__*/_react.default.createElement(ButtonWrapper, {
    icon: _icons.ResetIcon,
    text: "Reset value",
    onClick: onClearValue,
    mode: "ghost"
  }));
}
