// @sanity/desk-tool/lib/panes/userComponent/UserComponentPaneHeader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UserComponentPaneHeader = UserComponentPaneHeader;
var _react = _interopRequireDefault(require("react"));
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _components = require("../../components");
var _deskTool = require("../../contexts/deskTool");
var _paneRouter = require("../../contexts/paneRouter");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function UserComponentPaneHeader(props) {
  var actionHandlers = props.actionHandlers,
    index = props.index,
    menuItems = props.menuItems,
    menuItemGroups = props.menuItemGroups,
    title = props.title;
  var _useDeskTool = (0, _deskTool.useDeskTool)(),
    features = _useDeskTool.features;
  if (!(menuItems !== null && menuItems !== void 0 && menuItems.length) && !title) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement(_components.PaneHeader, {
    actions: /*#__PURE__*/_react.default.createElement(_components.PaneHeaderActions, {
      menuItems: menuItems,
      menuItemGroups: menuItemGroups,
      actionHandlers: actionHandlers
    }),
    backButton: features.backButton && index > 0 && /*#__PURE__*/_react.default.createElement(_ui.Button, {
      as: _paneRouter.BackLink,
      "data-as": "a",
      icon: _icons.ArrowLeftIcon,
      mode: "bleed"
    }),
    title: title
  });
}
