// @sanity/base/lib/components/collapseMenu/CollapseMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CollapseMenu = void 0;
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireWildcard(require("styled-components"));
var _CollapseOverflowMenu = require("./CollapseOverflowMenu");
var _ObserveElement = require("./ObserveElement");
var _CollapseMenuDivider = require("./CollapseMenuDivider");
var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6;
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var FOCUS_RING_PADDING = 3;
var OPTION_STYLE = (0, _styledComponents.css)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  list-style: none;\n  display: flex;\n  white-space: nowrap;\n\n  &[data-hidden='true'] {\n    opacity: 0;\n    visibility: hidden;\n  }\n"])));
var OuterFlex = (0, _styledComponents.default)(_ui.Flex)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  padding: ", "px;\n  margin: -", "px;\n  box-sizing: border-box;\n"])), FOCUS_RING_PADDING, FOCUS_RING_PADDING);
var RootFlex = (0, _styledComponents.default)(_ui.Flex)(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  border-radius: inherit;\n  position: relative;\n"])));
var RowFlex = (0, _styledComponents.default)(_ui.Flex)(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n  width: max-content;\n\n  &[data-hidden='true'] {\n    height: 0px;\n    visibility: hidden;\n  }\n"])));
var OptionObserveElement = (0, _styledComponents.default)(_ObserveElement.ObserveElement)(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["\n  ", "\n"])), OPTION_STYLE);
var OptionHiddenFlex = (0, _styledComponents.default)(_ui.Flex)(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["\n  ", "\n"])), OPTION_STYLE);
function _isReactElement(node) {
  return Boolean(node);
}
var CollapseMenu = /*#__PURE__*/(0, _react.forwardRef)(function CollapseMenu(props, ref) {
  var childrenProp = props.children,
    collapsed = props.collapsed,
    _props$collapseText = props.collapseText,
    collapseText = _props$collapseText === void 0 ? true : _props$collapseText,
    disableRestoreFocusOnClose = props.disableRestoreFocusOnClose,
    gap = props.gap,
    menuButtonProps = props.menuButtonProps,
    onMenuClose = props.onMenuClose;
  var _useState = (0, _react.useState)(null),
    _useState2 = _slicedToArray(_useState, 2),
    rootEl = _useState2[0],
    setRootEl = _useState2[1];
  var _useState3 = (0, _react.useState)(null),
    _useState4 = _slicedToArray(_useState3, 2),
    hiddenRowEl = _useState4[0],
    setHiddenRowEl = _useState4[1];
  var rootRect = (0, _ui.useElementRect)(rootEl);
  var _useState5 = (0, _react.useState)([]),
    _useState6 = _slicedToArray(_useState5, 2),
    menuOptions = _useState6[0],
    setMenuOptions = _useState6[1];
  var hasOverflow = (0, _react.useMemo)(() => {
    if (rootRect && rootEl && hiddenRowEl) {
      return rootRect.width < hiddenRowEl.scrollWidth;
    }
    return false;
  }, [hiddenRowEl, rootEl, rootRect]);
  var menuButton = (0, _react.useMemo)(() => (menuButtonProps === null || menuButtonProps === void 0 ? void 0 : menuButtonProps.button) || /*#__PURE__*/_react.default.createElement(_ui.Button, {
    icon: _icons.EllipsisVerticalIcon,
    mode: "bleed"
  }), [menuButtonProps]);
  var intersectionOptions = (0, _react.useMemo)(() => ({
    root: rootEl,
    threshold: 1,
    rootMargin: '2px'
  }), [rootEl]);
  var children = (0, _react.useMemo)(() => _react.default.Children.toArray(childrenProp).filter(_isReactElement), [childrenProp]);
  var menuOptionsArray = (0, _react.useMemo)(
  // eslint-disable-next-line max-nested-callbacks
  () => children.filter(_ref => {
    var key = _ref.key;
    return menuOptions.find(o => o.key === key);
  }), [children, menuOptions]);
  var menuIsVisible = (0, _react.useMemo)(() => collapsed || menuOptionsArray.length > 0, [collapsed, menuOptionsArray.length]);
  var isInMenu = (0, _react.useCallback)(childKey => menuOptionsArray.some(o => o.key === childKey), [menuOptionsArray]);
  var handleIntersection = (0, _react.useCallback)((e, child) => {
    var exists = isInMenu(child.key);
    if (!e.isIntersecting && !exists) {
      setMenuOptions(prev => [child, ...prev]);
    }
    if (e.isIntersecting && exists) {
      var updatedOptions = menuOptionsArray.filter(_ref2 => {
        var key = _ref2.key;
        return key !== child.key;
      });
      setMenuOptions(updatedOptions);
    }
  }, [isInMenu, menuOptionsArray]);
  var items = (0, _react.useMemo)(() => children.map(child => {
    var _child$props = child.props,
      collapsedProps = _child$props.collapsedProps,
      expandedProps = _child$props.expandedProps;
    var modeProps = hasOverflow ? collapsedProps : expandedProps;
    var text = hasOverflow && collapseText ? undefined : child.props.text;
    return /*#__PURE__*/(0, _react.cloneElement)(child, _objectSpread(_objectSpread({}, modeProps), {}, {
      text: text
    }));
  }), [children, collapseText, hasOverflow]);
  if (collapsed) {
    return /*#__PURE__*/_react.default.createElement(_CollapseOverflowMenu.CollapseOverflowMenu, {
      disableRestoreFocusOnClose: disableRestoreFocusOnClose,
      menuButton: menuButton,
      menuButtonProps: menuButtonProps,
      menuOptionsArray: children,
      onMenuClose: onMenuClose
    });
  }
  return /*#__PURE__*/_react.default.createElement(OuterFlex, {
    align: "center",
    "data-ui": "CollapseMenu",
    overflow: "hidden",
    sizing: "border",
    ref: ref
  }, /*#__PURE__*/_react.default.createElement(RootFlex, {
    direction: "column",
    flex: 1,
    justify: "center",
    ref: setRootEl
  }, /*#__PURE__*/_react.default.createElement(RowFlex, {
    gap: gap
  }, items.map((child, index) => {
    var _child$props2 = child.props,
      dividerBefore = _child$props2.dividerBefore,
      _child$props2$tooltip = _child$props2.tooltipText,
      tooltipText = _child$props2$tooltip === void 0 ? '' : _child$props2$tooltip,
      _child$props2$tooltip2 = _child$props2.tooltipProps,
      tooltipProps = _child$props2$tooltip2 === void 0 ? {} : _child$props2$tooltip2;
    var hidden = isInMenu(child.key);
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
      key: child.key
    }, dividerBefore && index !== 0 && /*#__PURE__*/_react.default.createElement(_CollapseMenuDivider.CollapseMenuDivider, {
      hidden: hidden
    }), /*#__PURE__*/_react.default.createElement(OptionObserveElement, {
      options: intersectionOptions
      // eslint-disable-next-line react/jsx-no-bind
      ,
      callback: e => handleIntersection(e[0], child),
      "aria-hidden": hidden,
      "data-hidden": hidden
    }, /*#__PURE__*/_react.default.createElement(_ui.Tooltip, _extends({
      portal: true,
      disabled: !tooltipText,
      content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
        padding: 2,
        sizing: "border"
      }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
        size: 1
      }, tooltipText))
    }, tooltipProps), /*#__PURE__*/_react.default.createElement(_ui.Flex, null, /*#__PURE__*/(0, _react.cloneElement)(child, {
      disabled: child.props.disabled || hidden,
      'aria-hidden': hidden
    })))));
  })), /*#__PURE__*/_react.default.createElement(RowFlex, {
    "data-hidden": true,
    "aria-hidden": "true",
    gap: gap,
    ref: setHiddenRowEl
  }, children.map((child, index) => {
    var dividerBefore = child.props.dividerBefore;
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
      key: child.key
    }, dividerBefore && index !== 0 && /*#__PURE__*/_react.default.createElement(_CollapseMenuDivider.CollapseMenuDivider, null), /*#__PURE__*/_react.default.createElement(OptionHiddenFlex, {
      key: child.key
    }, /*#__PURE__*/(0, _react.cloneElement)(child, {
      disabled: true,
      'aria-hidden': true
    })));
  }))), menuIsVisible && /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    marginLeft: gap
  }, /*#__PURE__*/_react.default.createElement(_CollapseOverflowMenu.CollapseOverflowMenu, {
    disableRestoreFocusOnClose: disableRestoreFocusOnClose,
    menuButton: menuButton,
    menuButtonProps: menuButtonProps,
    menuOptionsArray: menuOptionsArray,
    onMenuClose: onMenuClose
  })));
});
exports.CollapseMenu = CollapseMenu;
