// @sanity/default-login/lib/LoginDialogContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _ui = require("@sanity/ui");
var _configSanity = _interopRequireDefault(require("config:sanity"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _legacyParts = require("./legacyParts");
var _LoginProviderButton = require("./LoginProviderButton");
var _templateObject;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var BrandLogoWrapper = (0, _styledComponents.default)(_ui.Flex)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  svg {\n    display: block;\n    height: 2rem;\n    width: auto;\n    margin: 0 auto;\n  }\n"])));
var projectName = _configSanity.default.project && _configSanity.default.project.name || '';
class LoginDialogContent extends _react.default.PureComponent {
  constructor() {
    super(...arguments);
    _defineProperty(this, "handleLoginButtonClicked", (provider, event) => {
      var onLoginButtonClick = this.props.onLoginButtonClick;
      if (onLoginButtonClick) {
        this.props.onLoginButtonClick(provider, event);
      } else {
        console.warn('LoginDialogContent is missing the onLoginButtonClick property'); // eslint-disable-line no-console
      }
    });
  }

  render() {
    var _this$props = this.props,
      title = _this$props.title,
      description = _this$props.description,
      _this$props$providers = _this$props.providers,
      providers = _this$props$providers === void 0 ? [] : _this$props$providers,
      SanityLogo = _this$props.SanityLogo;
    return /*#__PURE__*/_react.default.createElement(_ui.Box, {
      paddingX: 4,
      paddingY: 5
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 4
    }, _legacyParts.BrandLogo && projectName && /*#__PURE__*/_react.default.createElement(BrandLogoWrapper, {
      justify: "center"
    }, /*#__PURE__*/_react.default.createElement(_legacyParts.BrandLogo, {
      projectName: projectName
    })), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 4
    }, !_legacyParts.BrandLogo && projectName && /*#__PURE__*/_react.default.createElement(_ui.Heading, {
      align: "center",
      as: "h1"
    }, projectName), title && /*#__PURE__*/_react.default.createElement(_ui.Text, {
      align: "center",
      size: 1,
      weight: "semibold"
    }, title), description && /*#__PURE__*/_react.default.createElement(_ui.Text, {
      align: "center",
      muted: true,
      size: 1
    }, description)), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 2,
      as: "ul"
    }, providers.map((provider, providerIndex) => /*#__PURE__*/_react.default.createElement(_LoginProviderButton.LoginProviderButton, {
      key: (provider === null || provider === void 0 ? void 0 : provider.name) || providerIndex,
      onLogin: this.handleLoginButtonClicked,
      provider: provider
    }))), SanityLogo && /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      justify: "center",
      marginTop: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, /*#__PURE__*/_react.default.createElement(SanityLogo, null)))));
  }
}
exports.default = LoginDialogContent;
_defineProperty(LoginDialogContent, "propTypes", {
  title: _propTypes.default.node.isRequired,
  description: _propTypes.default.node,
  // eslint-disable-next-line react/forbid-prop-types
  providers: _propTypes.default.array,
  onLoginButtonClick: _propTypes.default.func,
  SanityLogo: _propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.node, _propTypes.default.func])
});
_defineProperty(LoginDialogContent, "defaultProps", {
  description: null,
  providers: null,
  onLoginButtonClick: null,
  SanityLogo: null
});
