// @sanity/default-login/lib/LoginProviderButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LoginProviderButton = LoginProviderButton;
var _ui = require("@sanity/ui");
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireWildcard(require("styled-components"));
var _getProviderLogo = require("./util/getProviderLogo");
var _templateObject, _templateObject2;
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var Root = (0, _styledComponents.default)(_ui.Button)(_ref => {
  var theme = _ref.theme;
  var _theme$sanity$color$b = theme.sanity.color.button,
    bleed = _theme$sanity$color$b.bleed,
    ghost = _theme$sanity$color$b.ghost;
  return (0, _styledComponents.css)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n    @media (hover: hover) {\n      &:not([data-disabled='true']):hover {\n        --card-bg-color: ", ";\n        --card-fg-color: ", ";\n        --card-border-color: ", ";\n      }\n    }\n  "])), bleed.default.hovered.bg, bleed.default.hovered.fg, ghost.default.enabled.border);
});
var ProviderLogoWrapper = (0, _styledComponents.default)(_ui.Box)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  svg,\n  img {\n    border-radius: 50%;\n    height: 1.25em;\n    width: 1.25em;\n  }\n"])));
function LoginProviderButton(_ref2) {
  var onLogin = _ref2.onLogin,
    provider = _ref2.provider;
  var ProviderLogo = (0, _getProviderLogo.getProviderLogo)(provider);
  var handleClick = (0, _react.useCallback)(event => {
    onLogin(provider, event);
  }, [onLogin, provider]);
  return /*#__PURE__*/_react.default.createElement(Root, {
    mode: "ghost",
    paddingY: 3,
    onClick: handleClick
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    justify: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Inline, {
    space: 2
  }, /*#__PURE__*/_react.default.createElement(ProviderLogoWrapper, null, /*#__PURE__*/_react.default.createElement(ProviderLogo, null)), /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, null, provider === null || provider === void 0 ? void 0 : provider.title)))));
}
