// @sanity/default-layout/lib/main/NotFound.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NotFound = void 0;
var _react = _interopRequireDefault(require("react"));
var _router = require("@sanity/base/router");
var _ui = require("@sanity/ui");
var _spaces = require("../util/spaces");
var _useDefaultLayoutRouter = require("../useDefaultLayoutRouter");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var NotFound = props => {
  var router = (0, _useDefaultLayoutRouter.useDefaultLayoutRouter)();
  var rootState = _spaces.HAS_SPACES && router.state.space ? {
    space: router.state.space
  } : {};
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    height: "fill",
    paddingX: [5, 5, 7],
    paddingY: [5, 5, 6],
    sizing: "border"
  }, /*#__PURE__*/_react.default.createElement(_ui.Container, null, /*#__PURE__*/_react.default.createElement(_ui.Heading, {
    as: "h1"
  }, "Page not found"), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 4,
    paddingY: 4
  }, props.children && /*#__PURE__*/_react.default.createElement(_ui.Box, null, props.children), /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, null, /*#__PURE__*/_react.default.createElement(_router.StateLink, {
    state: rootState
  }, "Go to index"))))));
};
exports.NotFound = NotFound;
