// @sanity/default-layout/lib/main/RenderTool.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RenderTool = void 0;
var _react = _interopRequireWildcard(require("react"));
var _ui = require("@sanity/ui");
var _getRegisteredTools = require("../util/getRegisteredTools");
var _ErrorScreen = require("./ErrorScreen");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
var defaultUnknownError = {
  message: 'An unknown error occurred while rendering'
};
var RenderTool = /*#__PURE__*/(0, _react.memo)(function RenderTool(props) {
  var tools = (0, _getRegisteredTools.getRegisteredTools)();
  var activeToolName = props.tool;
  var activeToolNameRef = (0, _react.useRef)(activeToolName);
  var activeTool = tools.find(tool => tool.name === activeToolName);
  var _useState = (0, _react.useState)({
      error: null,
      showErrorDetails: __DEV__
    }),
    _useState2 = _slicedToArray(_useState, 2),
    state = _useState2[0],
    setState = _useState2[1];
  (0, _react.useEffect)(() => {
    var prevToolName = activeToolNameRef.current;
    if (state.error && prevToolName !== activeToolName) {
      // reset error state
      setState({
        error: null,
        showErrorDetails: __DEV__
      });
    }
  }, [activeToolName, state.error]);
  var handleCatch = (0, _react.useCallback)(errorParams => {
    setState(prevState => _objectSpread(_objectSpread({}, prevState), {}, {
      error: errorParams
    }));
  }, []);
  var handleRetry = (0, _react.useCallback)(() => {
    setState(prevState => _objectSpread(_objectSpread({}, prevState), {}, {
      error: null
    }));
  }, []);
  if (state.error) {
    var _state$error = state.error,
      error = _state$error.error,
      info = _state$error.info;
    var showErrorDetails = state.showErrorDetails;
    return /*#__PURE__*/_react.default.createElement(_ErrorScreen.RenderToolErrorScreen, {
      activeTool: activeTool
      // Some (rare) errors don't seem to have any Error instance attached
      // In these cases, default to an error-like object with a generic message
      ,
      error: error || defaultUnknownError,
      info: info,
      onRetry: handleRetry,
      showErrorDetails: showErrorDetails
    });
  }
  if (!tools.length) {
    return /*#__PURE__*/_react.default.createElement(_ui.Card, {
      height: "fill",
      paddingX: [5, 5, 7],
      paddingY: [5, 5, 6],
      sizing: "border"
    }, /*#__PURE__*/_react.default.createElement(_ui.Container, null, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginBottom: 5
    }, /*#__PURE__*/_react.default.createElement(_ui.Heading, {
      as: "h1"
    }, "No available tools")), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 4
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      muted: true
    }, "No tools implement the ", /*#__PURE__*/_react.default.createElement("code", null, "part:@sanity/base/tool"), ", so there is nothing to display."))));
  }
  if (!activeTool) {
    return /*#__PURE__*/_react.default.createElement(_ui.Card, {
      height: "fill",
      paddingX: [5, 5, 7],
      paddingY: [5, 5, 6],
      sizing: "border"
    }, /*#__PURE__*/_react.default.createElement(_ui.Container, null, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginBottom: 5
    }, /*#__PURE__*/_react.default.createElement(_ui.Heading, {
      as: "h1"
    }, "Tool not found: ", /*#__PURE__*/_react.default.createElement("code", null, props.tool))), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 4
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      muted: true
    }, "The list of tools installed in this Studio does not include ", /*#__PURE__*/_react.default.createElement("code", null, props.tool)))));
  }
  return /*#__PURE__*/_react.default.createElement(_ui.ErrorBoundary, {
    onCatch: handleCatch
  }, /*#__PURE__*/(0, _react.createElement)(activeTool.component, {
    tool: props.tool
  }));
});
exports.RenderTool = RenderTool;
