// @sanity/default-layout/lib/navbar/presence/PresenceMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PresenceMenu = PresenceMenu;
var _components = require("@sanity/base/components");
var _hooks = require("@sanity/base/hooks");
var _ui = require("@sanity/ui");
var _icons = require("@sanity/icons");
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _components2 = require("../components");
var _PresenceMenuItem = require("./PresenceMenuItem");
var _templateObject, _templateObject2, _templateObject3;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var AvatarStackCard = (0, _styledComponents.default)(_ui.Card)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  background: transparent;\n"])));
var FooterCard = (0, _styledComponents.default)(_ui.Card)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  position: sticky;\n  bottom: 0;\n"])));
var StyledMenu = (0, _styledComponents.default)(_ui.Menu)(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  max-width: 350px;\n  min-width: 250px;\n"])));
var PRESENCE_MENU_POPOVER_PROPS = {
  portal: true,
  constrainSize: true,
  scheme: 'light'
};
function PresenceMenu(props) {
  var collapse = props.collapse,
    maxAvatars = props.maxAvatars,
    projectId = props.projectId,
    _props$label = props.label,
    label = _props$label === void 0 ? 'Who is here' : _props$label;
  var users = (0, _hooks.useGlobalPresenceUsers)();
  var hasPresence = users.length > 0;
  var button = (0, _react.useMemo)(() => collapse ? /*#__PURE__*/_react.default.createElement(_components2.StatusButton, {
    icon: _icons.UsersIcon,
    mode: "bleed",
    statusTone: hasPresence ? 'positive' : undefined,
    "aria-label": label
  }) : /*#__PURE__*/_react.default.createElement(_ui.Button, {
    mode: "bleed",
    padding: 1,
    "aria-label": label
  }, /*#__PURE__*/_react.default.createElement(AvatarStackCard, null, /*#__PURE__*/_react.default.createElement(_ui.AvatarStack, {
    maxLength: maxAvatars
  }, users.map(user => /*#__PURE__*/_react.default.createElement(_components.UserAvatar, {
    key: user.id,
    user: user
  }))))), [collapse, maxAvatars, users, hasPresence]);
  return /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
    button: button,
    popover: PRESENCE_MENU_POPOVER_PROPS,
    id: "presence-menu",
    placement: "bottom",
    menu: hasPresence ? /*#__PURE__*/_react.default.createElement(PresenceMenuContent, {
      projectId: projectId
    }) : /*#__PURE__*/_react.default.createElement(StyledMenu, {
      padding: 0
    })
  });
}
function PresenceMenuContent(_ref) {
  var projectId = _ref.projectId;
  var presence = (0, _hooks.useGlobalPresence)();
  var hasPresence = presence.length > 0;
  return /*#__PURE__*/_react.default.createElement(StyledMenu, {
    padding: 0
  }, !hasPresence && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    paddingX: 3,
    paddingY: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    weight: "semibold",
    size: 2
  }, "No one else is here"), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    muted: true
  }, "Invite people to the project to see their online status."))), hasPresence && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 1,
    paddingBottom: 0
  }, presence.map(item => /*#__PURE__*/_react.default.createElement(_PresenceMenuItem.PresenceMenuItem, {
    presence: item,
    key: item.user.id
  }))), /*#__PURE__*/_react.default.createElement(FooterCard, {
    padding: 1,
    paddingTop: 0,
    radius: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.MenuDivider, null), /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
    as: "a",
    text: "Manage members",
    paddingY: 4,
    iconRight: _icons.CogIcon,
    href: "https://sanity.io/manage/project/".concat(projectId),
    target: "_blank",
    rel: "noopener noreferrer"
  }))));
}
