// @sanity/form-builder/lib/inputs/arrays/ArrayOfObjectsInput/item/CellItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CellItem = void 0;
var _icons = require("@sanity/icons");
var _presence = require("@sanity/base/presence");
var _react = _interopRequireWildcard(require("react"));
var _ui = require("@sanity/ui");
var _components = require("@sanity/base/components");
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _types = require("@sanity/types");
var _autoId = require("@reach/auto-id");
var _Preview = _interopRequireDefault(require("../../../../Preview"));
var _DragHandle = require("../../common/DragHandle");
var _randomKey = _interopRequireDefault(require("../../common/randomKey"));
var _ArrayInput = require("../ArrayInput");
var _InsertMenu = require("../InsertMenu");
var _ItemWithMissingType = require("./ItemWithMissingType");
var _excluded = ["isSortable", "value", "onClick", "onKeyPress", "onFocus", "onInsert", "insertableTypes", "type", "readOnly", "presence", "onRemove", "validation"];
var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var dragHandle = /*#__PURE__*/_react.default.createElement(_DragHandle.DragHandle, {
  grid: true,
  padding: 2,
  mode: "ghost"
});
var DragHandleCard = (0, _styledComponents.default)(_ui.Card)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  position: absolute;\n  top: 0;\n  left: 0;\n"])));
var PresenceFlex = (0, _styledComponents.default)(_ui.Flex)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  position: absolute;\n  top: 0;\n  right: 0;\n  height: 35px;\n"])));
var Root = (0, _styledComponents.default)(_ui.Card)(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  transition: border-color 250ms;\n  position: relative;\n\n  @media (hover: hover) {\n    ", " {\n      opacity: 0;\n    }\n\n    &:hover,\n    &:focus-within {\n      ", " {\n        opacity: 1;\n      }\n    }\n  }\n\n  &[aria-selected='true'] {\n    box-shadow: 0 0 0 2px var(--card-focus-ring-color);\n  }\n"])), DragHandleCard, DragHandleCard);
var FooterFlex = (0, _styledComponents.default)(_ui.Flex)(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n  min-height: 35px;\n"])));
var PreviewCard = (0, _styledComponents.default)(_ui.Card)(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["\n  border-top-right-radius: inherit;\n  border-top-left-radius: inherit;\n  height: 100%;\n\n  @media (hover: hover) {\n    &:hover {\n      filter: brightness(95%);\n    }\n  }\n\n  &:focus:focus-visible {\n    box-shadow: 0 0 0 2px var(--card-focus-ring-color);\n  }\n"])));
var MissingTypeBox = (0, _styledComponents.default)(_ui.Box)(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["\n  padding-bottom: 100%;\n"])));
var StyledItemWithMissingType = (0, _styledComponents.default)(_ItemWithMissingType.ItemWithMissingType)(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["\n  position: absolute;\n  top: 0;\n  left: 0;\n  height: 100%;\n"])));
var CellItem = /*#__PURE__*/_react.default.forwardRef(function ItemCell(props, ref) {
  var focusRef = _react.default.useRef();
  var isSortable = props.isSortable,
    value = props.value,
    onClick = props.onClick,
    onKeyPress = props.onKeyPress,
    onFocus = props.onFocus,
    onInsert = props.onInsert,
    insertableTypes = props.insertableTypes,
    type = props.type,
    readOnly = props.readOnly,
    presence = props.presence,
    onRemove = props.onRemove,
    validation = props.validation,
    rest = _objectWithoutProperties(props, _excluded);
  var hasError = validation.filter(_types.isValidationErrorMarker).length > 0;
  var hasWarning = validation.filter(_types.isValidationWarningMarker).length > 0;
  var hasKey = value._key;
  var tone = (0, _react.useMemo)(() => {
    if (!hasKey) {
      return 'caution';
    }
    if (hasError) {
      return 'critical';
    }
    if (hasWarning) {
      return 'caution';
    }
    return undefined;
  }, [hasError, hasWarning, hasKey]);
  var handleDuplicate = (0, _react.useCallback)(() => {
    onInsert === null || onInsert === void 0 ? void 0 : onInsert({
      item: _objectSpread(_objectSpread({}, value), {}, {
        _key: (0, _randomKey.default)()
      }),
      position: 'after',
      path: [{
        _key: value._key
      }],
      edit: false
    });
  }, [onInsert, value]);
  var handleInsert = (0, _react.useCallback)((pos, insertType) => {
    onInsert === null || onInsert === void 0 ? void 0 : onInsert({
      item: (0, _ArrayInput.createProtoValue)(insertType),
      position: pos,
      path: [{
        _key: value._key
      }]
    });
  }, [onInsert, value._key]);
  var id = (0, _autoId.useId)();
  return /*#__PURE__*/_react.default.createElement(Root, _extends({}, rest, {
    radius: 2,
    ref: ref,
    border: true,
    tone: tone
  }), type ? /*#__PURE__*/_react.default.createElement(PreviewCard, {
    tone: "inherit",
    "data-ui": "PreviewCard",
    forwardedAs: "button",
    type: "button",
    overflow: "auto",
    flex: 1,
    tabIndex: 0,
    onClick: onClick,
    ref: focusRef,
    onKeyPress: onKeyPress,
    onFocus: onFocus,
    __unstable_focusRing: true
  }, /*#__PURE__*/_react.default.createElement(_Preview.default, {
    layout: "media",
    value: value,
    type: type,
    withRadius: false,
    withBorder: false
  })) : /*#__PURE__*/_react.default.createElement(MissingTypeBox, {
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(StyledItemWithMissingType, {
    value: value,
    onFocus: onFocus,
    vertical: true
  })), /*#__PURE__*/_react.default.createElement(DragHandleCard, {
    margin: 1,
    radius: 2,
    display: "flex",
    tone: "inherit",
    "data-ui": "DragHandleCard"
  }, !readOnly && isSortable && dragHandle), /*#__PURE__*/_react.default.createElement(PresenceFlex, {
    align: "center",
    marginX: 1
  }, !readOnly && /*#__PURE__*/_react.default.createElement(_presence.FieldPresence, {
    presence: presence,
    maxAvatars: 1
  })), /*#__PURE__*/_react.default.createElement(FooterFlex, {
    align: "center",
    paddingX: 1,
    sizing: "border",
    justify: "space-between"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, null, value._key && validation.length > 0 && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1,
    paddingX: 1,
    paddingY: 3
  }, /*#__PURE__*/_react.default.createElement(_components.FormFieldValidationStatus, {
    __unstable_markers: validation,
    __unstable_showSummary: !value._ref,
    placement: "bottom",
    portal: true
  })), !hasKey && /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
    portal: true,
    content: /*#__PURE__*/_react.default.createElement(_ui.Card, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1
    }, "This item is missing a required ", /*#__PURE__*/_react.default.createElement("code", null, "_key"), " property."))
  }, /*#__PURE__*/_react.default.createElement(_ui.Badge, {
    mode: "outline",
    tone: "caution",
    margin: 1,
    padding: 2,
    fontSize: 1
  }, /*#__PURE__*/_react.default.createElement(_icons.WarningOutlineIcon, null), " key"))), !readOnly && /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
    button: /*#__PURE__*/_react.default.createElement(_ui.Button, {
      padding: 2,
      mode: "bleed",
      icon: _icons.EllipsisVerticalIcon
    }),
    id: "".concat(id, "-menuButton"),
    portal: true,
    menu: /*#__PURE__*/_react.default.createElement(_ui.Menu, null, /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      text: "Remove",
      tone: "critical",
      icon: _icons.TrashIcon,
      onClick: onRemove
    }), /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      text: "Duplicate",
      icon: _icons.CopyIcon,
      onClick: handleDuplicate
    }), /*#__PURE__*/_react.default.createElement(_InsertMenu.InsertMenu, {
      types: insertableTypes,
      onInsert: handleInsert
    }))
  }))));
});
exports.CellItem = CellItem;
