// @sanity/form-builder/lib/inputs/arrays/ArrayOfObjectsInput/item/ItemWithMissingType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ItemWithMissingType = ItemWithMissingType;
var _ui = require("@sanity/ui");
var _react = _interopRequireWildcard(require("react"));
var _icons = require("@sanity/icons");
var _content = require("@sanity/util/content");
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _excluded = ["value", "onFocus", "vertical"];
var _templateObject;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var PopoverCard = (0, _styledComponents.default)(_ui.Card)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  max-width: ", "px;\n"])), _ref => {
  var theme = _ref.theme;
  return theme.sanity.container[1];
});
function ItemWithMissingType(props) {
  var value = props.value,
    onFocus = props.onFocus,
    vertical = props.vertical,
    rest = _objectWithoutProperties(props, _excluded);
  var _React$useState = _react.default.useState(false),
    _React$useState2 = _slicedToArray(_React$useState, 2),
    showDetails = _React$useState2[0],
    setShowDetails = _React$useState2[1];
  var _React$useState3 = _react.default.useState(null),
    _React$useState4 = _slicedToArray(_React$useState3, 2),
    popoverRef = _React$useState4[0],
    setPopoverRef = _React$useState4[1];
  (0, _ui.useClickOutside)(() => setShowDetails(false), [popoverRef]);
  var handleKeyDown = _react.default.useCallback(e => {
    if (e.key === 'Escape' || e.key === 'Tab') {
      setShowDetails(false);
    }
  }, []);
  var handleShowDetails = (0, _react.useCallback)(() => {
    setShowDetails(v => !v);
  }, []);
  var typeName = (0, _content.resolveTypeName)(value);
  return /*#__PURE__*/_react.default.createElement(_ui.Popover, {
    open: showDetails,
    ref: setPopoverRef,
    onKeyDown: handleKeyDown,
    portal: true,
    constrainSize: true,
    tone: "default",
    content: /*#__PURE__*/_react.default.createElement(PopoverCard, {
      margin: 1,
      padding: 3,
      onKeyDown: handleKeyDown,
      tabIndex: 0,
      overflow: "auto"
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 4
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      weight: "semibold"
    }, "Why is this happening?")), /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1
    }, "The current schema does not declare items of type ", /*#__PURE__*/_react.default.createElement("code", null, typeName), " as valid for this list. This could mean that the type has been removed as a valid item type, or that someone else has added it to their own local schema that is not yet deployed."), /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1
    }, /*#__PURE__*/_react.default.createElement(_icons.BulbOutlineIcon, null), " You can still move or delete this item, but it cannot be edited since the schema definition for its type is nowhere to be found.")), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1,
      weight: "semibold"
    }, "JSON representation of this item:"), /*#__PURE__*/_react.default.createElement(_ui.Card, {
      padding: 2,
      overflow: "auto",
      border: true
    }, /*#__PURE__*/_react.default.createElement(_ui.Code, {
      size: 1,
      as: "pre"
    }, JSON.stringify(value, null, 2))))))
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, _extends({
    as: "button",
    type: "button",
    radius: 2,
    tone: "inherit",
    padding: 2,
    onFocus: onFocus,
    onClick: handleShowDetails,
    onKeyDown: handleKeyDown
  }, rest), vertical && /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_icons.UnknownIcon, null)), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    align: "center"
  }, "Item type ", /*#__PURE__*/_react.default.createElement("code", null, typeName), " not defined for this list")), !vertical && /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginRight: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, /*#__PURE__*/_react.default.createElement(_icons.UnknownIcon, null))), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    textOverflow: "ellipsis"
  }, "Item type ", /*#__PURE__*/_react.default.createElement("code", null, typeName), " not defined for this list")))));
}
