// @sanity/default-layout/lib/navbar/search/components/searchResultItem/DebugOverlay.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DebugOverlay = DebugOverlay;
var _ui = require("@sanity/ui");
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _templateObject;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var DebugScoreCard = (0, _styledComponents.default)(_ui.Card)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  cursor: help;\n  left: 0;\n  position: absolute;\n  top: 0;\n"])));
function DebugOverlay(_ref) {
  var data = _ref.data;
  var score = data.score;
  var tone = 'default';
  if (score <= 0.1) {
    tone = 'critical';
  } else if (score >= 0.5) {
    tone = 'primary';
  }
  var matchingStories = data.stories.filter(story => story.score > 0);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
    content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 2
    }, matchingStories.length ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, matchingStories.map(story => /*#__PURE__*/_react.default.createElement(_ui.Inline, {
      key: story.path,
      space: 3
    }, /*#__PURE__*/_react.default.createElement(_ui.Code, {
      size: 0,
      weight: "semibold"
    }, story.path), /*#__PURE__*/_react.default.createElement(_ui.Code, {
      size: 0
    }, story.why)))) : /*#__PURE__*/_react.default.createElement(_ui.Code, {
      size: 0
    }, "No matches"))),
    placement: "bottom-start",
    portal: true
  }, /*#__PURE__*/_react.default.createElement(DebugScoreCard, {
    padding: 1,
    shadow: 1,
    tone: tone
  }, /*#__PURE__*/_react.default.createElement(_ui.Code, {
    size: 0
  }, "score: ", score))));
}
