// @sanity/base/lib/components/previews/__workshop__/GeneralPreviewStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = GeneralPreviewStory;
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _uiWorkshop = require("@sanity/ui-workshop");
var _react = _interopRequireWildcard(require("react"));
var _DefaultPreview = require("../general/DefaultPreview");
var _DetailPreview = require("../general/DetailPreview");
var _MediaPreview = require("../general/MediaPreview");
var _constants = require("../constants");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var MEDIA_OPTIONS = {
  None: 'none',
  Image: 'image',
  Icon: 'icon',
  Text: 'text'
};
var LAYOUT_OPTIONS = {
  Default: 'default',
  Detail: 'detail',
  Media: 'media'
};
var previewComponents = {
  default: _DefaultPreview.DefaultPreview,
  detail: _DetailPreview.DetailPreview,
  media: _MediaPreview.MediaPreview
};
function GeneralPreviewStory() {
  var layout = (0, _uiWorkshop.useSelect)('Layout', LAYOUT_OPTIONS, 'default');
  var isPlaceholder = (0, _uiWorkshop.useBoolean)('Is placeholder', false);
  var interactive = (0, _uiWorkshop.useBoolean)('Interactive', false);
  var mediaKey = (0, _uiWorkshop.useSelect)('Media', MEDIA_OPTIONS) || 'none';
  var title = (0, _uiWorkshop.useString)('Title', 'Title', 'Props');
  var subtitle = (0, _uiWorkshop.useString)('Subtitle', 'Subtitle', 'Props');
  var description = (0, _uiWorkshop.useText)('Description', undefined, 'Props');
  var selected = (0, _uiWorkshop.useBoolean)('Selected', false, 'Props');
  var status = (0, _uiWorkshop.useBoolean)('Status', false);
  var progress = (0, _uiWorkshop.useNumber)('Progress (%)', 50);
  var media = (0, _react.useMemo)(() => {
    var _ref = _constants.PREVIEW_MEDIA_SIZE[layout] || _constants.PREVIEW_MEDIA_SIZE.default,
      width = _ref.width,
      height = _ref.height;
    if (mediaKey === 'image') {
      return /*#__PURE__*/_react.default.createElement("img", {
        src: "https://source.unsplash.com/".concat(width * 2, "x").concat(height * 2, "/?abstract")
      });
    }
    if (mediaKey === 'icon') {
      return /*#__PURE__*/_react.default.createElement(_icons.DocumentIcon, null);
    }
    if (mediaKey === 'text') {
      return 'T';
    }
    return false;
  }, [layout, mediaKey]);
  var component = previewComponents[layout];
  if (!component) {
    return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      align: "center",
      height: "fill",
      justify: "center",
      padding: 4,
      sizing: "border"
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, "Unknown layout: ", layout));
  }
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    height: "fill",
    tone: "transparent"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    height: "fill",
    justify: "center",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/_react.default.createElement(_ui.Container, {
    width: 0
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    "data-as": interactive ? 'button' : undefined,
    padding: 2,
    radius: 2,
    selected: interactive ? selected : undefined,
    style: {
      lineHeight: 0
    }
  }, /*#__PURE__*/(0, _react.createElement)(component, {
    description,
    isPlaceholder,
    media,
    progress,
    status: status && /*#__PURE__*/_react.default.createElement(_ui.Text, {
      muted: true,
      size: 1
    }, /*#__PURE__*/_react.default.createElement(_icons.EditIcon, null)),
    title,
    subtitle
  })))));
}
