// @sanity/base/lib/components/previews/__workshop__/TemplatePreviewStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TemplatePreviewStory;
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _uiWorkshop = require("@sanity/ui-workshop");
var _react = _interopRequireWildcard(require("react"));
var _TemplatePreview = require("../template/TemplatePreview");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var MEDIA_OPTIONS = {
  None: 'none',
  Image: 'image',
  Icon: 'icon'
};
var LAYOUT_OPTIONS = {
  'Create document': 'createDocument'
};
var mediaValues = {
  image: /*#__PURE__*/_react.default.createElement("img", {
    src: "https://source.unsplash.com/70x70/?abstract"
  }),
  icon: /*#__PURE__*/_react.default.createElement(_icons.DocumentIcon, null)
};
var previewComponents = {
  createDocument: _TemplatePreview.TemplatePreview
};
function TemplatePreviewStory() {
  var layout = (0, _uiWorkshop.useSelect)('Layout', LAYOUT_OPTIONS, 'createDocument');
  var isPlaceholder = (0, _uiWorkshop.useBoolean)('Is placeholder', false);
  var mediaKey = (0, _uiWorkshop.useSelect)('Media', MEDIA_OPTIONS) || 'none';
  var title = (0, _uiWorkshop.useString)('Title', 'Title', 'Props');
  var subtitle = (0, _uiWorkshop.useString)('Subtitle', 'Subtitle', 'Props');
  var description = (0, _uiWorkshop.useText)('Description', undefined, 'Props');
  var media = mediaValues[mediaKey] || false;
  var component = previewComponents[layout];
  if (!component) {
    return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      align: "center",
      height: "fill",
      justify: "center",
      padding: 4,
      sizing: "border"
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, "Unknown layout: ", layout));
  }
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    height: "fill",
    tone: "transparent"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    height: "fill",
    justify: "center",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/_react.default.createElement(_ui.Container, {
    width: 0
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    padding: 3,
    radius: 2
  }, /*#__PURE__*/(0, _react.createElement)(component, {
    description,
    isPlaceholder,
    media,
    title,
    subtitle
  })))));
}
