// @sanity/desk-tool/lib/components/pane/__workshop__/SplitPanesStory/ListPane/ListPane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListPane = ListPane;
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _react = _interopRequireWildcard(require("react"));
var _usePaneLayout2 = require("../../../usePaneLayout");
var _Pane = require("../../../Pane");
var _PaneContent = require("../../../PaneContent");
var _PaneHeader = require("../../../PaneHeader");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function ListPane(props) {
  var active = props.active,
    childId = props.childId,
    index = props.index,
    node = props.node,
    setPath = props.setPath;
  var _usePaneLayout = (0, _usePaneLayout2.usePaneLayout)(),
    layoutCollapsed = _usePaneLayout.collapsed;
  var handleBackClick = (0, _react.useCallback)(() => {
    setPath(p => p.slice(0, index));
  }, [index, setPath]);
  return /*#__PURE__*/_react.default.createElement(_Pane.Pane, {
    currentMaxWidth: 350,
    flex: 1,
    id: String(index),
    minWidth: 320,
    maxWidth: 640
  }, /*#__PURE__*/_react.default.createElement(_PaneHeader.PaneHeader, {
    actions: /*#__PURE__*/_react.default.createElement(_ui.Button, {
      icon: _icons.EllipsisVerticalIcon,
      mode: "bleed"
    }),
    backButton: index > 0 && layoutCollapsed && /*#__PURE__*/_react.default.createElement(_ui.Button, {
      icon: _icons.ArrowLeftIcon,
      mode: "bleed",
      onClick: handleBackClick
    }),
    title: node.title
  }), /*#__PURE__*/_react.default.createElement(_PaneContent.PaneContent, null, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    padding: 2,
    space: 1
  }, node.items.map(item => /*#__PURE__*/_react.default.createElement(_ui.Card, {
    as: "button",
    key: item.id,
    onClick: () => setPath(p => p.slice(0, index + 1).concat([item.id])),
    padding: 3,
    radius: 2,
    pressed: !active && childId === item.id,
    selected: active && childId === item.id
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, null, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, item.title)), /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, null, /*#__PURE__*/_react.default.createElement(_icons.ChevronRightIcon, null)))))))));
}
