// @sanity/portable-text-editor/lib/dts/src/patch/PatchEvent.js
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { flatten } from 'lodash';
import { prefixPath, set, unset, setIfMissing, insert, diffMatchPatch } from './patches';
var PatchEvent = /** @class */ (function () {
    function PatchEvent(patches) {
        this.patches = patches;
    }
    PatchEvent.from = function () {
        var patches = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            patches[_i] = arguments[_i];
        }
        return new PatchEvent(flatten(patches));
    };
    PatchEvent.prototype.prepend = function () {
        var patches = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            patches[_i] = arguments[_i];
        }
        return PatchEvent.from(__spreadArray(__spreadArray([], __read(flatten(patches)), false), __read(this.patches), false));
    };
    PatchEvent.prototype.append = function () {
        var patches = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            patches[_i] = arguments[_i];
        }
        return PatchEvent.from(__spreadArray(__spreadArray([], __read(this.patches), false), __read(flatten(patches)), false));
    };
    PatchEvent.prototype.prefixAll = function (segment) {
        return PatchEvent.from(this.patches.map(function (patch) { return prefixPath(patch, segment); }));
    };
    return PatchEvent;
}());
export default PatchEvent;
export { PatchEvent, set, unset, setIfMissing, insert, diffMatchPatch };
