// @sanity/structure/lib/Document.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocumentBuilder = void 0;
exports.documentFromEditor = documentFromEditor;
exports.documentFromEditorWithInitialValue = documentFromEditorWithInitialValue;
exports.getDefaultDocumentNode = getDefaultDocumentNode;
var _camelCase2 = _interopRequireDefault(require("lodash/camelCase"));
var _uniq2 = _interopRequireDefault(require("lodash/uniq"));
var _initialValueTemplates = require("@sanity/initial-value-templates");
var _SerializeError = require("./SerializeError");
var _validateId = require("./util/validateId");
var _View = require("./views/View");
var _views = require("./views");
var _userDefinedStructure = require("./userDefinedStructure");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
var resolveDocumentChild = (itemId, _ref) => {
  var params = _ref.params,
    path = _ref.path;
  var type = params.type;
  var parentPath = path.slice(0, path.length - 1);
  var currentSegment = path[path.length - 1];
  if (!type) {
    throw new _SerializeError.SerializeError("Invalid link. Your link must contain a `type`.", parentPath, currentSegment);
  }
  return getDefaultDocumentNode({
    documentId: itemId,
    schemaType: type
  });
};
class DocumentBuilder {
  constructor(spec) {
    _defineProperty(this, "spec", void 0);
    this.spec = spec ? spec : {};
  }
  id(id) {
    return this.clone({
      id
    });
  }
  getId() {
    return this.spec.id;
  }
  title(title) {
    return this.clone({
      title,
      id: this.spec.id || (0, _camelCase2.default)(title)
    });
  }
  getTitle() {
    return this.spec.title;
  }
  child(child) {
    return this.clone({
      child
    });
  }
  getChild() {
    return this.spec.child;
  }
  documentId(documentId) {
    // Let's try to be a bit helpful and assign an ID from document ID if none is specified
    var paneId = this.spec.id || documentId;
    return this.clone({
      id: paneId,
      options: _objectSpread(_objectSpread({}, this.spec.options || {}), {}, {
        id: documentId
      })
    });
  }
  getDocumentId() {
    return this.spec.options && this.spec.options.id;
  }
  schemaType(documentType) {
    return this.clone({
      options: _objectSpread(_objectSpread({}, this.spec.options || {}), {}, {
        type: typeof documentType === 'string' ? documentType : documentType.name
      })
    });
  }
  getSchemaType() {
    return this.spec.options && this.spec.options.type;
  }
  initialValueTemplate(templateId, parameters) {
    return this.clone({
      options: _objectSpread(_objectSpread({}, this.spec.options || {}), {}, {
        template: templateId,
        templateParameters: parameters
      })
    });
  }
  getInitalValueTemplate() {
    return this.spec.options && this.spec.options.template;
  }
  getInitialValueTemplateParameters() {
    return this.spec.options && this.spec.options.templateParameters;
  }
  views(views) {
    return this.clone({
      views
    });
  }
  getViews() {
    return this.spec.views || [];
  }
  serialize() {
    var _ref2 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
        path: []
      },
      _ref2$path = _ref2.path,
      path = _ref2$path === void 0 ? [] : _ref2$path,
      index = _ref2.index,
      hint = _ref2.hint;
    var urlId = path[index || path.length - 1];

    // Try to grab document ID / editor ID from URL if not defined
    var id = this.spec.id || urlId && "".concat(urlId) || '';
    var options = _objectSpread({
      id,
      type: undefined,
      template: undefined,
      templateParameters: undefined
    }, this.spec.options);
    if (typeof id !== 'string' || !id) {
      throw new _SerializeError.SerializeError('`id` is required for document nodes', path, index, hint).withHelpUrl(_SerializeError.HELP_URL.ID_REQUIRED);
    }
    if (!options || !options.id) {
      throw new _SerializeError.SerializeError('document id (`id`) is required for document nodes', path, id, hint).withHelpUrl(_SerializeError.HELP_URL.DOCUMENT_ID_REQUIRED);
    }
    if (!options || !options.type) {
      // eslint-disable-next-line no-console
      console.log("\u26A0\uFE0F Structure warning: document type (`schemaType`) will be required for document nodes in the near future! At:\n\n".concat(path.join(' > '), "\n"));
    }
    var views = (this.spec.views && this.spec.views.length > 0 ? this.spec.views : [(0, _views.form)()]).map((item, i) => (0, _View.maybeSerializeView)(item, i, path));
    var viewIds = views.map(view => view.id);
    var dupes = (0, _uniq2.default)(viewIds.filter((viewId, i) => viewIds.includes(viewId, i + 1)));
    if (dupes.length > 0) {
      throw new _SerializeError.SerializeError("document node has views with duplicate IDs: ".concat(dupes.join(',  ')), path, id, hint);
    }
    return _objectSpread(_objectSpread({}, this.spec), {}, {
      child: this.spec.child || resolveDocumentChild,
      id: (0, _validateId.validateId)(id, path, index),
      type: 'document',
      options: getDocumentOptions(options),
      views
    });
  }
  clone() {
    var withSpec = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    var builder = new DocumentBuilder();
    var options = _objectSpread(_objectSpread({}, this.spec.options || {}), withSpec.options || {});
    builder.spec = _objectSpread(_objectSpread(_objectSpread({}, this.spec), withSpec), {}, {
      options
    });
    return builder;
  }
}
exports.DocumentBuilder = DocumentBuilder;
function getDocumentOptions(spec) {
  var opts = {
    id: spec.id || '',
    type: spec.type || '*'
  };
  if (spec.template) {
    opts.template = spec.template;
  }
  if (spec.templateParameters) {
    opts.templateParameters = spec.templateParameters;
  }
  return opts;
}
function documentFromEditor(spec) {
  var doc = spec && spec.type ?
  // Use user-defined document fragment as base if possible
  getDefaultDocumentNode({
    schemaType: spec.type
  }) :
  // Fall back to plain old document builder
  new DocumentBuilder();
  if (spec) {
    var _spec$options = spec.options,
      id = _spec$options.id,
      type = _spec$options.type,
      template = _spec$options.template,
      templateParameters = _spec$options.templateParameters;
    doc = doc.id(spec.id).documentId(id);
    if (type) {
      doc = doc.schemaType(type);
    }
    if (template) {
      doc = doc.initialValueTemplate(template, templateParameters);
    }
    if (spec.child) {
      doc = doc.child(spec.child);
    }
  }
  return doc;
}
function documentFromEditorWithInitialValue(templateId, parameters) {
  var template = (0, _initialValueTemplates.getTemplateById)(templateId);
  if (!template) {
    throw new Error("Template with ID \"".concat(templateId, "\" not defined"));
  }
  return getDefaultDocumentNode({
    schemaType: template.schemaType
  }).initialValueTemplate(templateId, parameters);
}
function getDefaultDocumentNode(options) {
  var documentId = options.documentId,
    schemaType = options.schemaType;
  var userDefined = (0, _userDefinedStructure.getUserDefinedDefaultDocumentBuilder)(options);
  var builder = userDefined || new DocumentBuilder();
  if (!builder.getId()) {
    builder = builder.id('documentEditor');
  }
  if (documentId) {
    builder = builder.documentId(documentId.replace(/^drafts\./, ''));
  }
  return builder.schemaType(schemaType);
}
