// @sanity/base/lib/components/previews/portableText/BlockImagePreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BlockImagePreview = BlockImagePreview;
var _ui = require("@sanity/ui");
var _react = _interopRequireWildcard(require("react"));
var _useDevicePixelRatio = require("use-device-pixel-ratio");
var _Media = require("../_common/Media");
var _constants = require("../constants");
var _BlockImagePreview = require("./BlockImagePreview.styled");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
var DEFAULT_MEDIA_DIMENSIONS = _objectSpread(_objectSpread({}, _constants.PREVIEW_MEDIA_SIZE.blockImage), {}, {
  fit: 'fillmax',
  dpr: (0, _useDevicePixelRatio.getDevicePixelRatio)()
});
function BlockImagePreview(props) {
  var actions = props.actions,
    title = props.title,
    subtitle = props.subtitle,
    description = props.description,
    _props$fallbackTitle = props.fallbackTitle,
    fallbackTitle = _props$fallbackTitle === void 0 ? 'Untitled' : _props$fallbackTitle,
    _props$mediaDimension = props.mediaDimensions,
    mediaDimensions = _props$mediaDimension === void 0 ? DEFAULT_MEDIA_DIMENSIONS : _props$mediaDimension,
    media = props.media,
    children = props.children,
    status = props.status;
  var getRatio = (0, _react.useCallback)(dimensions => {
    var height = dimensions.height,
      width = dimensions.width;
    return height / width * 100;
  }, []);
  return /*#__PURE__*/_react.default.createElement(_BlockImagePreview.RootBox, null, /*#__PURE__*/_react.default.createElement(_ui.Stack, null, /*#__PURE__*/_react.default.createElement(_BlockImagePreview.HeaderFlex, {
    paddingLeft: 2,
    paddingRight: 1,
    paddingY: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    flex: 1,
    space: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    textOverflow: "ellipsis",
    weight: "semibold"
  }, title || fallbackTitle), subtitle && /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true,
    size: 1,
    textOverflow: "ellipsis"
  }, subtitle)), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    gap: 1,
    paddingLeft: 1
  }, status && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    paddingX: 2,
    paddingY: 3
  }, typeof status === 'function' ? status({
    layout: 'block'
  }) : status), actions)), /*#__PURE__*/_react.default.createElement(_BlockImagePreview.MediaCard, {
    $ratio: getRatio(mediaDimensions),
    __unstable_checkered: true,
    display: "flex",
    sizing: "border",
    tone: "inherit"
  }, /*#__PURE__*/_react.default.createElement(_Media.Media, {
    border: false,
    dimensions: mediaDimensions,
    layout: "blockImage",
    media: media,
    radius: 0,
    responsive: true
  }))), description && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    paddingX: 2,
    paddingY: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true,
    size: 1
  }, typeof description === 'function' ? description({
    layout: 'block'
  }) : description)), children && /*#__PURE__*/_react.default.createElement("div", null, children));
}
