// @sanity/desk-tool/lib/panes/document/statusBar/DocumentStatusBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocumentStatusBar = DocumentStatusBar;
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _ui = require("@sanity/ui");
var _useDocumentPane2 = require("../useDocumentPane");
var _DocumentStatusBarActions = require("./DocumentStatusBarActions");
var _DocumentSparkline = require("./sparkline/DocumentSparkline");
var _templateObject;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var DocumentActionsBox = (0, _styledComponents.default)(_ui.Box)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  min-width: 10em;\n  max-width: 16em;\n"])));
function DocumentStatusBar(props) {
  var actionsBoxRef = props.actionsBoxRef;
  var _useDocumentPane = (0, _useDocumentPane2.useDocumentPane)(),
    badges = _useDocumentPane.badges,
    historyController = _useDocumentPane.historyController;
  var showingRevision = historyController.onOlderRevision();
  return (0, _react.useMemo)(() => /*#__PURE__*/_react.default.createElement(_ui.Box, {
    paddingLeft: 2,
    paddingRight: [2, 3],
    paddingY: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: [1, 2]
  }, badges && /*#__PURE__*/_react.default.createElement(_DocumentSparkline.DocumentSparkline, null)), /*#__PURE__*/_react.default.createElement(DocumentActionsBox, {
    flex: 1,
    marginLeft: [1, 3],
    ref: actionsBoxRef
  }, showingRevision ? /*#__PURE__*/_react.default.createElement(_DocumentStatusBarActions.HistoryStatusBarActions, null) : /*#__PURE__*/_react.default.createElement(_DocumentStatusBarActions.DocumentStatusBarActions, null)))), [actionsBoxRef, badges, showingRevision]);
}
