// @sanity/form-builder/lib/inputs/ReferenceInput/ArrayItemReferenceInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ArrayItemReferenceInput = void 0;
var _react = _interopRequireWildcard(require("react"));
var _types = require("@sanity/types");
var _icons = require("@sanity/icons");
var _rxjs = require("rxjs");
var _operators = require("rxjs/operators");
var _ui = require("@sanity/ui");
var _components = require("@sanity/base/components");
var _presence = require("@sanity/base/presence");
var _internal = require("@sanity/base/_internal");
var _reactRx = require("react-rx");
var _uuid = require("@sanity/uuid");
var _autoId = require("@reach/auto-id");
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _PatchEvent = _interopRequireWildcard(require("../../PatchEvent"));
var _empty = require("../../utils/empty");
var _useDidUpdate = require("../../hooks/useDidUpdate");
var _isNonNullable = require("../../utils/isNonNullable");
var _AlertStrip = require("../../AlertStrip");
var _RowWrapper = require("../arrays/ArrayOfObjectsInput/item/components/RowWrapper");
var _DragHandle = require("../arrays/common/DragHandle");
var _randomKey = _interopRequireDefault(require("../arrays/common/randomKey"));
var _InsertMenu = require("../arrays/ArrayOfObjectsInput/InsertMenu");
var _useOnClickOutside = require("../../hooks/useOnClickOutside");
var _OptionPreview = require("./OptionPreview");
var _useReferenceInfo = require("./useReferenceInfo");
var _PreviewReferenceValue = require("./PreviewReferenceValue");
var _CreateButton = require("./CreateButton");
var _ReferenceAutocomplete = require("./ReferenceAutocomplete");
var _AutocompleteContainer = require("./AutocompleteContainer");
var _templateObject;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var StyledPreviewCard = (0, _styledComponents.default)(_components.PreviewCard)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  /* this is a hack to avoid layout jumps while previews are loading\n  there's probably better ways of solving this */\n  min-height: 36px;\n"])));
var INITIAL_SEARCH_STATE = {
  hits: [],
  isLoading: false
};
var NO_FILTER = () => true;
var dragHandle = /*#__PURE__*/_react.default.createElement(_DragHandle.DragHandle, {
  paddingX: 1,
  paddingY: 3
});
function valueHasRef(value) {
  return typeof value._ref === 'string';
}
var ArrayItemReferenceInput = /*#__PURE__*/(0, _react.forwardRef)(function ReferenceInput(props, forwardedRef) {
  var _loadableReferenceInf, _value$_strengthenOnP, _loadableReferenceInf2, _loadableReferenceInf3, _loadableReferenceInf4, _loadableReferenceInf5, _loadableReferenceInf6;
  var type = props.type,
    value = props.value,
    markers = props.markers,
    readOnly = props.readOnly,
    liveEdit = props.liveEdit,
    onSearch = props.onSearch,
    onChange = props.onChange,
    insertableTypes = props.insertableTypes,
    _props$focusPath = props.focusPath,
    focusPath = _props$focusPath === void 0 ? _empty.EMPTY_ARRAY : _props$focusPath,
    onFocus = props.onFocus,
    presence = props.presence,
    createOptions = props.createOptions,
    isSortable = props.isSortable,
    level = props.level,
    onBlur = props.onBlur,
    onInsert = props.onInsert,
    selectedState = props.selectedState,
    EditReferenceLink = props.editReferenceLinkComponent,
    onEditReference = props.onEditReference,
    getReferenceInfo = props.getReferenceInfo;
  var _useState = (0, _react.useState)(INITIAL_SEARCH_STATE),
    _useState2 = _slicedToArray(_useState, 2),
    searchState = _useState2[0],
    setSearchState = _useState2[1];
  var handleCreateNew = option => {
    var id = (0, _uuid.uuid)();
    var patches = [(0, _PatchEvent.setIfMissing)({}), (0, _PatchEvent.set)(type.name, ['_type']), (0, _PatchEvent.set)(id, ['_ref']), (0, _PatchEvent.set)(true, ['_weak']), (0, _PatchEvent.set)({
      type: option.type,
      weak: type.weak,
      template: option.template
    }, ['_strengthenOnPublish'])];
    onChange(_PatchEvent.default.from(patches));
    onEditReference({
      id,
      type: option.type,
      template: option.template
    });
    onFocus === null || onFocus === void 0 ? void 0 : onFocus([]);
  };
  var handleDuplicate = (0, _react.useCallback)(() => {
    onInsert === null || onInsert === void 0 ? void 0 : onInsert({
      item: _objectSpread(_objectSpread({}, value), {}, {
        _key: (0, _randomKey.default)()
      }),
      position: 'after',
      path: [{
        _key: value._key
      }],
      edit: false
    });
  }, [onInsert, value]);
  var handleInsert = (0, _react.useCallback)(pos => {
    onInsert === null || onInsert === void 0 ? void 0 : onInsert({
      item: {
        _type: type.name,
        _key: (0, _randomKey.default)()
      },
      position: pos,
      path: [{
        _key: value._key
      }]
    });
  }, [onInsert, type.name, value._key]);
  var handleChange = (0, _react.useCallback)(id => {
    if (!id) {
      onChange(_PatchEvent.default.from((0, _PatchEvent.unset)()));
      onFocus === null || onFocus === void 0 ? void 0 : onFocus([]);
      return;
    }
    var hit = searchState.hits.find(h => h.id === id);
    if (!hit) {
      throw new Error('Selected an item that wasnt part of the result set');
    }
    // if there's no published version of this document, set the reference to weak

    var patches = [(0, _PatchEvent.setIfMissing)({}), (0, _PatchEvent.set)(type.name, ['_type']), (0, _PatchEvent.set)((0, _internal.getPublishedId)(id), ['_ref']), hit.published && !type.weak ? (0, _PatchEvent.unset)(['_weak']) : (0, _PatchEvent.set)(true, ['_weak']), hit.published ? (0, _PatchEvent.unset)(['_strengthenOnPublish']) : (0, _PatchEvent.set)({
      type: hit === null || hit === void 0 ? void 0 : hit.type,
      weak: type.weak
    }, ['_strengthenOnPublish'])].filter(_isNonNullable.isNonNullable);
    onChange(_PatchEvent.default.from(patches));
    onFocus === null || onFocus === void 0 ? void 0 : onFocus([]);
  }, [searchState.hits, type.name, type.weak, onChange, onFocus]);
  var handleClear = (0, _react.useCallback)(() => {
    onChange(_PatchEvent.default.from((0, _PatchEvent.unset)()));
  }, [onChange]);
  var handlePreviewKeyPress = (0, _react.useCallback)(event => {
    if (event.key !== 'Enter' && event.key !== 'Space') {
      // enable "search for reference"-mode
      onFocus === null || onFocus === void 0 ? void 0 : onFocus(['_ref']);
    }
  }, [onFocus]);
  var handleCancelEdit = (0, _react.useCallback)(() => {
    if (!(value !== null && value !== void 0 && value._ref)) {
      onChange(_PatchEvent.default.from((0, _PatchEvent.unset)()));
    }
    onFocus === null || onFocus === void 0 ? void 0 : onFocus([]);
  }, [onChange, onFocus, value === null || value === void 0 ? void 0 : value._ref]);
  var handleAutocompleteKeyDown = (0, _react.useCallback)(event => {
    // escape
    if (event.keyCode === 27) {
      handleCancelEdit();
    }
  }, [handleCancelEdit]);
  var getReferenceInfoMemo = (0, _react.useCallback)(id => getReferenceInfo(id, type), [getReferenceInfo, type]);
  var loadableReferenceInfo = (0, _useReferenceInfo.useReferenceInfo)(value === null || value === void 0 ? void 0 : value._ref, getReferenceInfoMemo);
  var refTypeName = ((_loadableReferenceInf = loadableReferenceInfo.result) === null || _loadableReferenceInf === void 0 ? void 0 : _loadableReferenceInf.type) || (value === null || value === void 0 ? void 0 : (_value$_strengthenOnP = value._strengthenOnPublish) === null || _value$_strengthenOnP === void 0 ? void 0 : _value$_strengthenOnP.type);
  var refType = refTypeName ? type.to.find(toType => toType.name === refTypeName) : undefined;

  // --- focus handling
  var hasFocusAtRef = focusPath.length === 1 && (focusPath[0] === '_ref' || focusPath[0] === '$');
  var focusElementRef = (0, _ui.useForwardedRef)(forwardedRef);
  (0, _useDidUpdate.useDidUpdate)({
    hasFocusAt: hasFocusAtRef,
    ref: value === null || value === void 0 ? void 0 : value._ref
  }, (prev, current) => {
    var refUpdated = prev.ref !== current.ref;
    var focusAtUpdated = prev.hasFocusAt !== current.hasFocusAt;
    if ((focusAtUpdated || refUpdated) && current.hasFocusAt) {
      var _focusElementRef$curr;
      // if search mode changed and we're having focus always ensure the
      // ref element gets focus
      (_focusElementRef$curr = focusElementRef.current) === null || _focusElementRef$curr === void 0 ? void 0 : _focusElementRef$curr.focus();
    }
  });
  var weakIs = value !== null && value !== void 0 && value._weak ? 'weak' : 'strong';
  var weakShouldBe = type.weak === true ? 'weak' : 'strong';
  var hasRef = valueHasRef(value);
  // If the reference value is marked with _strengthenOnPublish,
  // we allow weak references if the reference points to a document that has a draft but not a published
  // In all other cases we should display a "weak mismatch" warning
  var weakWarningOverride = hasRef && !loadableReferenceInfo.isLoading && (value === null || value === void 0 ? void 0 : value._strengthenOnPublish);
  var handleFixStrengthMismatch = (0, _react.useCallback)(() => {
    onChange(_PatchEvent.default.from(type.weak === true ? (0, _PatchEvent.set)(true, ['_weak']) : (0, _PatchEvent.unset)(['_weak'])));
  }, [onChange, type]);
  var referenceExists = hasRef && ((_loadableReferenceInf2 = loadableReferenceInfo.result) === null || _loadableReferenceInf2 === void 0 ? void 0 : (_loadableReferenceInf3 = _loadableReferenceInf2.preview) === null || _loadableReferenceInf3 === void 0 ? void 0 : (_loadableReferenceInf4 = _loadableReferenceInf3.published) === null || _loadableReferenceInf4 === void 0 ? void 0 : _loadableReferenceInf4._id);
  var handleRemoveStrengthenOnPublish = (0, _react.useCallback)(() => {
    onChange(_PatchEvent.default.from([type.weak === true ? (0, _PatchEvent.set)(true, ['_weak']) : (0, _PatchEvent.unset)(['_weak']), (0, _PatchEvent.unset)(['_strengthenOnPublish'])]));
  }, [onChange, type]);
  var _useToast = (0, _ui.useToast)(),
    push = _useToast.push;
  var validation = (0, _react.useMemo)(() => markers.filter(_types.isValidationMarker), [markers]);
  var errors = (0, _react.useMemo)(() => validation.filter(_types.isValidationErrorMarker), [validation]);
  var pressed = selectedState === 'pressed';
  var selected = selectedState === 'selected';
  var handleFocus = (0, _react.useCallback)(event => {
    if (onFocus && event.currentTarget === focusElementRef.current) {
      onFocus([]);
    }
  }, [onFocus, focusElementRef]);
  var handleAutocompleteFocus = (0, _react.useCallback)(event => {
    if (onFocus && event.currentTarget === focusElementRef.current) {
      onFocus(['_ref']);
    }
  }, [onFocus, focusElementRef]);
  var handleQueryChange = (0, _reactRx.useObservableCallback)(inputValue$ => {
    return inputValue$.pipe((0, _operators.filter)(_isNonNullable.isNonNullable), (0, _operators.distinctUntilChanged)(), (0, _operators.switchMap)(searchString => (0, _rxjs.concat)((0, _rxjs.of)({
      isLoading: true
    }), onSearch(searchString).pipe((0, _operators.map)(hits => ({
      hits,
      isLoading: false,
      searchString
    })), (0, _operators.catchError)(error => {
      push({
        title: 'Reference search failed',
        description: error.message,
        status: 'error',
        id: "reference-search-fail-".concat(inputId)
      });
      console.error(error);
      return (0, _rxjs.of)({
        hits: []
      });
    })))), (0, _operators.scan)((prevState, nextState) => _objectSpread(_objectSpread({}, prevState), nextState), INITIAL_SEARCH_STATE), (0, _operators.tap)(setSearchState));
  }, []);
  var handleAutocompleteOpenButtonClick = (0, _react.useCallback)(() => {
    handleQueryChange('');
  }, [handleQueryChange]);
  var showWeakRefMismatch = !loadableReferenceInfo.isLoading && hasRef && weakIs !== weakShouldBe && !weakWarningOverride;
  var inputId = (0, _autoId.useId)();
  var handleCreateButtonKeyDown = (0, _react.useCallback)(e => {
    if (e.key === 'Escape') {
      var _focusElementRef$curr2;
      (_focusElementRef$curr2 = focusElementRef.current) === null || _focusElementRef$curr2 === void 0 ? void 0 : _focusElementRef$curr2.focus();
    }
  }, [focusElementRef]);
  var renderOption = (0, _react.useCallback)(option => {
    var _option$hit$draft, _option$hit$published;
    var id = ((_option$hit$draft = option.hit.draft) === null || _option$hit$draft === void 0 ? void 0 : _option$hit$draft._id) || ((_option$hit$published = option.hit.published) === null || _option$hit$published === void 0 ? void 0 : _option$hit$published._id);
    return /*#__PURE__*/_react.default.createElement(_components.PreviewCard, {
      as: "button",
      type: "button",
      radius: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      paddingX: 3,
      paddingY: 1
    }, /*#__PURE__*/_react.default.createElement(_OptionPreview.OptionPreview, {
      getReferenceInfo: getReferenceInfoMemo,
      id: id,
      type: type
    })));
  }, [type, getReferenceInfoMemo]);
  var OpenLink = (0, _react.useMemo)(() =>
  /*#__PURE__*/
  // eslint-disable-next-line @typescript-eslint/no-shadow
  (0, _react.forwardRef)(function OpenLink(restProps, _ref) {
    var template = ((value === null || value === void 0 ? void 0 : value._strengthenOnPublish) || {}).template;
    return /*#__PURE__*/_react.default.createElement(_components.IntentLink, _extends({}, restProps, {
      intent: "edit",
      params: [{
        id: value === null || value === void 0 ? void 0 : value._ref,
        type: refType === null || refType === void 0 ? void 0 : refType.name,
        template: template === null || template === void 0 ? void 0 : template.id
      }, {
        params: template === null || template === void 0 ? void 0 : template.params
      }],
      target: "_blank",
      rel: "noopener noreferrer",
      ref: _ref
    }));
  }), [refType === null || refType === void 0 ? void 0 : refType.name, value === null || value === void 0 ? void 0 : value._ref, value === null || value === void 0 ? void 0 : value._strengthenOnPublish]);
  var preview = ((_loadableReferenceInf5 = loadableReferenceInfo.result) === null || _loadableReferenceInf5 === void 0 ? void 0 : _loadableReferenceInf5.preview.draft) || ((_loadableReferenceInf6 = loadableReferenceInfo.result) === null || _loadableReferenceInf6 === void 0 ? void 0 : _loadableReferenceInf6.preview.published);
  var isEditing = hasFocusAtRef;

  // --- click outside handling
  var clickOutsideBoundaryRef = (0, _react.useRef)();
  var autocompletePortalRef = (0, _react.useRef)();
  var createButtonMenuPortalRef = (0, _react.useRef)();
  (0, _useOnClickOutside.useOnClickOutside)([clickOutsideBoundaryRef, autocompletePortalRef, createButtonMenuPortalRef], () => {
    if (isEditing) {
      handleCancelEdit();
    }
  });
  var autocompletePopoverReferenceElementRef = (0, _react.useRef)();
  return /*#__PURE__*/_react.default.createElement(_RowWrapper.RowWrapper, {
    radius: 2,
    padding: 1,
    tone: isEditing ? ' default' : readOnly ? 'transparent' : loadableReferenceInfo.error || errors.length > 0 ? 'critical' : showWeakRefMismatch ? 'caution' : 'default'
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, !isEditing && isSortable && !readOnly && /*#__PURE__*/_react.default.createElement(_ui.Card, {
    className: "dragHandle",
    tone: "inherit",
    marginRight: 1
  }, dragHandle), isEditing ? /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1,
    padding: 1,
    ref: clickOutsideBoundaryRef
  }, /*#__PURE__*/_react.default.createElement(_components.FormField, {
    __unstable_markers: markers,
    __unstable_presence: presence,
    __unstable_changeIndicator: false,
    inputId: inputId,
    title: type.title,
    level: level,
    description: type.description
  }, /*#__PURE__*/_react.default.createElement(_AutocompleteContainer.AutocompleteContainer, {
    ref: autocompletePopoverReferenceElementRef
  }, /*#__PURE__*/_react.default.createElement(_ReferenceAutocomplete.ReferenceAutocomplete, {
    "data-testid": "autocomplete",
    loading: searchState.isLoading,
    portalRef: autocompletePortalRef,
    ref: focusElementRef,
    id: inputId || '',
    options: searchState.hits.map(hit => ({
      value: hit.id,
      hit: hit
    })),
    onFocus: handleAutocompleteFocus,
    onBlur: onBlur,
    radius: 1,
    placeholder: "Type to search",
    onKeyDown: handleAutocompleteKeyDown,
    readOnly: readOnly,
    disabled: loadableReferenceInfo.isLoading,
    onQueryChange: handleQueryChange,
    searchString: searchState.searchString,
    onChange: handleChange,
    referenceElement: autocompletePopoverReferenceElementRef.current,
    filterOption: NO_FILTER,
    renderOption: renderOption,
    openButton: {
      onClick: handleAutocompleteOpenButtonClick
    }
  }), !readOnly && createOptions.length > 0 && /*#__PURE__*/_react.default.createElement(_CreateButton.CreateButton, {
    menuRef: createButtonMenuPortalRef,
    id: "".concat(inputId, "-selectTypeMenuButton"),
    createOptions: createOptions,
    onCreate: handleCreateNew,
    onKeyDown: handleCreateButtonKeyDown
  })))) : /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center"
  }, hasRef ? /*#__PURE__*/_react.default.createElement(StyledPreviewCard, {
    flex: 1,
    padding: 1,
    paddingRight: 3,
    radius: 2,
    forwardedAs: EditReferenceLink,
    documentId: value === null || value === void 0 ? void 0 : value._ref,
    documentType: refType === null || refType === void 0 ? void 0 : refType.name,
    "data-as": "a",
    tone: selected ? 'default' : 'inherit',
    __unstable_focusRing: true,
    tabIndex: 0,
    selected: selected,
    pressed: pressed,
    onKeyPress: handlePreviewKeyPress,
    onFocus: handleFocus,
    "data-selected": selected ? true : undefined,
    "data-pressed": pressed ? true : undefined,
    ref: focusElementRef
  }, /*#__PURE__*/_react.default.createElement(_PreviewReferenceValue.PreviewReferenceValue, {
    value: value,
    referenceInfo: loadableReferenceInfo,
    type: type
  })) : /*#__PURE__*/_react.default.createElement(_ui.Card, {
    flex: 1,
    padding: 3,
    marginRight: 1,
    radius: 2,
    tone: "inherit",
    as: "button",
    __unstable_focusRing: true,
    tabIndex: 0,
    onClick: () => onFocus === null || onFocus === void 0 ? void 0 : onFocus(['_ref']),
    ref: focusElementRef
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginY: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true
  }, "Empty reference"))), /*#__PURE__*/_react.default.createElement(_ui.Inline, {
    marginLeft: !readOnly && presence.length > 0 ? 2 : undefined
  }, !readOnly && presence.length > 0 && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_presence.FieldPresence, {
    presence: presence,
    maxAvatars: 1
  })), validation.length > 0 && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1,
    paddingX: 1,
    paddingY: 3
  }, /*#__PURE__*/_react.default.createElement(_components.FormFieldValidationStatus, {
    __unstable_markers: validation
  })), !value._key && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
    content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      muted: true,
      size: 1
    }, "This item is missing the required ", /*#__PURE__*/_react.default.createElement("code", null, "_key"), " property.")),
    placement: "top"
  }, /*#__PURE__*/_react.default.createElement(_ui.Badge, {
    mode: "outline",
    tone: "caution"
  }, "Missing key")))))), !isEditing && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
    button: /*#__PURE__*/_react.default.createElement(_ui.Button, {
      paddingY: 3,
      paddingX: 2,
      mode: "bleed",
      icon: _icons.EllipsisVerticalIcon
    }),
    id: "".concat(inputId, "-menuButton"),
    menu: /*#__PURE__*/_react.default.createElement(_ui.Menu, null, !readOnly && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      text: "Remove",
      tone: "critical",
      icon: _icons.TrashIcon,
      onClick: handleClear
    }), /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      text: "Replace",
      icon: _icons.SyncIcon,
      onClick: () => {
        onFocus === null || onFocus === void 0 ? void 0 : onFocus(['_ref']);
      }
    }), /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      text: "Duplicate",
      icon: _icons.CopyIcon,
      onClick: handleDuplicate
    }), /*#__PURE__*/_react.default.createElement(_InsertMenu.InsertMenu, {
      onInsert: handleInsert,
      types: insertableTypes
    })), !readOnly && hasRef && /*#__PURE__*/_react.default.createElement(_ui.MenuDivider, null), hasRef && /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      as: OpenLink,
      "data-as": "a",
      text: "Open in new tab",
      icon: _icons.LaunchIcon
    })),
    placement: "right",
    popover: {
      portal: true,
      tone: 'default'
    }
  }))), liveEdit && referenceExists && value._strengthenOnPublish && /*#__PURE__*/_react.default.createElement(_AlertStrip.AlertStrip, {
    padding: 1,
    title: type.weak ? 'Finalize reference' : 'Convert to strong reference',
    status: "info",
    "data-testid": "alert-reference-published"
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, /*#__PURE__*/_react.default.createElement("strong", null, loadableReferenceInfo.result.preview.published.title), " is published and this reference should now be", ' ', type.weak ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "finalized") : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "converted to a strong reference"), "."), /*#__PURE__*/_react.default.createElement(_ui.Button, {
    onClick: handleRemoveStrengthenOnPublish,
    text: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Convert to strong reference"),
    tone: "positive"
  }))), showWeakRefMismatch && /*#__PURE__*/_react.default.createElement(_AlertStrip.AlertStrip, {
    padding: 1,
    title: "Reference strength mismatch",
    status: "warning",
    "data-testid": "alert-reference-strength-mismatch"
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, "This reference is ", /*#__PURE__*/_react.default.createElement("em", null, weakIs), ", but according to the current schema it should be", ' ', /*#__PURE__*/_react.default.createElement("em", null, weakShouldBe, ".")), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, type.weak ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "It will not be possible to delete the \"", preview === null || preview === void 0 ? void 0 : preview.title, "\"-document without first removing this reference.") : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "This makes it possible to delete the \"", preview === null || preview === void 0 ? void 0 : preview.title, "\"-document without first deleting this reference, leaving this field referencing a nonexisting document.")), /*#__PURE__*/_react.default.createElement(_ui.Button, {
    onClick: handleFixStrengthMismatch,
    text: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Convert to ", weakShouldBe, " reference"),
    tone: "caution"
  }))), loadableReferenceInfo.error && /*#__PURE__*/_react.default.createElement(_AlertStrip.AlertStrip, {
    padding: 1,
    title: "Unable to load reference metadata",
    status: "warning",
    "data-testid": "alert-reference-info-failed"
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true,
    size: 1
  }, "Error: ", loadableReferenceInfo.error.message), /*#__PURE__*/_react.default.createElement(_ui.Button, {
    onClick: loadableReferenceInfo.retry,
    text: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Retry"),
    tone: "primary"
  }))));
});
exports.ArrayItemReferenceInput = ArrayItemReferenceInput;
