// @sanity/desk-tool/lib/panes/document/statusBar/__workshop__/ReviewChangesButtonStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ReviewChangesButtonStory;
var _ui = require("@sanity/ui");
var _uiWorkshop = require("@sanity/ui-workshop");
var _react = _interopRequireDefault(require("react"));
var _ReviewChangesButton = require("../sparkline/ReviewChangesButton");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var STATUS_OPTIONS = {
  changes: 'changes',
  saved: 'saved',
  syncing: 'syncing'
};
function ReviewChangesButtonStory() {
  var collapsed = (0, _uiWorkshop.useBoolean)('Collapsed', false);
  var status = (0, _uiWorkshop.useSelect)('Status', STATUS_OPTIONS) || 'changes';
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    height: "fill"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    height: "fill",
    justify: "center",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/_react.default.createElement(_ReviewChangesButton.ReviewChangesButton, {
    status: status,
    lastUpdated: "just now",
    collapsed: collapsed
  })));
}
