// @sanity/form-builder/lib/inputs/PortableText/__workshop__/customSchema/schema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.schema = exports.portableTextType = exports.blockType = void 0;
var _schema = _interopRequireDefault(require("@sanity/schema"));
var _schemaType = _interopRequireDefault(require("all:part:@sanity/base/schema-type"));
var _baseTypes = require("../_common/baseTypes");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var imageType = {
  type: 'image',
  name: 'blockImage'
};
var linkType = {
  type: 'object',
  name: 'link',
  fields: [{
    type: 'string',
    name: 'href'
  }]
};
var someAnnotationType = {
  type: 'object',
  name: 'someAnnotation',
  options: {
    modal: {
      size: 'medium'
    }
  },
  fields: [{
    type: 'string',
    name: 'title'
  }]
};
var someObject = {
  type: 'object',
  name: 'someObject',
  options: {
    // editModal: 'fullscreen',
    modal: {
      type: 'dialog',
      size: 'full'
    }
  },
  fields: [{
    type: 'string',
    name: 'title'
  }]
};
function extractTextFromBlocks(blocks) {
  if (!blocks) {
    return '';
  }
  return blocks.filter(val => val._type === 'block').map(block => {
    return block.children.filter(child => child._type === 'span').map(span => span.text).join('');
  }).join('');
}
var blockType = {
  type: 'block',
  name: 'myBlockType',
  styles: [{
    title: 'Paragraph',
    value: 'normal'
  }, {
    title: 'Heading 1',
    value: 'h1'
  }, {
    title: 'Heading 2',
    value: 'h2'
  }, {
    title: 'Heading 3',
    value: 'h3'
  }, {
    title: 'Heading 4',
    value: 'h4'
  }, {
    title: 'Heading 5',
    value: 'h5'
  }, {
    title: 'Heading 6',
    value: 'h6'
  }, {
    title: 'Quotation',
    value: 'blockquote'
  }],
  marks: {
    annotations: [someAnnotationType, linkType]
  },
  of: [someObject, imageType],
  validation: rule => rule.custom(block => {
    var length = extractTextFromBlocks([block]).length;
    return length < 10 ? 'Please write a longer paragraph.' : null;
  }).error()
};
exports.blockType = blockType;
var ptType = {
  type: 'array',
  name: 'body',
  of: [blockType, someObject, imageType]
};
var schema = _schema.default.compile({
  name: 'default',
  types: _schemaType.default.concat([imageType, someObject, ptType]).concat(_baseTypes.baseTypes)
});
exports.schema = schema;
var portableTextType = schema.get('body');
exports.portableTextType = portableTextType;
