// @sanity/base/lib/datastores/authState/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "authStateChangedInThisWindow$", {
  enumerable: true,
  get: function get() {
    return _state.authStateChangedInThisWindow$;
  }
});
Object.defineProperty(exports, "authToken$", {
  enumerable: true,
  get: function get() {
    return _token.authToken$;
  }
});
Object.defineProperty(exports, "authTokenIsAllowed", {
  enumerable: true,
  get: function get() {
    return _config.authTokenIsAllowed;
  }
});
Object.defineProperty(exports, "broadcastAuthStateChanged", {
  enumerable: true,
  get: function get() {
    return _state.broadcastAuthStateChanged;
  }
});
Object.defineProperty(exports, "clearToken", {
  enumerable: true,
  get: function get() {
    return _token.clearToken;
  }
});
Object.defineProperty(exports, "fetchToken", {
  enumerable: true,
  get: function get() {
    return _token.fetchToken;
  }
});
Object.defineProperty(exports, "getToken", {
  enumerable: true,
  get: function get() {
    return _token.getToken;
  }
});
Object.defineProperty(exports, "saveToken", {
  enumerable: true,
  get: function get() {
    return _token.saveToken;
  }
});
var _config = require("./config");
var _token = require("./token");
var _state = require("./state");
