// @sanity/base/lib/datastores/authState/state.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.broadcastAuthStateChanged = exports.authStateChangedInThisWindow$ = exports.authStateChangedInOtherWindow$ = void 0;
var _rxjs = require("rxjs");
var _operators = require("rxjs/operators");
var _crossWindowMessaging = require("../crossWindowMessaging");
var MSG_AUTH_STATE_CHANGED = 'authStateChange';
var authStateChangedInThisWindow$ = new _rxjs.Subject();
exports.authStateChangedInThisWindow$ = authStateChangedInThisWindow$;
var authStateChangedInOtherWindow$ = _crossWindowMessaging.otherWindowMessages$.pipe((0, _operators.filter)(isAuthChangeMessage));
exports.authStateChangedInOtherWindow$ = authStateChangedInOtherWindow$;
var broadcastAuthStateChanged = userId => {
  var message = {
    type: MSG_AUTH_STATE_CHANGED,
    id: userId
  };
  (0, _crossWindowMessaging.crossWindowBroadcast)(message);
  authStateChangedInThisWindow$.next(true);
};
exports.broadcastAuthStateChanged = broadcastAuthStateChanged;
function isAuthChangeMessage(message) {
  return typeof message === 'object' && 'type' in message && message.type === MSG_AUTH_STATE_CHANGED;
}
