// @sanity/default-layout/lib/update/__workshop__/ChangelogDialogStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ChangelogDialogStory;
var _uiWorkshop = require("@sanity/ui-workshop");
var _react = _interopRequireDefault(require("react"));
var _ = require("..");
var _mockData = require("./_mock/mockData");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function ChangelogDialogStory() {
  var withUpgradeAccordion = (0, _uiWorkshop.useBoolean)('Show upgrade accordion', true);
  var _ref = _mockData.CHANGELOG_MOCK_DATA,
    changelog = _ref.changelog,
    currentVersion = _ref.currentVersion,
    latestVersion = _ref.latestVersion;
  return /*#__PURE__*/_react.default.createElement(_.ChangelogDialog, {
    changelog: changelog,
    currentVersion: currentVersion,
    latestVersion: latestVersion,
    dialogProps: {
      onClose: () => null,
      footer: withUpgradeAccordion && /*#__PURE__*/_react.default.createElement(_.UpgradeAccordion, null)
    }
  });
}
