// @sanity/base/lib/datastores/grants/debug/exampleRoles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.exampleRoles = void 0;
exports.getDebugRolesByNames = getDebugRolesByNames;
var exampleRoles = {
  // basic
  administrator: {
    name: 'administrator',
    title: 'Administrator'
  },
  viewer: {
    name: 'Viewer',
    title: 'Viewer'
  },
  editor: {
    name: 'editor',
    title: 'Editor'
  },
  developer: {
    name: 'developer',
    title: 'Developer'
  },
  contributor: {
    name: 'contributor',
    title: 'Contributor'
  },
  // custom
  restricted: {
    name: 'restricted',
    title: 'Restricted'
  },
  requiresApproval: {
    name: 'requiresApproval',
    title: 'Requires approval'
  },
  // legacy
  read: {
    name: 'read',
    title: 'Read'
  },
  write: {
    name: 'write',
    title: 'Write'
  }
};
exports.exampleRoles = exampleRoles;
function getDebugRolesByNames(roleNames) {
  return roleNames.filter(roleName => roleName in exampleRoles).map(roleName => exampleRoles[roleName]);
}
