// @sanity/form-builder/lib/inputs/CrossDatasetReferenceInput/DisabledFeatureWarning.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DisabledFeatureWarning = DisabledFeatureWarning;
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _templateObject;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var ButtonWrapper = (0, _styledComponents.default)(_ui.Button)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  width: 100%;\n"])));
function DisabledFeatureWarning(_ref) {
  var value = _ref.value,
    onClearValue = _ref.onClearValue;
  var hasRef = Boolean(value === null || value === void 0 ? void 0 : value._ref);
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    tone: "caution",
    padding: 4,
    border: true,
    radius: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    gap: 4,
    marginBottom: hasRef ? 4 : undefined
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, /*#__PURE__*/_react.default.createElement(_icons.WarningOutlineIcon, null))), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "h2",
    size: 1,
    weight: "semibold"
  }, "Unavailable feature: Cross dataset reference"), hasRef && /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    size: 1
  }, "This feature has been disabled. Read how to enable it in", ' ', /*#__PURE__*/_react.default.createElement("a", {
    href: "https://www.sanity.io/docs/cross-dataset-references",
    target: "_blank",
    rel: "noreferrer"
  }, "documentation")), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    size: 1
  }, "You can still clear this field's existing reference, but that cannot be revoked as long as the feature is disabled.")), !hasRef && /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    size: 1
  }, "Read how to enable it in", ' ', /*#__PURE__*/_react.default.createElement("a", {
    href: "https://www.sanity.io/docs/cross-dataset-references",
    target: "_blank",
    rel: "noreferrer"
  }, "documentation")))), onClearValue && hasRef && /*#__PURE__*/_react.default.createElement(ButtonWrapper, {
    icon: _icons.ResetIcon,
    text: "Reset value",
    onClick: onClearValue,
    mode: "ghost"
  }));
}
