// @sanity/default-layout/lib/navbar/search/contexts/search/reducer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.initialSearchState = initialSearchState;
exports.searchReducer = searchReducer;
var _types = require("../../types");
var _debug = require("../../utils/debug");
var _isRecentSearchTerms = require("../../utils/isRecentSearchTerms");
var _selectors = require("../../utils/selectors");
var _excluded = ["__recent"];
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function initialSearchState(currentUser, recentSearches) {
  return {
    currentUser,
    debug: (0, _debug.isDebugMode)(),
    filtersVisible: false,
    ordering: _types.ORDER_RELEVANCE,
    pageIndex: 0,
    recentSearches: recentSearches || [],
    result: {
      error: null,
      hasMore: null,
      hits: [],
      loaded: false,
      loading: false
    },
    terms: {
      query: '',
      types: []
    }
  };
}
var debug = (0, _debug.debugWithName)('searchReducer');
function searchReducer(state, action) {
  var prefix = 'ðŸ”';
  if (action.type.startsWith('SEARCH_REQUEST')) {
    prefix = 'ðŸš¨';
  }
  if (action.type.startsWith('RECENT_SEARCHES')) {
    prefix = 'ðŸ’¾';
  }
  debug(prefix, action);
  switch (action.type) {
    case 'FILTERS_HIDE':
      return _objectSpread(_objectSpread({}, state), {}, {
        filtersVisible: false
      });
    case 'FILTERS_SHOW':
      return _objectSpread(_objectSpread({}, state), {}, {
        filtersVisible: true
      });
    case 'FILTERS_TOGGLE':
      return _objectSpread(_objectSpread({}, state), {}, {
        filtersVisible: !state.filtersVisible
      });
    case 'PAGE_INCREMENT':
      return _objectSpread(_objectSpread({}, state), {}, {
        pageIndex: state.pageIndex + 1,
        terms: stripRecent(state.terms)
      });
    case 'RECENT_SEARCHES_SET':
      return _objectSpread(_objectSpread({}, state), {}, {
        recentSearches: action.recentSearches
      });
    case 'SEARCH_CLEAR':
      return _objectSpread(_objectSpread({}, state), {}, {
        pageIndex: 0,
        result: _objectSpread(_objectSpread({}, state.result), {}, {
          hasMore: null,
          hits: []
        })
      });
    case 'SEARCH_ORDERING_RESET':
      return _objectSpread(_objectSpread({}, state), {}, {
        ordering: _types.ORDER_RELEVANCE,
        terms: stripRecent(state.terms)
      });
    case 'SEARCH_ORDERING_SET':
      return _objectSpread(_objectSpread({}, state), {}, {
        ordering: action.ordering,
        terms: stripRecent(state.terms)
      });
    case 'SEARCH_REQUEST_COMPLETE':
      return _objectSpread(_objectSpread({}, state), {}, {
        result: _objectSpread(_objectSpread({}, state.result), {}, {
          error: null,
          hasMore: action.hits.length > 0,
          hits: state.pageIndex > 0 ? [...state.result.hits, ...action.hits] : action.hits,
          loaded: true,
          loading: false
        })
      });
    case 'SEARCH_REQUEST_ERROR':
      return _objectSpread(_objectSpread({}, state), {}, {
        result: _objectSpread(_objectSpread({}, state.result), {}, {
          error: action.error,
          loaded: false,
          loading: false
        })
      });
    case 'SEARCH_REQUEST_START':
      return _objectSpread(_objectSpread({}, state), {}, {
        result: _objectSpread(_objectSpread({}, state.result), {}, {
          loaded: false,
          loading: true
        })
      });
    case 'TERMS_QUERY_SET':
      return _objectSpread(_objectSpread({}, state), {}, {
        pageIndex: 0,
        result: _objectSpread(_objectSpread({}, state.result), {}, {
          loaded: false
        }),
        terms: stripRecent(_objectSpread(_objectSpread({}, state.terms), {}, {
          query: action.query
        }))
      });
    case 'TERMS_SET':
      return _objectSpread(_objectSpread({}, state), {}, {
        pageIndex: 0,
        result: _objectSpread(_objectSpread({}, state.result), {}, {
          loaded: false
        }),
        terms: action.terms
      });
    case 'TERMS_TYPE_ADD':
      return _objectSpread(_objectSpread({}, state), {}, {
        pageIndex: 0,
        result: _objectSpread(_objectSpread({}, state.result), {}, {
          loaded: false
        }),
        terms: stripRecent(_objectSpread(_objectSpread({}, state.terms), {}, {
          types: [...state.terms.types, action.schemaType].sort(_selectors.sortTypes)
        }))
      });
    case 'TERMS_TYPE_REMOVE':
      return _objectSpread(_objectSpread({}, state), {}, {
        pageIndex: 0,
        result: _objectSpread(_objectSpread({}, state.result), {}, {
          loaded: false
        }),
        terms: stripRecent(_objectSpread(_objectSpread({}, state.terms), {}, {
          types: state.terms.types.filter(s => s !== action.schemaType)
        }))
      });
    case 'TERMS_TYPES_CLEAR':
      return _objectSpread(_objectSpread({}, state), {}, {
        pageIndex: 0,
        result: _objectSpread(_objectSpread({}, state.result), {}, {
          loaded: false
        }),
        terms: stripRecent(_objectSpread(_objectSpread({}, state.terms), {}, {
          types: []
        }))
      });
    default:
      return state;
  }
}

/**
 * This function is used to strip __recent from terms, generally whenever there's a change in
 * search terms or options that would otherwise trigger an additional search request.
 * (e.g. updating the search query, changing a sort filter, adding / removing document types)
 *
 * This is done so we can better disambiguate between requests sent as a result of clicking a 'recent search'
 * for purposes of measurement.
 *
 * @todo remove this (and associated tests) once client-side instrumentation is available
 */
function stripRecent(terms) {
  if ((0, _isRecentSearchTerms.isRecentSearchTerms)(terms)) {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    var __recent = terms.__recent,
      rest = _objectWithoutProperties(terms, _excluded);
    return rest;
  }
  return terms;
}
