// @sanity/form-builder/lib/__workshop__/_common/schema/simpleDummySchema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = getSchema;
var _data = require("../data");
function getSchema(props) {
  var _props$hiddenGroup = props.hiddenGroup,
    hiddenGroup = _props$hiddenGroup === void 0 ? false : _props$hiddenGroup;
  return (0, _data.wrapSchema)({
    name: 'dummy',
    type: 'document',
    groups: [{
      name: 'group1',
      title: 'Group 1',
      hidden: hiddenGroup
    }],
    fields: [{
      name: 'title',
      title: 'Title',
      type: 'string'
    }, {
      name: 'person',
      type: 'object',
      group: ['group1'],
      fieldsets: [{
        name: 'social',
        title: 'Social media handles [collapsed by default]',
        options: {
          collapsible: true,
          collapsed: true
        }
      }],
      groups: [{
        name: 'instagram',
        title: 'Instagram'
      }],
      fields: [{
        name: 'name',
        title: 'Name',
        type: 'string'
      }, {
        name: 'conditionalName',
        title: 'Name (conditional read only)',
        type: 'string',
        readOnly: _ref => {
          var parent = _ref.parent;
          return (parent === null || parent === void 0 ? void 0 : parent.name) === 'ro';
        }
      }, {
        name: 'twitter',
        title: 'Twitter',
        type: 'string',
        fieldset: 'social',
        validation: Rule => Rule.required()
      }, {
        name: 'instagram',
        title: 'Instagram',
        type: 'string',
        fieldset: 'social',
        group: ['instagram']
      }, {
        name: 'facebook',
        title: 'Facebook',
        type: 'string',
        fieldset: 'social'
      }]
    }]
  });
}
