// @sanity/default-layout/lib/schemaErrors/InitialValueTemplateError.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = InitialValueTemplateError;
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireDefault(require("react"));
var _ui = require("@sanity/ui");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function InitialValueTemplateError(_ref) {
  var errors = _ref.errors;
  return /*#__PURE__*/_react.default.createElement(_ui.Dialog, {
    header: "Initial value template error",
    id: "initial-value-error-dialog",
    width: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, "Failed to load initial value templates:"), errors.map(error => /*#__PURE__*/_react.default.createElement(_ui.Text, {
    key: error.message
  }, /*#__PURE__*/_react.default.createElement("code", null, error.message))))));
}
InitialValueTemplateError.propTypes = {
  errors: _propTypes.default.arrayOf(_propTypes.default.shape({
    message: _propTypes.default.string.isRequired
  })).isRequired
};
