// @sanity/form-builder/lib/inputs/PortableText/__workshop__/textBlocks/TextBlocksStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TextBlocksStory = TextBlocksStory;
var _ui = require("@sanity/ui");
var _uiWorkshop = require("@sanity/ui-workshop");
var _react = _interopRequireWildcard(require("react"));
var _text = require("../../text");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
var errorMarker = {
  type: 'validation',
  level: 'error',
  path: [{
    _key: 'a'
  }],
  item: {
    message: 'There is an error with this text block'
  }
};
var customMarker = {
  type: 'test',
  path: [{
    _key: 'a'
  }],
  item: {
    message: '1 comment'
  }
};
var blockBase = {
  _type: 'myBlockType',
  _key: 'a',
  style: 'normal',
  markDefs: [],
  children: [{
    _type: 'span',
    _key: 'a1',
    text: "This is a custom portable text block above an empty image block! It's the block below. There should be a nice margin below me?",
    marks: []
  }]
};
function TextBlocksStory() {
  var hasErrors = (0, _uiWorkshop.useBoolean)('Has errors', false);
  var hasMarkers = (0, _uiWorkshop.useBoolean)('Has markers', false);
  var markers = (0, _react.useMemo)(() => [...(hasErrors ? [errorMarker] : []), ...(hasMarkers ? [customMarker] : [])], [hasErrors, hasMarkers]);
  var onChange = (0, _uiWorkshop.useAction)('onChange');
  var attributes = (0, _react.useMemo)(() => ({
    focused: false,
    selected: false,
    path: []
  }), []);
  var readOnly = false;
  return /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Container, {
    width: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    padding: 3,
    shadow: 1,
    overflow: "auto",
    style: {
      maxHeight: '100%'
    }
  }, /*#__PURE__*/_react.default.createElement(_text.TextBlock, {
    markers: markers,
    block: _objectSpread(_objectSpread({}, blockBase), {}, {
      style: 'h1'
    }),
    compareValue: undefined,
    onChange: onChange,
    attributes: attributes,
    readOnly: readOnly
  }, "Heading 1"), /*#__PURE__*/_react.default.createElement(_text.TextBlock, {
    markers: markers,
    block: _objectSpread(_objectSpread({}, blockBase), {}, {
      style: 'normal'
    }),
    compareValue: undefined,
    onChange: onChange,
    attributes: attributes,
    readOnly: readOnly
  }, "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book."), /*#__PURE__*/_react.default.createElement(_text.TextBlock, {
    markers: markers,
    block: _objectSpread(_objectSpread({}, blockBase), {}, {
      style: 'normal',
      listItem: 'number',
      level: 1
    }),
    compareValue: undefined,
    onChange: onChange,
    attributes: attributes,
    readOnly: readOnly
  }, "Lorem Ipsum is simply dummy text of the printing and typesetting industry."), /*#__PURE__*/_react.default.createElement(_text.TextBlock, {
    markers: markers,
    block: _objectSpread(_objectSpread({}, blockBase), {}, {
      style: 'normal',
      listItem: 'number',
      level: 1
    }),
    compareValue: undefined,
    onChange: onChange,
    attributes: attributes,
    readOnly: readOnly
  }, "Lorem Ipsum is simply dummy text of the printing and typesetting industry."), /*#__PURE__*/_react.default.createElement(_text.TextBlock, {
    markers: markers,
    block: _objectSpread(_objectSpread({}, blockBase), {}, {
      style: 'normal',
      listItem: 'number',
      level: 3
    }),
    compareValue: undefined,
    onChange: onChange,
    attributes: attributes,
    readOnly: readOnly
  }, "Lorem Ipsum is simply dummy text of the printing and typesetting industry."), /*#__PURE__*/_react.default.createElement(_text.TextBlock, {
    markers: markers,
    block: _objectSpread(_objectSpread({}, blockBase), {}, {
      style: 'normal',
      listItem: 'number',
      level: 3
    }),
    compareValue: undefined,
    onChange: onChange,
    attributes: attributes,
    readOnly: readOnly
  }, "Lorem Ipsum is simply dummy text of the printing and typesetting industry."), /*#__PURE__*/_react.default.createElement(_text.TextBlock, {
    markers: markers,
    block: _objectSpread(_objectSpread({}, blockBase), {}, {
      style: 'h2'
    }),
    compareValue: undefined,
    onChange: onChange,
    attributes: attributes,
    readOnly: readOnly
  }, "Heading 2"), /*#__PURE__*/_react.default.createElement(_text.TextBlock, {
    markers: markers,
    block: _objectSpread(_objectSpread({}, blockBase), {}, {
      style: 'normal'
    }),
    compareValue: undefined,
    onChange: onChange,
    attributes: attributes,
    readOnly: readOnly
  }, "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book."), /*#__PURE__*/_react.default.createElement(_text.TextBlock, {
    markers: markers,
    block: _objectSpread(_objectSpread({}, blockBase), {}, {
      style: 'normal',
      listItem: 'bullet',
      level: 1
    }),
    compareValue: undefined,
    onChange: onChange,
    attributes: attributes,
    readOnly: readOnly
  }, "Lorem Ipsum is simply dummy text of the printing and typesetting industry."), /*#__PURE__*/_react.default.createElement(_text.TextBlock, {
    markers: markers,
    block: _objectSpread(_objectSpread({}, blockBase), {}, {
      style: 'normal',
      listItem: 'bullet',
      level: 2
    }),
    compareValue: undefined,
    onChange: onChange,
    attributes: attributes,
    readOnly: readOnly
  }, "Lorem Ipsum is simply dummy text of the printing and typesetting industry."), /*#__PURE__*/_react.default.createElement(_text.TextBlock, {
    block: _objectSpread(_objectSpread({}, blockBase), {}, {
      style: 'h3'
    }),
    markers: [],
    compareValue: undefined,
    onChange: onChange,
    attributes: attributes,
    readOnly: readOnly
  }, "Heading 3"), /*#__PURE__*/_react.default.createElement(_text.TextBlock, {
    block: _objectSpread(_objectSpread({}, blockBase), {}, {
      style: 'normal'
    }),
    markers: markers,
    compareValue: undefined,
    onChange: onChange,
    attributes: attributes,
    readOnly: readOnly
  }, "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book."), /*#__PURE__*/_react.default.createElement(_text.TextBlock, {
    block: _objectSpread(_objectSpread({}, blockBase), {}, {
      style: 'blockquote'
    }),
    markers: markers,
    compareValue: undefined,
    onChange: onChange,
    attributes: attributes,
    readOnly: readOnly
  }, "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book."))));
}
