// @sanity/desk-tool/lib/components/pane/__workshop__/SplitPanesStory/DocumentPane/DocumentViewPanel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocumentViewPanel = DocumentViewPanel;
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _react = _interopRequireDefault(require("react"));
var _PaneContent = require("../../../PaneContent");
var _PaneFooter = require("../../../PaneFooter");
var _PaneHeader = require("../../../PaneHeader");
var _usePaneLayout2 = require("../../../usePaneLayout");
var _DocumentActions = require("./DocumentActions");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function DocumentViewPanel(props) {
  var onBackClick = props.onBackClick,
    reviewChanges = props.reviewChanges,
    title = props.title,
    toggleReviewChanges = props.toggleReviewChanges;
  var _usePaneLayout = (0, _usePaneLayout2.usePaneLayout)(),
    layoutCollapsed = _usePaneLayout.collapsed;
  return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    direction: "column",
    flex: 2,
    height: "fill",
    overflow: layoutCollapsed ? undefined : 'hidden'
  }, /*#__PURE__*/_react.default.createElement(_PaneHeader.PaneHeader, {
    backButton: layoutCollapsed && /*#__PURE__*/_react.default.createElement(_ui.Button, {
      icon: _icons.ArrowLeftIcon,
      mode: "bleed",
      onClick: onBackClick
    }),
    title: title
  }), /*#__PURE__*/_react.default.createElement(_PaneContent.PaneContent, null, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    height: layoutCollapsed ? undefined : 'fill',
    overflow: layoutCollapsed ? undefined : 'auto'
  }, /*#__PURE__*/_react.default.createElement(_ui.Container, {
    width: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    padding: 4,
    space: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true
  }, "Lorem ipsum dolor ", /*#__PURE__*/_react.default.createElement("a", {
    href: "#"
  }, "sit amet"), ", consectetur adipiscing elit. Donec vitae odio tellus. Etiam non metus at ante varius viverra. Pellentesque in iaculis lectus. Mauris laoreet et dui quis rhoncus. Quisque tempus velit tellus. In vel maximus quam. Vestibulum eu interdum diam. Aliquam luctus mattis justo, lacinia accumsan nunc tincidunt ac. Curabitur molestie malesuada velit et venenatis. Ut eget nunc sit amet ligula dapibus vehicula ut sed velit. Donec eget erat consectetur, scelerisque nibh nec, fermentum ipsum. Nulla at erat eu felis fermentum pellentesque."), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true
  }, "Suspendisse lacinia mi nibh, sit amet ultricies neque vulputate eget. Vivamus nisl augue, sodales vitae velit ac, vehicula dignissim leo. Suspendisse ornare efficitur porttitor. Cras placerat, augue in tempus malesuada, urna dolor volutpat erat, vitae pulvinar odio sapien sed odio. Nulla est enim, rutrum sit amet felis non, blandit tempus lectus. Pellentesque nulla tellus, fermentum vel mauris vitae, efficitur tincidunt nulla. Etiam venenatis ante ac condimentum porta. Praesent vel ornare ex. Fusce scelerisque, dolor quis faucibus bibendum, mi tortor gravida velit, id blandit ante nisl nec mauris. Donec ultrices interdum ipsum. Vestibulum venenatis malesuada quam, quis porta elit. Donec volutpat massa quis dolor fermentum lobortis. Morbi quis eros at lectus ornare aliquet."), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true
  }, "Interdum et malesuada fames ac ante ipsum primis in faucibus. Maecenas dictum velit in nunc pretium, in gravida elit faucibus. Vivamus at porta turpis, a suscipit eros. Donec sit amet mattis risus. Pellentesque auctor est id leo efficitur, eu eleifend diam suscipit. Nulla nec congue ante. Praesent elementum dolor felis, in dignissim massa tincidunt nec. Duis consectetur congue commodo. Quisque id facilisis dui. Lorem ipsum dolor sit amet, consectetur adipiscing elit."), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true
  }, "Donec a tristique turpis. Pellentesque eleifend eu libero ac laoreet. Fusce volutpat hendrerit ex. Suspendisse ornare tempor est nec faucibus. Morbi semper ullamcorper arcu nec elementum. Nam sagittis tempor sapien, nec blandit ante lacinia id. Vestibulum dapibus odio quam, eu placerat diam pulvinar eu. Nulla posuere dolor volutpat, facilisis felis ac, tristique nulla. Mauris eget luctus ex. Curabitur at imperdiet metus. Donec rutrum dui ac viverra vulputate. Curabitur aliquam erat quis turpis lobortis sollicitudin. Integer ut eleifend lectus."), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true
  }, "In hac habitasse platea dictumst. Maecenas venenatis nisi est, auctor efficitur nisi lacinia nec. Nam condimentum elit quis est mattis, eu fermentum neque feugiat. Nulla arcu neque, porttitor varius vulputate bibendum, facilisis sed nisi. Praesent id urna id lacus scelerisque convallis sed vel quam. Vestibulum est massa, sollicitudin eget ipsum et, porta pellentesque magna. Cras vulputate semper odio at facilisis. Mauris eget purus congue, sollicitudin nisi sed, euismod dui."))))), /*#__PURE__*/_react.default.createElement(_PaneFooter.PaneFooter, {
    padding: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, null, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
    disabled: layoutCollapsed,
    icon: _icons.RestoreIcon,
    mode: "bleed",
    onClick: toggleReviewChanges,
    selected: reviewChanges
  })), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1,
    style: {
      maxWidth: '260px'
    }
  }, /*#__PURE__*/_react.default.createElement(_DocumentActions.DocumentActions, null)))));
}
