// @sanity/desk-tool/lib/components/pane/__workshop__/SplitPanesStory/DocumentPane/ReviewChangesPanel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReviewChangesPanel = ReviewChangesPanel;
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _react = _interopRequireDefault(require("react"));
var _PaneHeader = require("../../../PaneHeader");
var _usePane2 = require("../../../usePane");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function ReviewChangesPanel(props) {
  var _usePane = (0, _usePane2.usePane)(),
    collapsed = _usePane.collapsed;
  return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    direction: "column",
    flex: 1,
    hidden: collapsed,
    style: {
      borderLeft: '1px dashed var(--card-border-color)',
      overflow: 'hidden',
      minWidth: 320
    }
  }, /*#__PURE__*/_react.default.createElement(_PaneHeader.PaneHeader, {
    actions: /*#__PURE__*/_react.default.createElement(_ui.Button, {
      icon: _icons.CloseIcon,
      mode: "bleed",
      onClick: props.onClose
    }),
    title: "Review changes"
  }));
}
