// @sanity/base/lib/components/__workshop__/CollapseMenuStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CollapseMenuStory;
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _uiWorkshop = require("@sanity/ui-workshop");
var _react = _interopRequireDefault(require("react"));
var _ = require("..");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var GAP_OPTIONS = {
  '0': 0,
  '1': 1,
  '2': 2,
  '3': 3,
  '4': 4
};
function CollapseMenuStory() {
  var collapsed = (0, _uiWorkshop.useBoolean)('Collapsed', false);
  var gap = (0, _uiWorkshop.useSelect)('Gap', GAP_OPTIONS, 1);
  var collapseText = (0, _uiWorkshop.useBoolean)('Collapse text', true);
  return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    height: "fill",
    justify: "center",
    padding: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    shadow: 1,
    radius: 3,
    padding: 1
  }, /*#__PURE__*/_react.default.createElement(_.CollapseMenu, {
    gap: gap,
    collapsed: collapsed,
    collapseText: collapseText,
    menuButtonProps: {
      button: /*#__PURE__*/_react.default.createElement(_ui.Button, {
        icon: _icons.EllipsisVerticalIcon,
        mode: "bleed",
        padding: 2,
        fontSize: 1
      })
    }
  }, [...Array(5).keys()].map(num => /*#__PURE__*/_react.default.createElement(_.CollapseMenuButton, {
    key: num,
    fontSize: 1,
    padding: 2,
    text: "Button ".concat(num + 1),
    dividerBefore: Boolean(num % 2),
    icon: _icons.IceCreamIcon,
    mode: "bleed",
    collapsedProps: {
      tooltipText: 'Collapsed'
    },
    expandedProps: {
      tooltipText: 'Expanded'
    }
  })))));
}
