// @sanity/base/lib/components/__workshop__/PreviewCardStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PreviewCardStory;
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _uiWorkshop = require("@sanity/ui-workshop");
var _react = _interopRequireDefault(require("react"));
var _ = require("..");
var _presence = require("../../presence");
var _PreviewCard = require("../PreviewCard");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function PreviewCardStory() {
  var selected = (0, _uiWorkshop.useBoolean)('Selected', true);
  return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    height: "fill",
    justify: "center",
    padding: 4,
    sizing: "border"
  }, /*#__PURE__*/_react.default.createElement(_ui.Container, {
    width: 0
  }, /*#__PURE__*/_react.default.createElement(_PreviewCard.PreviewCard, {
    as: "button",
    paddingY: 2,
    paddingX: 3,
    radius: 2,
    selected: selected
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "center",
    justify: "center",
    gap: 4
  }, /*#__PURE__*/_react.default.createElement(_presence.DocumentPreviewPresence, {
    presence: ['id-1', 'id-2', 'id-3', 'id-4'].map(num => ({
      id: num,
      imageUrl: 'https://source.unsplash.com/96x96/?face'
    }))
  }), /*#__PURE__*/_react.default.createElement(_.TextWithTone, {
    tone: "positive"
  }, /*#__PURE__*/_react.default.createElement(_icons.PublishIcon, null)), /*#__PURE__*/_react.default.createElement(_.TextWithTone, {
    tone: "caution"
  }, /*#__PURE__*/_react.default.createElement(_icons.EditIcon, null))))));
}
