// @sanity/desk-tool/lib/panes/error/ErrorPane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ErrorPane = ErrorPane;
var _react = _interopRequireDefault(require("react"));
var _ui = require("@sanity/ui");
var _pane = require("../../components/pane");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/**
 * @internal
 */
function ErrorPane(props) {
  var children = props.children,
    flex = props.flex,
    minWidth = props.minWidth,
    paneKey = props.paneKey,
    _props$title = props.title,
    title = _props$title === void 0 ? 'Error' : _props$title,
    _props$tone = props.tone,
    tone = _props$tone === void 0 ? 'critical' : _props$tone;
  return /*#__PURE__*/_react.default.createElement(_pane.Pane, {
    flex: flex,
    id: paneKey,
    minWidth: minWidth,
    tone: tone
  }, /*#__PURE__*/_react.default.createElement(_pane.PaneHeader, {
    title: title
  }), /*#__PURE__*/_react.default.createElement(_pane.PaneContent, {
    overflow: "auto"
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    paddingX: 4,
    paddingY: 5
  }, children)));
}
