// @sanity/form-builder/lib/inputs/PortableText/__workshop__/listCounter/ListCounterStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListCounterStory = ListCounterStory;
var _ui = require("@sanity/ui");
var _uiWorkshop = require("@sanity/ui-workshop");
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireWildcard(require("styled-components"));
var _templateObject, _templateObject2, _templateObject3, _templateObject4;
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var LEVELS = [0, 1, 2, 3, 4, 5, 6, 7, 8];
var BULLET_MARKERS = ['●', '○', '■'];
var NUMBER_FORMATS = ['number', 'lower-alpha', 'lower-roman'];
var FONT_SIZE_OPTIONS = {
  '0': 0,
  '1': 1,
  '2': 2,
  '3': 3,
  '4': 4
};
var SPACE_OPTIONS = {
  '0': 0,
  '1': 1,
  '2': 2,
  '3': 3,
  '4': 4,
  '5': 5,
  '6': 6,
  '7': 7,
  '8': 8,
  '9': 9
};
var items = [{
  level: 0
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'bullet'
}, {
  level: 1,
  listItem: 'bullet'
}, {
  level: 2,
  listItem: 'bullet'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 1,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 1,
  listItem: 'number'
}, {
  level: 0
}, {
  level: 0
}, {
  level: 0,
  listItem: 'bullet'
}, {
  level: 1,
  listItem: 'number'
}, {
  level: 1,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'bullet'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 1,
  listItem: 'number'
}, {
  level: 2,
  listItem: 'number'
}, {
  level: 3,
  listItem: 'number'
}, {
  level: 4,
  listItem: 'number'
}, {
  level: 5,
  listItem: 'number'
}, {
  level: 5,
  listItem: 'number'
}, {
  level: 5,
  listItem: 'number'
}, {
  level: 5,
  listItem: 'number'
}, {
  level: 4,
  listItem: 'number'
}, {
  level: 3,
  listItem: 'number'
}, {
  level: 2,
  listItem: 'number'
}, {
  level: 1,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'number'
}, {
  level: 0,
  listItem: 'bullet'
}];
var EditableWrapper = (0, _styledComponents.default)(_ui.Stack)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  counter-reset: ", ";\n\n  &[data-debug] {\n    outline: 1px solid cyan;\n  }\n\n  ", "\n\n  & > [data-list-item='number'] + *:not([data-list-item='number']) {\n    counter-reset: ", ";\n  }\n\n  ", "\n\n  & > [data-list-item='bullet'] {\n    counter-reset: ", ";\n  }\n"])), LEVELS.map(l => "list-level-".concat(l)).join(' '), LEVELS.map(l => {
  return (0, _styledComponents.css)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n      & > [data-level='", "'][data-list-item='number'] {\n        counter-increment: list-level-", ";\n      }\n    "])), l, l);
}), LEVELS.map(l => "list-level-".concat(l)).join(' '), LEVELS.slice(1).map(l => {
  return (0, _styledComponents.css)(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n      & > [data-level='", "'] + [data-level='", "'] {\n        counter-reset: list-level-", ";\n      }\n    "])), l, l - 1, l);
}), LEVELS.map(l => "list-level-".concat(l)).join(' '));
function ListCounterStory() {
  var debug = (0, _uiWorkshop.useBoolean)('Debug', false);
  var fontSize = (0, _uiWorkshop.useSelect)('Font size', FONT_SIZE_OPTIONS, 2);
  var space = (0, _uiWorkshop.useSelect)('Space', SPACE_OPTIONS, 3);
  return /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    paddingLeft: 6,
    paddingRight: 4,
    paddingY: [4, 5, 6]
  }, /*#__PURE__*/_react.default.createElement(_ui.Container, {
    width: 1
  }, /*#__PURE__*/_react.default.createElement(EditableWrapper, {
    "data-debug": debug ? '' : undefined,
    space: space
  }, items.map((item, itemIndex) => /*#__PURE__*/_react.default.createElement(Block, {
    fontSize: fontSize,
    index: itemIndex,
    key: itemIndex,
    value: item
  })))));
}
var BlockRoot = _styledComponents.default.div(props => {
  var $level = props.$level;
  return (0, _styledComponents.css)(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n    padding-left: ", "px;\n\n    [data-debug] & {\n      outline: 1px solid rgba(0 0 0 / 0.25);\n      outline-offset: -1px;\n    }\n\n    &[data-list-item] {\n      padding-left: ", "px;\n    }\n\n    &[data-list-item] > div > [data-prefix] {\n      position: absolute;\n      margin-left: -4.5rem;\n      width: 3.75rem;\n      text-align: right;\n      box-sizing: border-box;\n\n      [data-debug] & {\n        background: rgba(255 0 0 / 0.2);\n      }\n    }\n\n    &[data-list-item='number'] > div > [data-prefix] {\n      width: 4rem;\n      font-variant-numeric: tabular-nums;\n\n      & > span:before {\n        content: ", " '.';\n        content: ", "\n          '.';\n      }\n    }\n\n    &[data-list-item='bullet'] > div > [data-prefix] {\n      /* padding-right: 0.25rem; */\n\n      & > span {\n        position: relative;\n        top: -0.1875em;\n\n        &:before {\n          content: '", "';\n          font-size: 0.46666em;\n        }\n      }\n    }\n  "])), $level * 32, 32 + $level * 32, "counter(list-level-".concat($level, ")"), "counter(list-level-".concat($level, ", ").concat(NUMBER_FORMATS[$level % NUMBER_FORMATS.length], ")"), BULLET_MARKERS[$level % BULLET_MARKERS.length]);
});
function Block(props) {
  var fontSize = props.fontSize,
    index = props.index,
    value = props.value;
  return /*#__PURE__*/_react.default.createElement(BlockRoot, {
    $level: value.level || 0,
    "data-level": value.level || 0,
    "data-list-item": value.listItem
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    align: "flex-start"
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "span",
    "data-prefix": "",
    size: fontSize
  }), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    as: "span",
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "span",
    "data-text": "",
    size: fontSize
  }, "Block ", /*#__PURE__*/_react.default.createElement("code", null, "#", index), value.level !== undefined && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, ", Level ", /*#__PURE__*/_react.default.createElement("code", null, "#", value.level))))));
}
