// @sanity/field/lib/diff/components/GroupChange.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GroupChange = GroupChange;
var _reactHooks = require("@sanity/react-hooks");
var _react = _interopRequireWildcard(require("react"));
var _hooks = require("@sanity/base/hooks");
var _ui = require("@sanity/ui");
var _internal = require("@sanity/base/_internal");
var _undoChange = require("../changes/undoChange");
var _helpers = require("../helpers");
var _diff = require("../../types/portableText/diff");
var _useHover3 = require("../../utils/useHover");
var _paths = require("../../paths");
var _DiffContext = require("../contexts/DiffContext");
var _ChangeBreadcrumb = require("./ChangeBreadcrumb");
var _ChangeResolver = require("./ChangeResolver");
var _DocumentChangeContext = require("./DocumentChangeContext");
var _RevertChangesButton = require("./RevertChangesButton");
var _GroupChange = require("./GroupChange.styled");
var _excluded = ["change", "readOnly", "hidden"];
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
function GroupChange(_ref) {
  var _useConditionalReadOn;
  var group = _ref.change,
    readOnly = _ref.readOnly,
    hidden = _ref.hidden,
    restProps = _objectWithoutProperties(_ref, _excluded);
  var titlePath = group.titlePath,
    changes = group.changes,
    groupPath = group.path;
  var _useContext = (0, _react.useContext)(_DiffContext.DiffContext),
    diffPath = _useContext.path;
  var _useContext2 = (0, _react.useContext)(_DocumentChangeContext.DocumentChangeContext),
    documentId = _useContext2.documentId,
    schemaType = _useContext2.schemaType,
    FieldWrapper = _useContext2.FieldWrapper,
    rootDiff = _useContext2.rootDiff,
    isComparingCurrent = _useContext2.isComparingCurrent;
  var conditionalReadOnly = (_useConditionalReadOn = (0, _internal.useConditionalReadOnly)()) !== null && _useConditionalReadOn !== void 0 ? _useConditionalReadOn : readOnly;
  var isPortableText = changes.every(change => (0, _helpers.isFieldChange)(change) && (0, _diff.isPTSchemaType)(change.schemaType));
  var isNestedInDiff = (0, _paths.pathsAreEqual)(diffPath, groupPath);
  var _useHover = (0, _useHover3.useHover)(),
    _useHover2 = _slicedToArray(_useHover, 2),
    hoverRef = _useHover2[0],
    isHoveringRevert = _useHover2[1];
  var docOperations = (0, _reactHooks.useDocumentOperation)(documentId, schemaType.name);
  var _useState = (0, _react.useState)(false),
    _useState2 = _slicedToArray(_useState, 2),
    confirmRevertOpen = _useState2[0],
    setConfirmRevertOpen = _useState2[1];
  var _useState3 = (0, _react.useState)(null),
    _useState4 = _slicedToArray(_useState3, 2),
    revertButtonElement = _useState4[0],
    setRevertButtonElement = _useState4[1];
  var _useDocumentPairPermi = (0, _hooks.unstable_useDocumentPairPermissions)({
      id: documentId,
      type: schemaType.name,
      permission: 'update'
    }),
    _useDocumentPairPermi2 = _slicedToArray(_useDocumentPairPermi, 2),
    permissions = _useDocumentPairPermi2[0],
    isPermissionsLoading = _useDocumentPairPermi2[1];
  var handleRevertChanges = (0, _react.useCallback)(() => (0, _undoChange.undoChange)(group, rootDiff, docOperations), [group, rootDiff, docOperations]);
  var handleRevertChangesConfirm = (0, _react.useCallback)(() => {
    setConfirmRevertOpen(true);
  }, []);
  var closeRevertChangesConfirmDialog = (0, _react.useCallback)(() => {
    setConfirmRevertOpen(false);
  }, []);
  (0, _ui.useClickOutside)(() => setConfirmRevertOpen(false), [revertButtonElement]);
  var setRevertButtonRef = (0, _react.useCallback)(el => {
    hoverRef.current = el;
    setRevertButtonElement(el);
  }, [hoverRef]);
  var content = (0, _react.useMemo)(() => /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 1,
    as: _GroupChange.GroupChangeContainer,
    "data-revert-group-hover": isHoveringRevert ? '' : undefined,
    "data-portable-text": isPortableText ? '' : undefined,
    "data-revert-all-groups-hover": restProps['data-revert-all-changes-hover'] === '' ? '' : undefined
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    as: _GroupChange.ChangeListWrapper,
    space: 5
  }, changes.map(change => /*#__PURE__*/_react.default.createElement(_ChangeResolver.ChangeResolver, {
    key: change.key,
    change: change,
    readOnly: conditionalReadOnly,
    hidden: hidden
  }))), isComparingCurrent && !isPermissionsLoading && (permissions === null || permissions === void 0 ? void 0 : permissions.granted) && /*#__PURE__*/_react.default.createElement(_GroupChange.PopoverWrapper, {
    content: /*#__PURE__*/_react.default.createElement(_ui.Box, null, "Are you sure you want to revert the changes?", /*#__PURE__*/_react.default.createElement(_ui.Grid, {
      columns: 2,
      gap: 2,
      marginTop: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
      mode: "ghost",
      onClick: closeRevertChangesConfirmDialog
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      align: "center"
    }, "Cancel")), /*#__PURE__*/_react.default.createElement(_ui.Button, {
      tone: "critical",
      onClick: handleRevertChanges
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      align: "center"
    }, "Revert change")))),
    portal: true,
    padding: 4,
    placement: 'left',
    open: confirmRevertOpen,
    ref: setRevertButtonElement
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_RevertChangesButton.RevertChangesButton, {
    onClick: handleRevertChangesConfirm,
    ref: setRevertButtonRef,
    selected: confirmRevertOpen,
    disabled: conditionalReadOnly,
    "data-testid": "group-change-revert-button-".concat(group.fieldsetName)
  })))), [conditionalReadOnly, hidden, confirmRevertOpen, isPermissionsLoading, permissions]);
  return hidden ? null : /*#__PURE__*/_react.default.createElement(_ui.Stack, _extends({
    space: 1
  }, restProps), /*#__PURE__*/_react.default.createElement(_ChangeBreadcrumb.ChangeBreadcrumb, {
    titlePath: titlePath
  }), isNestedInDiff || !FieldWrapper ? content : /*#__PURE__*/_react.default.createElement(FieldWrapper, {
    path: group.path,
    hasHover: isHoveringRevert
  }, content));
}
