// @sanity/form-builder/lib/inputs/PortableText/object/InlineObjectToolbarPopover.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InlineObjectToolbarPopover = InlineObjectToolbarPopover;
var _react = _interopRequireWildcard(require("react"));
var _ui = require("@sanity/ui");
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _icons = require("@sanity/icons");
var _templateObject;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var ToolbarPopover = (0, _styledComponents.default)(_ui.Popover)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  &[data-popper-reference-hidden='true'] {\n    display: none !important;\n  }\n"])));
var POPOVER_FALLBACK_PLACEMENTS = ['top', 'bottom'];
function InlineObjectToolbarPopover(props) {
  var open = props.open,
    onEdit = props.onEdit,
    onDelete = props.onDelete,
    referenceElement = props.referenceElement,
    scrollElement = props.scrollElement,
    setOpen = props.setOpen,
    title = props.title;
  var _useTheme = (0, _ui.useTheme)(),
    sanity = _useTheme.sanity;
  var editButtonRef = (0, _react.useRef)();
  var popoverScheme = sanity.color.dark ? 'light' : 'dark';
  var isTabbing = (0, _react.useRef)(false);

  // Close floating toolbar on Escape
  // Focus to edit button on Tab
  (0, _ui.useGlobalKeyDown)((0, _react.useCallback)(event => {
    if (!open) {
      return;
    }
    if (event.key === 'Escape') {
      event.preventDefault();
      event.stopPropagation();
      isTabbing.current = false;
      setOpen(false);
    }
    if (event.key === 'Tab') {
      if (!isTabbing.current) {
        event.preventDefault();
        event.stopPropagation();
        editButtonRef.current.focus();
        isTabbing.current = true;
      }
    }
  }, [open, setOpen]));
  (0, _react.useEffect)(() => {
    if (open && isTabbing.current) {
      var _editButtonRef$curren;
      (_editButtonRef$curren = editButtonRef.current) === null || _editButtonRef$curren === void 0 ? void 0 : _editButtonRef$curren.focus();
    }
  }, [open]);
  return /*#__PURE__*/_react.default.createElement("div", {
    contentEditable: false
  }, /*#__PURE__*/_react.default.createElement(ToolbarPopover, {
    boundaryElement: scrollElement,
    constrainSize: true,
    content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 1
    }, /*#__PURE__*/_react.default.createElement(_ui.Inline, {
      space: 1
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      weight: "semibold",
      size: 1
    }, title)), /*#__PURE__*/_react.default.createElement(_ui.Button, {
      icon: _icons.EditIcon,
      mode: "bleed",
      onClick: onEdit,
      padding: 2,
      ref: editButtonRef
    }), /*#__PURE__*/_react.default.createElement(_ui.Button, {
      icon: _icons.TrashIcon,
      mode: "bleed",
      padding: 2,
      onClick: onDelete,
      tone: "critical"
    }))),
    fallbackPlacements: POPOVER_FALLBACK_PLACEMENTS,
    open: open,
    placement: "top",
    portal: "editor",
    referenceElement: referenceElement,
    scheme: popoverScheme
  }));
}
