// @sanity/form-builder/lib/inputs/files/common/UploadWarning.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UploadWarning = UploadWarning;
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _templateObject;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var ButtonWrapper = (0, _styledComponents.default)(_ui.Button)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  width: 100%;\n"])));
function UploadWarning(_ref) {
  var onClearStale = _ref.onClearStale;
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    tone: "caution",
    padding: 4,
    border: true,
    radius: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    gap: 4,
    marginBottom: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, /*#__PURE__*/_react.default.createElement(_icons.WarningOutlineIcon, null))), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    weight: "semibold"
  }, "Incomplete upload"), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, "An upload has made no progress in the last 6m and likely got interrupted. You can safely clear the incomplete upload and try uploading again."))), /*#__PURE__*/_react.default.createElement(ButtonWrapper, {
    icon: _icons.ResetIcon,
    text: "Clear upload",
    onClick: onClearStale,
    mode: "ghost"
  }));
}
