// @sanity/base/lib/_exports/_internal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  observeForPreview: true,
  observePaths: true,
  prepareForPreview: true,
  getPreviewPaths: true,
  observeDocumentTypeFromId: true,
  unstable_observePathsDocumentPair: true,
  unstable_observeDocumentPairAvailability: true,
  AvailabilityReason: true,
  getSearchableTypes: true,
  createWeightedSearch: true,
  createHookFromObservableFactory: true,
  collate: true,
  getDraftId: true,
  getPublishedId: true,
  getIdPair: true,
  FieldStatus: true,
  ChangeIndicatorValueProvider: true,
  resizeObserver: true,
  getNewDocumentOptions: true,
  remoteSnapshots: true
};
Object.defineProperty(exports, "AvailabilityReason", {
  enumerable: true,
  get: function get() {
    return _types.AvailabilityReason;
  }
});
Object.defineProperty(exports, "ChangeIndicatorValueProvider", {
  enumerable: true,
  get: function get() {
    return _ChangeIndicator.ChangeIndicatorValueProvider;
  }
});
Object.defineProperty(exports, "FieldStatus", {
  enumerable: true,
  get: function get() {
    return _FieldStatus.default;
  }
});
Object.defineProperty(exports, "collate", {
  enumerable: true,
  get: function get() {
    return _draftUtils.collate;
  }
});
Object.defineProperty(exports, "createHookFromObservableFactory", {
  enumerable: true,
  get: function get() {
    return _createHookFromObservableFactory.createHookFromObservableFactory;
  }
});
Object.defineProperty(exports, "createWeightedSearch", {
  enumerable: true,
  get: function get() {
    return _createWeightedSearch.createWeightedSearch;
  }
});
Object.defineProperty(exports, "getDraftId", {
  enumerable: true,
  get: function get() {
    return _draftUtils.getDraftId;
  }
});
Object.defineProperty(exports, "getIdPair", {
  enumerable: true,
  get: function get() {
    return _draftUtils.getIdPair;
  }
});
Object.defineProperty(exports, "getNewDocumentOptions", {
  enumerable: true,
  get: function get() {
    return _getNewDocumentOptions.getNewDocumentOptions;
  }
});
Object.defineProperty(exports, "getPreviewPaths", {
  enumerable: true,
  get: function get() {
    return _preview.getPreviewPaths;
  }
});
Object.defineProperty(exports, "getPublishedId", {
  enumerable: true,
  get: function get() {
    return _draftUtils.getPublishedId;
  }
});
Object.defineProperty(exports, "getSearchableTypes", {
  enumerable: true,
  get: function get() {
    return _utils.getSearchableTypes;
  }
});
Object.defineProperty(exports, "observeDocumentTypeFromId", {
  enumerable: true,
  get: function get() {
    return _preview.observeDocumentTypeFromId;
  }
});
Object.defineProperty(exports, "observeForPreview", {
  enumerable: true,
  get: function get() {
    return _preview.observeForPreview;
  }
});
Object.defineProperty(exports, "observePaths", {
  enumerable: true,
  get: function get() {
    return _preview.observePaths;
  }
});
Object.defineProperty(exports, "prepareForPreview", {
  enumerable: true,
  get: function get() {
    return _preview.prepareForPreview;
  }
});
Object.defineProperty(exports, "remoteSnapshots", {
  enumerable: true,
  get: function get() {
    return _remoteSnapshots.remoteSnapshots;
  }
});
Object.defineProperty(exports, "resizeObserver", {
  enumerable: true,
  get: function get() {
    return _resizeObserver.resizeObserver;
  }
});
Object.defineProperty(exports, "unstable_observeDocumentPairAvailability", {
  enumerable: true,
  get: function get() {
    return _availability.observeDocumentPairAvailability;
  }
});
Object.defineProperty(exports, "unstable_observePathsDocumentPair", {
  enumerable: true,
  get: function get() {
    return _documentPair.observePathsDocumentPair;
  }
});
var _preview = require("../preview");
var _documentPair = require("../preview/documentPair");
var _availability = require("../preview/availability");
var _types = require("../preview/types");
var _utils = require("../search/common/utils");
var _createWeightedSearch = require("../search/weighted/createWeightedSearch");
var _createHookFromObservableFactory = require("../util/createHookFromObservableFactory");
var _draftUtils = require("../util/draftUtils");
var _FieldStatus = _interopRequireDefault(require("../__legacy/@sanity/components/fieldsets/FieldStatus"));
var _ChangeIndicator = require("../change-indicators/ChangeIndicator");
var _resizeObserver = require("../util/resizeObserver");
var _getNewDocumentOptions = require("../util/getNewDocumentOptions");
var _remoteSnapshots = require("../datastores/document/document-pair/remoteSnapshots");
var _types2 = require("../actions/utils/types");
Object.keys(_types2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types2[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _types2[key];
    }
  });
});
var _conditionalReadOnly = require("../conditional-property/conditionalReadOnly");
Object.keys(_conditionalReadOnly).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _conditionalReadOnly[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _conditionalReadOnly[key];
    }
  });
});
var _types3 = require("../module-status/types");
Object.keys(_types3).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types3[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _types3[key];
    }
  });
});
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
