// @sanity/base/lib/_exports/hooks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "unstable_useConditionalProperty", {
  enumerable: true,
  get: function get() {
    return _conditionalProperty.unstable_useConditionalProperty;
  }
});
Object.defineProperty(exports, "unstable_useDocumentPairPermissions", {
  enumerable: true,
  get: function get() {
    return _grants.unstable_useDocumentPairPermissions;
  }
});
Object.defineProperty(exports, "unstable_useDocumentValuePermissions", {
  enumerable: true,
  get: function get() {
    return _grants.unstable_useDocumentValuePermissions;
  }
});
Object.defineProperty(exports, "unstable_useTemplatePermissions", {
  enumerable: true,
  get: function get() {
    return _grants.unstable_useTemplatePermissions;
  }
});
Object.defineProperty(exports, "useCurrentUser", {
  enumerable: true,
  get: function get() {
    return _hooks2.useCurrentUser;
  }
});
Object.defineProperty(exports, "useDocumentPresence", {
  enumerable: true,
  get: function get() {
    return _hooks.useDocumentPresence;
  }
});
Object.defineProperty(exports, "useDocumentPresenceUsers", {
  enumerable: true,
  get: function get() {
    return _hooks.useDocumentPresenceUsers;
  }
});
Object.defineProperty(exports, "useDocumentType", {
  enumerable: true,
  get: function get() {
    return _hooks3.useDocumentType;
  }
});
Object.defineProperty(exports, "useDocumentValues", {
  enumerable: true,
  get: function get() {
    return _useDocumentValues.useDocumentValues;
  }
});
Object.defineProperty(exports, "useGlobalPresence", {
  enumerable: true,
  get: function get() {
    return _hooks.useGlobalPresence;
  }
});
Object.defineProperty(exports, "useGlobalPresenceUsers", {
  enumerable: true,
  get: function get() {
    return _hooks.useGlobalPresenceUsers;
  }
});
Object.defineProperty(exports, "useModuleStatus", {
  enumerable: true,
  get: function get() {
    return _moduleStatus.useModuleStatus;
  }
});
Object.defineProperty(exports, "useTimeAgo", {
  enumerable: true,
  get: function get() {
    return _useTimeAgo.useTimeAgo;
  }
});
Object.defineProperty(exports, "useUser", {
  enumerable: true,
  get: function get() {
    return _hooks2.useUser;
  }
});
Object.defineProperty(exports, "useUserColor", {
  enumerable: true,
  get: function get() {
    return _hooks4.useUserColor;
  }
});
var _hooks = require("../datastores/presence/hooks");
var _hooks2 = require("../datastores/user/hooks");
var _hooks3 = require("../datastores/document/hooks");
var _grants = require("../datastores/grants");
var _hooks4 = require("../user-color/hooks");
var _useTimeAgo = require("../time/useTimeAgo");
var _useDocumentValues = require("../datastores/document/useDocumentValues");
var _moduleStatus = require("../module-status");
var _conditionalProperty = require("../conditional-property");
