// @sanity/desk-tool/lib/components/DraftStatus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DraftStatus = void 0;
var _react = _interopRequireDefault(require("react"));
var _ui = require("@sanity/ui");
var _components = require("@sanity/base/components");
var _icons = require("@sanity/icons");
var _TimeAgo = require("./TimeAgo");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var DraftStatus = _ref => {
  var document = _ref.document;
  return /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
    portal: true,
    content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1
    }, document ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Edited ", (document === null || document === void 0 ? void 0 : document._updatedAt) && /*#__PURE__*/_react.default.createElement(_TimeAgo.TimeAgo, {
      time: document === null || document === void 0 ? void 0 : document._updatedAt
    })) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "No unpublished edits")))
  }, /*#__PURE__*/_react.default.createElement(_components.TextWithTone, {
    tone: "caution",
    dimmed: !document,
    muted: !document,
    size: 1
  }, /*#__PURE__*/_react.default.createElement(_icons.EditIcon, null)));
};
exports.DraftStatus = DraftStatus;
