// @sanity/desk-tool/lib/components/PublishedStatus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PublishedStatus = void 0;
var _react = _interopRequireDefault(require("react"));
var _ui = require("@sanity/ui");
var _components = require("@sanity/base/components");
var _icons = require("@sanity/icons");
var _TimeAgo = require("./TimeAgo");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var PublishedStatus = _ref => {
  var document = _ref.document;
  return /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
    portal: true,
    content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1
    }, document ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Published ", document._updatedAt && /*#__PURE__*/_react.default.createElement(_TimeAgo.TimeAgo, {
      time: document._updatedAt
    })) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Not published")))
  }, /*#__PURE__*/_react.default.createElement(_components.TextWithTone, {
    tone: "positive",
    dimmed: !document,
    muted: !document,
    size: 1
  }, /*#__PURE__*/_react.default.createElement(_icons.PublishIcon, null)));
};
exports.PublishedStatus = PublishedStatus;
