// @sanity/form-builder/lib/inputs/PortableText/__workshop__/editObjects/schemaType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.schemaType = exports.schema = void 0;
var _schema = _interopRequireDefault(require("@sanity/schema"));
var _schemaType = _interopRequireDefault(require("all:part:@sanity/base/schema-type"));
var _baseTypes = require("../_common/baseTypes");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var linkType = {
  type: 'object',
  name: 'link',
  title: 'Link',
  fields: [],
  options: {
    editModal: 'fold' // 'fullscreen' | 'popover' | 'fold'
  }
};

var myObjectBlockType = {
  type: 'object',
  name: 'myObjectBlock',
  title: 'My object block',
  fields: [],
  options: {
    editModal: 'fullscreen' // 'fullscreen' | 'popover' | 'fold'
  }
};

var myInlineObjectType = {
  type: 'object',
  name: 'myInlineObject',
  title: 'My inline object',
  fields: [],
  options: {
    editModal: 'popover' // 'fullscreen' | 'popover' | 'fold'
  }
};

var blockType = {
  type: 'block',
  marks: {
    annotations: [linkType]
  },
  of: [myInlineObjectType, myObjectBlockType]
};
var bodyType = {
  type: 'array',
  name: 'body',
  of: [blockType, myObjectBlockType]
};
var schema = _schema.default.compile({
  name: 'default',
  types: _schemaType.default.concat([bodyType]).concat(_baseTypes.baseTypes)
});
exports.schema = schema;
var schemaType = schema.get('body');
exports.schemaType = schemaType;
