// @sanity/default-layout/lib/navbar/search/SearchDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SearchDialog = SearchDialog;
var _autoId = require("@reach/auto-id");
var _ui = require("@sanity/ui");
var _react = _interopRequireWildcard(require("react"));
var _reactFocusLock = _interopRequireDefault(require("react-focus-lock"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _RecentSearches = require("./components/RecentSearches");
var _SearchHeader = require("./components/SearchHeader");
var _SearchResults = require("./components/SearchResults");
var _TypeFilters = require("./components/TypeFilters");
var _commandList = require("./contexts/commandList");
var _search = require("./contexts/search");
var _useMeasureSearchResultsIndex = require("./hooks/useMeasureSearchResultsIndex");
var _useSearchHotkeys = require("./hooks/useSearchHotkeys");
var _hasSearchableTerms = require("./utils/hasSearchableTerms");
var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var DialogContentCard = (0, _styledComponents.default)(_ui.Card)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  height: 100%;\n"])));
var InnerCard = (0, _styledComponents.default)(_ui.Card)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  flex-direction: column;\n  overflow: hidden;\n  pointer-events: all;\n  position: relative;\n"])));
var SearchDialogBox = (0, _styledComponents.default)(_ui.Box)(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n  height: 100%;\n  left: 0;\n  overflow: hidden;\n  pointer-events: none;\n  position: fixed;\n  top: 0;\n  width: 100%;\n  z-index: 1;\n"])));
var SearchContentBox = (0, _styledComponents.default)(_ui.Box)(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n  height: 100%;\n  width: 100%;\n  overflow-x: hidden;\n  overflow-y: auto;\n"])));
var StyledDialog = (0, _styledComponents.default)(_ui.Dialog)(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["\n  [data-ui='DialogCard'] > [data-ui='Card'] {\n    flex: 1;\n  }\n"])));
function SearchDialog(_ref) {
  var onClose = _ref.onClose,
    onOpen = _ref.onOpen,
    open = _ref.open;
  var _useState = (0, _react.useState)(null),
    _useState2 = _slicedToArray(_useState, 2),
    childContainerElement = _useState2[0],
    setChildContainerRef = _useState2[1];
  var _useState3 = (0, _react.useState)(null),
    _useState4 = _slicedToArray(_useState3, 2),
    containerElement = _useState4[0],
    setContainerRef = _useState4[1];
  var _useState5 = (0, _react.useState)(null),
    _useState6 = _slicedToArray(_useState5, 2),
    headerInputElement = _useState6[0],
    setHeaderInputRef = _useState6[1];
  var _useState7 = (0, _react.useState)(null),
    _useState8 = _slicedToArray(_useState7, 2),
    pointerOverlayElement = _useState8[0],
    setPointerOverlayRef = _useState8[1];
  var isMountedRef = (0, _react.useRef)(false);
  var _useSearchState = (0, _search.useSearchState)(),
    dispatch = _useSearchState.dispatch,
    _useSearchState$state = _useSearchState.state,
    filtersVisible = _useSearchState$state.filtersVisible,
    recentSearches = _useSearchState$state.recentSearches,
    result = _useSearchState$state.result,
    terms = _useSearchState$state.terms;
  var hasValidTerms = (0, _hasSearchableTerms.hasSearchableTerms)(terms);

  /**
   * Measure top-most visible search result index
   */
  var _useMeasureSearchResu = (0, _useMeasureSearchResultsIndex.useMeasureSearchResultsIndex)(childContainerElement),
    lastSearchIndex = _useMeasureSearchResu.lastSearchIndex,
    resetLastSearchIndex = _useMeasureSearchResu.resetLastSearchIndex,
    setLastSearchIndex = _useMeasureSearchResu.setLastSearchIndex;

  /**
   * Store top-most search result scroll index on close
   */
  var handleClose = (0, _react.useCallback)(() => {
    setLastSearchIndex();
    onClose();
  }, [onClose, setLastSearchIndex]);

  /**
   * Bind hotkeys to open / close actions
   */
  (0, _useSearchHotkeys.useSearchHotkeys)({
    onClose: handleClose,
    onOpen,
    open
  });

  /**
   * Reset last search index when new results are loaded, or visiting recent searches
   * @todo Revise if/when we introduce pagination
   */
  (0, _react.useEffect)(() => {
    if ((!hasValidTerms || result.loaded) && isMountedRef.current) {
      resetLastSearchIndex();
    }
  }, [hasValidTerms, resetLastSearchIndex, result.loaded]);

  /**
   * Reset ordering when popover is closed (without valid search terms)
   */
  (0, _react.useEffect)(() => {
    if (!hasValidTerms && isMountedRef.current && !open) {
      dispatch({
        type: 'SEARCH_ORDERING_RESET'
      });
    }
  }, [dispatch, hasValidTerms, open]);

  /**
   * Store mounted state (must be last)
   */
  (0, _react.useEffect)(() => {
    if (!(isMountedRef !== null && isMountedRef !== void 0 && isMountedRef.current)) {
      isMountedRef.current = true;
    }
  }, []);
  var dialogId = (0, _autoId.useId)();
  if (!open) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement(_commandList.CommandListProvider, {
    ariaChildrenLabel: hasValidTerms ? 'Search results' : 'Recent searches',
    ariaHeaderLabel: "Search",
    autoFocus: true,
    childContainerElement: childContainerElement,
    childCount: hasValidTerms ? result.hits.length : recentSearches.length,
    containerElement: containerElement,
    headerInputElement: headerInputElement,
    id: dialogId || '',
    "data-testid": "search-results-dialog",
    initialSelectedIndex: hasValidTerms ? lastSearchIndex : 0,
    level: 0,
    pointerOverlayElement: pointerOverlayElement,
    virtualList: hasValidTerms
  }, /*#__PURE__*/_react.default.createElement(_ui.Portal, null, /*#__PURE__*/_react.default.createElement(_reactFocusLock.default, {
    autoFocus: false,
    returnFocus: true
  }, /*#__PURE__*/_react.default.createElement(SearchDialogBox, {
    ref: setContainerRef
  }, /*#__PURE__*/_react.default.createElement(InnerCard, {
    display: "flex",
    height: "fill",
    scheme: "light",
    tone: "default"
  }, /*#__PURE__*/_react.default.createElement(_SearchHeader.SearchHeader, {
    onClose: handleClose,
    setHeaderInputRef: setHeaderInputRef
  }), /*#__PURE__*/_react.default.createElement(SearchContentBox, {
    flex: 1
  }, hasValidTerms ? /*#__PURE__*/_react.default.createElement(_SearchResults.SearchResults, {
    onClose: handleClose,
    setChildContainerRef: setChildContainerRef,
    setPointerOverlayRef: setPointerOverlayRef
  }) : /*#__PURE__*/_react.default.createElement(_RecentSearches.RecentSearches, {
    setChildContainerRef: setChildContainerRef,
    setPointerOverlayRef: setPointerOverlayRef
  })), filtersVisible && /*#__PURE__*/_react.default.createElement(SearchDialogFilters, null))))));
}
function SearchDialogFilters() {
  var _useSearchState2 = (0, _search.useSearchState)(),
    dispatch = _useSearchState2.dispatch;
  var handleClose = (0, _react.useCallback)(() => {
    dispatch({
      type: 'FILTERS_HIDE'
    });
  }, [dispatch]);
  return /*#__PURE__*/_react.default.createElement(_reactFocusLock.default, {
    autoFocus: false
  }, /*#__PURE__*/_react.default.createElement(StyledDialog, {
    cardRadius: 1,
    header: "Filter",
    height: "fill",
    id: "search-filter",
    onClose: handleClose,
    width: 2
  }, /*#__PURE__*/_react.default.createElement(DialogContentCard, {
    tone: "default"
  }, /*#__PURE__*/_react.default.createElement(_TypeFilters.TypeFilters, null))));
}
