// @sanity/default-layout/lib/navbar/search/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ORDER_UPDATED_DESC = exports.ORDER_UPDATED_ASC = exports.ORDER_RELEVANCE = exports.ORDER_CREATED_DESC = exports.ORDER_CREATED_ASC = void 0;
var ORDER_RELEVANCE = {
  customMeasurementLabel: 'relevance',
  sort: {
    direction: 'desc',
    field: '_updatedAt'
  },
  title: 'Relevance'
};
exports.ORDER_RELEVANCE = ORDER_RELEVANCE;
var ORDER_CREATED_ASC = {
  ignoreScore: true,
  sort: {
    direction: 'asc',
    field: '_createdAt'
  },
  title: 'Created: Oldest first'
};
exports.ORDER_CREATED_ASC = ORDER_CREATED_ASC;
var ORDER_CREATED_DESC = {
  ignoreScore: true,
  sort: {
    direction: 'desc',
    field: '_createdAt'
  },
  title: 'Created: Newest first'
};
exports.ORDER_CREATED_DESC = ORDER_CREATED_DESC;
var ORDER_UPDATED_ASC = {
  ignoreScore: true,
  sort: {
    direction: 'asc',
    field: '_updatedAt'
  },
  title: 'Updated: Oldest first'
};
exports.ORDER_UPDATED_ASC = ORDER_UPDATED_ASC;
var ORDER_UPDATED_DESC = {
  ignoreScore: true,
  sort: {
    direction: 'desc',
    field: '_updatedAt'
  },
  title: 'Updated: Newest first'
};
exports.ORDER_UPDATED_DESC = ORDER_UPDATED_DESC;
