// @sanity/desk-tool/lib/_parts/base-document-actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DeleteAction", {
  enumerable: true,
  get: function get() {
    return _DeleteAction.DeleteAction;
  }
});
Object.defineProperty(exports, "DiscardChangesAction", {
  enumerable: true,
  get: function get() {
    return _DiscardChangesAction.DiscardChangesAction;
  }
});
Object.defineProperty(exports, "DuplicateAction", {
  enumerable: true,
  get: function get() {
    return _DuplicateAction.DuplicateAction;
  }
});
Object.defineProperty(exports, "PublishAction", {
  enumerable: true,
  get: function get() {
    return _PublishAction.PublishAction;
  }
});
Object.defineProperty(exports, "UnpublishAction", {
  enumerable: true,
  get: function get() {
    return _UnpublishAction.UnpublishAction;
  }
});
exports.default = defaultResolveDocumentActions;
var _PublishAction = require("../actions/PublishAction");
var _DiscardChangesAction = require("../actions/DiscardChangesAction");
var _UnpublishAction = require("../actions/UnpublishAction");
var _DuplicateAction = require("../actions/DuplicateAction");
var _DeleteAction = require("../actions/DeleteAction");
function defaultResolveDocumentActions() {
  return [_PublishAction.PublishAction, _DiscardChangesAction.DiscardChangesAction, _UnpublishAction.UnpublishAction, _DuplicateAction.DuplicateAction, _DeleteAction.DeleteAction];
}
