// @sanity/form-builder/lib/inputs/files/ImageInput/ImagePreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ImagePreview = ImagePreview;
var _react = _interopRequireWildcard(require("react"));
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _ImagePreview = require("./ImagePreview.styled");
var _excluded = ["drag", "readOnly", "isRejected", "src"];
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
/*
  Used for setting the initial image height - specifically for images
  that are small and so can take less space in the document
*/
var getImageSize = src => {
  var imageUrlParams = new URLSearchParams(src.split('?')[1]);
  var rect = imageUrlParams.get('rect');
  if (rect) {
    return [rect.split(',')[2], rect.split(',')[3]].map(Number);
  }
  return src.split('-')[1].split('.')[0].split('x').map(Number);
};
function ImagePreview(props) {
  var drag = props.drag,
    readOnly = props.readOnly,
    isRejected = props.isRejected,
    src = props.src,
    rest = _objectWithoutProperties(props, _excluded);
  var _useState = (0, _react.useState)(false),
    _useState2 = _slicedToArray(_useState, 2),
    isLoaded = _useState2[0],
    setLoaded = _useState2[1];
  var _useState3 = (0, _react.useState)(null),
    _useState4 = _slicedToArray(_useState3, 2),
    rootElement = _useState4[0],
    setRootElement = _useState4[1];
  var rootRect = (0, _ui.useElementRect)(rootElement);
  var rootWidth = (rootRect === null || rootRect === void 0 ? void 0 : rootRect.width) || 0;
  var acceptTone = isRejected || readOnly ? 'critical' : 'primary';
  var tone = drag ? acceptTone : 'default';
  var maxHeightToPx = _ImagePreview.MAX_DEFAULT_HEIGHT * document.documentElement.clientHeight / 100; // convert from vh to px, max height of the input

  var _getImageSize = getImageSize(src),
    _getImageSize2 = _slicedToArray(_getImageSize, 2),
    imageWidth = _getImageSize2[0],
    imageHeight = _getImageSize2[1];
  var imageRatio = imageWidth / imageHeight;

  // is the image wider than root? if so calculate the resized height
  var renderedImageHeight = imageWidth > rootWidth ? rootWidth / imageRatio : imageHeight;

  /*
    if the rendered image is smaller than the max height then it doesn't require a height set
    otherwise, set the max height (to prevent a large image in the document)
  */
  var rootHeight = renderedImageHeight < maxHeightToPx ? null : "".concat(_ImagePreview.MAX_DEFAULT_HEIGHT, "vh");
  (0, _react.useEffect)(() => {
    /* set for when the src is being switched when the image input already had a image src
    - meaning it already had an asset */
    setLoaded(false);
  }, [src]);
  var onLoadChange = (0, _react.useCallback)(() => {
    setLoaded(true);
  }, []);
  return /*#__PURE__*/_react.default.createElement(_ImagePreview.RatioBox, _extends({}, rest, {
    ref: setRootElement,
    style: {
      height: rootHeight
    },
    tone: "transparent"
  }), /*#__PURE__*/_react.default.createElement(_ui.Card, {
    "data-container": true,
    tone: "inherit"
  }, !isLoaded && /*#__PURE__*/_react.default.createElement(OverlayComponent, {
    cardTone: "transparent",
    drag: true,
    content: /*#__PURE__*/_react.default.createElement(_ImagePreview.SpinnerWrapper, null)
  }), /*#__PURE__*/_react.default.createElement("img", {
    src: src,
    "data-testid": "hotspot-image-input",
    alt: props.alt,
    onLoad: onLoadChange,
    referrerPolicy: "strict-origin-when-cross-origin"
  })), drag && /*#__PURE__*/_react.default.createElement(OverlayComponent, {
    cardTone: tone,
    drag: drag,
    content: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginBottom: 3
    }, /*#__PURE__*/_react.default.createElement(_ui.Heading, null, /*#__PURE__*/_react.default.createElement(HoverIcon, {
      isRejected: isRejected,
      readOnly: readOnly
    }))), /*#__PURE__*/_react.default.createElement(HoverText, {
      isRejected: isRejected,
      readOnly: readOnly
    }))
  }));
}
function HoverIcon(_ref) {
  var isRejected = _ref.isRejected,
    readOnly = _ref.readOnly;
  if (isRejected) {
    return /*#__PURE__*/_react.default.createElement(_icons.AccessDeniedIcon, null);
  }
  if (readOnly) {
    return /*#__PURE__*/_react.default.createElement(_icons.ReadOnlyIcon, null);
  }
  return /*#__PURE__*/_react.default.createElement(_icons.ImageIcon, null);
}
function HoverText(_ref2) {
  var isRejected = _ref2.isRejected,
    readOnly = _ref2.readOnly;
  var message = 'Drop image to upload';
  if (isRejected) {
    message = 'Cannot upload this file here';
  }
  if (readOnly) {
    message = 'This field is read only';
  }
  return /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, message);
}
function OverlayComponent(_ref3) {
  var cardTone = _ref3.cardTone,
    drag = _ref3.drag,
    content = _ref3.content;
  return /*#__PURE__*/_react.default.createElement(_ImagePreview.Overlay, {
    justify: "flex-end",
    padding: 3,
    tone: cardTone,
    drag: drag
  }, /*#__PURE__*/_react.default.createElement(_ImagePreview.FlexOverlay, {
    direction: "column",
    align: "center",
    justify: "center"
  }, content));
}
