// @sanity/form-builder/lib/inputs/PortableText/text/AnnotationToolbarPopover.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AnnotationToolbarPopover = AnnotationToolbarPopover;
var _react = _interopRequireWildcard(require("react"));
var _ui = require("@sanity/ui");
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _icons = require("@sanity/icons");
var _portableTextEditor = require("@sanity/portable-text-editor");
var _templateObject;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var ToolbarPopover = (0, _styledComponents.default)(_ui.Popover)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  &[data-popper-reference-hidden='true'] {\n    display: none !important;\n  }\n"])));
var POPOVER_FALLBACK_PLACEMENTS = ['top', 'bottom'];
function AnnotationToolbarPopover(props) {
  var scrollElement = props.scrollElement,
    annotationElement = props.annotationElement,
    textElement = props.textElement,
    title = props.title,
    onEdit = props.onEdit,
    onDelete = props.onDelete;
  var _useState = (0, _react.useState)(false),
    _useState2 = _slicedToArray(_useState, 2),
    open = _useState2[0],
    setOpen = _useState2[1];
  var _useState3 = (0, _react.useState)(null),
    _useState4 = _slicedToArray(_useState3, 2),
    cursorRect = _useState4[0],
    setCursorRect = _useState4[1];
  var _useState5 = (0, _react.useState)(null),
    _useState6 = _slicedToArray(_useState5, 2),
    selection = _useState6[0],
    setSelection = _useState6[1];
  var isClosingRef = (0, _react.useRef)(false);
  var rangeRef = (0, _react.useRef)(null);
  var editButtonRef = (0, _react.useRef)();
  var isTabbing = (0, _react.useRef)(false);
  var _useTheme = (0, _ui.useTheme)(),
    sanity = _useTheme.sanity;
  var editor = (0, _portableTextEditor.usePortableTextEditor)();
  var popoverScheme = sanity.color.dark ? 'light' : 'dark';

  // This is a "virtual element" (supported by Popper.js)
  var cursorElement = (0, _react.useMemo)(() => {
    if (!cursorRect) {
      return null;
    }
    return {
      getBoundingClientRect: () => {
        return cursorRect;
      }
    };
  }, [cursorRect]);
  (0, _react.useEffect)(() => {
    if (!open) {
      return undefined;
    }
    var handleScroll = () => {
      if (rangeRef.current) {
        setCursorRect(rangeRef.current.getBoundingClientRect());
      }
    };
    scrollElement.addEventListener('scroll', handleScroll, {
      passive: true
    });
    return () => {
      scrollElement.removeEventListener('scroll', handleScroll);
    };
  }, [open, scrollElement]);

  // Close floating toolbar on Escape
  (0, _ui.useGlobalKeyDown)((0, _react.useCallback)(event => {
    if (!open) {
      return;
    }
    if (event.key === 'Escape') {
      event.preventDefault();
      event.stopPropagation();
      setOpen(false);
      isTabbing.current = false;
      _portableTextEditor.PortableTextEditor.focus(editor);
    }
    if (event.key === 'Tab') {
      if (!isTabbing.current) {
        event.preventDefault();
        event.stopPropagation();
        editButtonRef.current.focus();
        isTabbing.current = true;
      }
    }
  }, [editor, open]));

  // Detect selection changes
  (0, _react.useEffect)(() => {
    function handleSelectionChange() {
      if (!textElement) return;
      var winSelection = window.getSelection();
      var anchorNode = winSelection.anchorNode,
        anchorOffset = winSelection.anchorOffset,
        focusNode = winSelection.focusNode,
        focusOffset = winSelection.focusOffset;
      setSelection({
        anchorNode,
        anchorOffset,
        focusNode,
        focusOffset
      });
    }
    document.addEventListener('selectionchange', handleSelectionChange, {
      passive: true
    });
    return () => {
      document.removeEventListener('selectionchange', handleSelectionChange);
    };
  }, [textElement]);

  // Open popover when selection is within annotations
  (0, _react.useEffect)(() => {
    if (!selection) return;
    if (isClosingRef.current) return;
    var anchorNode = selection.anchorNode,
      focusNode = selection.focusNode;
    // Safari would close the popover by loosing range when button is focused.
    // If we are focused and currently tabbing to the action buttons, just return here.
    if (isTabbing.current) {
      return;
    }
    if (annotationElement && annotationElement.contains(anchorNode) && anchorNode === focusNode) {
      var range = window.getSelection().getRangeAt(0);
      var rect = range.getBoundingClientRect();
      rangeRef.current = range;
      if (rect) {
        setCursorRect(rect);
      }
      setOpen(true);
    } else {
      setOpen(false);
      setCursorRect(null);
      rangeRef.current = null;
    }
  }, [selection, annotationElement]);
  if (!open) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement(ToolbarPopover, {
    boundaryElement: scrollElement,
    constrainSize: true,
    content: /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 1
    }, /*#__PURE__*/_react.default.createElement(_ui.Inline, {
      space: 1
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      weight: "semibold",
      size: 1
    }, title)), /*#__PURE__*/_react.default.createElement(_ui.Button, {
      ref: editButtonRef,
      icon: _icons.EditIcon,
      mode: "bleed",
      onClick: onEdit,
      padding: 2
    }), /*#__PURE__*/_react.default.createElement(_ui.Button, {
      icon: _icons.TrashIcon,
      mode: "bleed",
      padding: 2,
      onClick: onDelete,
      tone: "critical"
    }))),
    fallbackPlacements: POPOVER_FALLBACK_PLACEMENTS,
    open: open,
    placement: "top",
    portal: "editor",
    referenceElement: cursorElement,
    scheme: popoverScheme
  });
}
