// @sanity/form-builder/lib/inputs/PortableText/text/textStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TEXT_STYLES = exports.Normal = exports.Heading6 = exports.Heading5 = exports.Heading4 = exports.Heading3 = exports.Heading2 = exports.Heading1 = exports.BlockQuote = void 0;
var _ui = require("@sanity/ui");
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _excluded = ["children"],
  _excluded2 = ["children"],
  _excluded3 = ["children"],
  _excluded4 = ["children"],
  _excluded5 = ["children"],
  _excluded6 = ["children"],
  _excluded7 = ["children"],
  _excluded8 = ["children"];
var _templateObject, _templateObject2;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
// Without this container, editing with Android breaks due to how Text is styled via responsiveFont in @sanity/ui
var TextContainer = _styledComponents.default.div(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  display: block;\n"])));
var Normal = _ref => {
  var children = _ref.children,
    rest = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/_react.default.createElement(_ui.Text, _extends({
    "data-testid": "text-style--normal"
  }, rest), /*#__PURE__*/_react.default.createElement(TextContainer, null, children));
};
exports.Normal = Normal;
var Heading1 = _ref2 => {
  var children = _ref2.children,
    rest = _objectWithoutProperties(_ref2, _excluded2);
  return /*#__PURE__*/_react.default.createElement(_ui.Heading, _extends({
    as: "h1",
    "data-testid": "text-style--h1",
    size: 5
  }, rest), /*#__PURE__*/_react.default.createElement(TextContainer, null, children));
};
exports.Heading1 = Heading1;
var Heading2 = _ref3 => {
  var children = _ref3.children,
    rest = _objectWithoutProperties(_ref3, _excluded3);
  return /*#__PURE__*/_react.default.createElement(_ui.Heading, _extends({
    as: "h2",
    "data-testid": "text-style--h2",
    size: 4
  }, rest), /*#__PURE__*/_react.default.createElement(TextContainer, null, children));
};
exports.Heading2 = Heading2;
var Heading3 = _ref4 => {
  var children = _ref4.children,
    rest = _objectWithoutProperties(_ref4, _excluded4);
  return /*#__PURE__*/_react.default.createElement(_ui.Heading, _extends({
    as: "h3",
    "data-testid": "text-style--h3",
    size: 3
  }, rest), /*#__PURE__*/_react.default.createElement(TextContainer, null, children));
};
exports.Heading3 = Heading3;
var Heading4 = _ref5 => {
  var children = _ref5.children,
    rest = _objectWithoutProperties(_ref5, _excluded5);
  return /*#__PURE__*/_react.default.createElement(_ui.Heading, _extends({
    as: "h4",
    "data-testid": "text-style--h4",
    size: 2
  }, rest), /*#__PURE__*/_react.default.createElement(TextContainer, null, children));
};
exports.Heading4 = Heading4;
var Heading5 = _ref6 => {
  var children = _ref6.children,
    rest = _objectWithoutProperties(_ref6, _excluded6);
  return /*#__PURE__*/_react.default.createElement(_ui.Heading, _extends({
    as: "h5",
    "data-testid": "text-style--h5",
    size: 1
  }, rest), /*#__PURE__*/_react.default.createElement(TextContainer, null, children));
};
exports.Heading5 = Heading5;
var Heading6 = _ref7 => {
  var children = _ref7.children,
    rest = _objectWithoutProperties(_ref7, _excluded7);
  return /*#__PURE__*/_react.default.createElement(_ui.Heading, _extends({
    as: "h6",
    "data-testid": "text-style--h6",
    size: 0
  }, rest), /*#__PURE__*/_react.default.createElement(TextContainer, null, children));
};
exports.Heading6 = Heading6;
var BlockQuoteRoot = _styledComponents.default.blockquote(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  position: relative;\n  display: block;\n  margin: 0;\n  padding-left: ", "px;\n\n  &::before {\n    content: '';\n    position: absolute;\n    left: 0;\n    top: -4px;\n    bottom: -4px;\n    width: 3px;\n    background: var(--card-border-color);\n  }\n"])), _ref8 => {
  var theme = _ref8.theme;
  return theme.sanity.space[3];
});
var BlockQuote = _ref9 => {
  var children = _ref9.children,
    rest = _objectWithoutProperties(_ref9, _excluded8);
  return /*#__PURE__*/_react.default.createElement(BlockQuoteRoot, _extends({
    "data-testid": "text-style--blockquote"
  }, rest), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p"
  }, children));
};
exports.BlockQuote = BlockQuote;
var TEXT_STYLES = {
  normal: Normal,
  h1: Heading1,
  h2: Heading2,
  h3: Heading3,
  h4: Heading4,
  h5: Heading5,
  h6: Heading6,
  blockquote: BlockQuote
};
exports.TEXT_STYLES = TEXT_STYLES;
