// @sanity/default-layout/lib/navbar/search/hooks/useSearch.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useSearch = useSearch;
var _internal = require("@sanity/base/_internal");
var _isEqual = _interopRequireDefault(require("lodash/isEqual"));
var _react = require("react");
var _reactRx = require("react-rx");
var _rxjs = require("rxjs");
var _operators = require("rxjs/operators");
var _versionedClient = require("../../../versionedClient");
var _hasSearchableTerms = require("../utils/hasSearchableTerms");
var _selectors = require("../utils/selectors");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
var INITIAL_SEARCH_STATE = {
  error: null,
  hits: [],
  loading: false,
  options: {},
  terms: {
    query: '',
    types: []
  }
};
function nonNullable(v) {
  return v !== null;
}
function sanitizeRequest(request) {
  return _objectSpread(_objectSpread({}, request), {}, {
    terms: _objectSpread(_objectSpread({}, request.terms), {}, {
      query: request.terms.query.trim()
    })
  });
}
function useSearch(_ref) {
  var initialState = _ref.initialState,
    onComplete = _ref.onComplete,
    onError = _ref.onError,
    onStart = _ref.onStart,
    schema = _ref.schema;
  var _useState = (0, _react.useState)(initialState),
    _useState2 = _slicedToArray(_useState, 2),
    searchState = _useState2[0],
    setSearchState = _useState2[1];
  var searchWeighted = (0, _react.useMemo)(() => (0, _internal.createWeightedSearch)((0, _selectors.getSearchableOmnisearchTypes)(schema), _versionedClient.searchClient, {
    tag: 'search.global',
    unique: true
  }), [schema]);
  var handleQueryChange = (0, _reactRx.useObservableCallback)(inputValue$ => {
    return inputValue$.pipe((0, _operators.filter)(nonNullable), (0, _operators.map)(sanitizeRequest), (0, _operators.distinctUntilChanged)(_isEqual.default), (0, _operators.filter)(request => (0, _hasSearchableTerms.hasSearchableTerms)(request.terms)), (0, _operators.debounceTime)(300), (0, _operators.tap)(onStart), (0, _operators.switchMap)(request => (0, _rxjs.concat)((0, _rxjs.of)(_objectSpread(_objectSpread({}, INITIAL_SEARCH_STATE), {}, {
      loading: true,
      options: request.options,
      terms: request.terms
    })), searchWeighted(request.terms, request.options).pipe((0, _operators.map)(hits => ({
      hits
    })), (0, _operators.tap)(_ref2 => {
      var hits = _ref2.hits;
      return onComplete === null || onComplete === void 0 ? void 0 : onComplete(hits);
    }), (0, _operators.catchError)(error => {
      onError === null || onError === void 0 ? void 0 : onError(error);
      return (0, _rxjs.of)(_objectSpread(_objectSpread({}, INITIAL_SEARCH_STATE), {}, {
        error,
        loading: false,
        options: request.options,
        terms: request.terms
      }));
    })), (0, _rxjs.of)({
      loading: false
    }))), (0, _operators.scan)((prevState, nextState) => {
      return _objectSpread(_objectSpread({}, prevState), nextState);
    }, INITIAL_SEARCH_STATE), (0, _operators.tap)(setSearchState));
  }, []);
  var handleClearSearch = (0, _react.useCallback)(() => {
    setSearchState(INITIAL_SEARCH_STATE);
    handleQueryChange({
      terms: INITIAL_SEARCH_STATE.terms
    }); // cancel current request
  }, [handleQueryChange]);
  var handleSearch = (0, _react.useCallback)(searchRequest => handleQueryChange(searchRequest), [handleQueryChange]);
  return {
    handleSearch,
    handleClearSearch,
    searchState
  };
}
