// @sanity/form-builder/lib/inputs/arrays/ArrayOfObjectsInput/ConfirmDeleteButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ConfirmDeleteButton = ConfirmDeleteButton;
var _react = _interopRequireDefault(require("react"));
var _icons = require("@sanity/icons");
var _ui = require("@sanity/ui");
var _autoId = require("@reach/auto-id");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function ConfirmDeleteButton(props) {
  var onConfirm = props.onConfirm,
    _props$placement = props.placement,
    placement = _props$placement === void 0 ? 'left' : _props$placement,
    disabled = props.disabled;
  var id = (0, _autoId.useId)();
  return /*#__PURE__*/_react.default.createElement(_ui.MenuButton, {
    button: /*#__PURE__*/_react.default.createElement(_ui.Button, {
      icon: _icons.EllipsisVerticalIcon,
      mode: "bleed",
      padding: 2,
      disabled: disabled
    }),
    id: id || '',
    popover: {
      scheme: 'light',
      tone: 'default',
      portal: true
    },
    menu: /*#__PURE__*/_react.default.createElement(_ui.Menu, null, /*#__PURE__*/_react.default.createElement(_ui.MenuItem, {
      icon: _icons.TrashIcon,
      onClick: onConfirm,
      text: props.title,
      tone: "critical"
    })),
    placement: placement
  });
}
