// @sanity/block-tools/lib/dts/src/constants.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { uniq } from 'lodash';
export var BLOCK_DEFAULT_STYLE = 'normal';
export var DEFAULT_BLOCK = Object.freeze({
    _type: 'block',
    markDefs: [],
    style: BLOCK_DEFAULT_STYLE,
});
export var DEFAULT_SPAN = Object.freeze({
    _type: 'span',
    marks: [],
});
export var HTML_BLOCK_TAGS = {
    p: DEFAULT_BLOCK,
    blockquote: __assign(__assign({}, DEFAULT_BLOCK), { style: 'blockquote' }),
};
export var HTML_SPAN_TAGS = {
    span: { object: 'text' },
};
export var HTML_LIST_CONTAINER_TAGS = {
    ol: { object: null },
    ul: { object: null },
};
export var HTML_HEADER_TAGS = {
    h1: __assign(__assign({}, DEFAULT_BLOCK), { style: 'h1' }),
    h2: __assign(__assign({}, DEFAULT_BLOCK), { style: 'h2' }),
    h3: __assign(__assign({}, DEFAULT_BLOCK), { style: 'h3' }),
    h4: __assign(__assign({}, DEFAULT_BLOCK), { style: 'h4' }),
    h5: __assign(__assign({}, DEFAULT_BLOCK), { style: 'h5' }),
    h6: __assign(__assign({}, DEFAULT_BLOCK), { style: 'h6' }),
};
export var HTML_MISC_TAGS = {
    br: __assign(__assign({}, DEFAULT_BLOCK), { style: BLOCK_DEFAULT_STYLE }),
};
export var HTML_DECORATOR_TAGS = {
    b: 'strong',
    strong: 'strong',
    i: 'em',
    em: 'em',
    u: 'underline',
    s: 'strike-through',
    strike: 'strike-through',
    del: 'strike-through',
    code: 'code',
};
export var HTML_LIST_ITEM_TAGS = {
    li: __assign(__assign({}, DEFAULT_BLOCK), { style: BLOCK_DEFAULT_STYLE, level: 1, listItem: 'bullet' }),
};
export var ELEMENT_MAP = __assign(__assign(__assign(__assign(__assign(__assign({}, HTML_BLOCK_TAGS), HTML_SPAN_TAGS), HTML_LIST_CONTAINER_TAGS), HTML_LIST_ITEM_TAGS), HTML_HEADER_TAGS), HTML_MISC_TAGS);
export var DEFAULT_SUPPORTED_STYLES = uniq(Object.values(ELEMENT_MAP)
    .filter(function (tag) { return 'style' in tag; })
    .map(function (tag) { return tag.style; }));
export var DEFAULT_SUPPORTED_DECORATORS = uniq(Object.values(HTML_DECORATOR_TAGS));
export var DEFAULT_SUPPORTED_ANNOTATIONS = ['link'];
