// @sanity/form-builder/lib/__workshop__/_common/FormBuilderTester.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormBuilderTester = FormBuilderTester;
var _react = _interopRequireDefault(require("react"));
var _ui = require("@sanity/ui");
var _FormBuilderContext = _interopRequireDefault(require("../../FormBuilderContext"));
var _ReviewChangesProvider = require("../../sanity/contexts/reviewChanges/ReviewChangesProvider");
var _inputResolver = require("./inputResolver");
var _resolvePreviewComponent = require("./resolvePreviewComponent");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function FormBuilderTester(props) {
  var value = props.value,
    patchChannel = props.patchChannel,
    _props$isChangesOpen = props.isChangesOpen,
    isChangesOpen = _props$isChangesOpen === void 0 ? false : _props$isChangesOpen;
  return /*#__PURE__*/_react.default.createElement(_ui.ThemeProvider, {
    theme: _ui.studioTheme
  }, /*#__PURE__*/_react.default.createElement(_ui.LayerProvider, null, /*#__PURE__*/_react.default.createElement(_ui.ToastProvider, null, /*#__PURE__*/_react.default.createElement(_ReviewChangesProvider.ReviewChangesContextProvider, {
    changesOpen: isChangesOpen
  }, /*#__PURE__*/_react.default.createElement(_FormBuilderContext.default, {
    value: value,
    patchChannel: patchChannel,
    schema: props.schema,
    resolveInputComponent: _inputResolver.inputResolver,
    resolvePreviewComponent: _resolvePreviewComponent.resolvePreviewComponent
  }, props.children)))));
}
