// @sanity/default-layout/lib/update/changelog/ChangelogItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChangelogItem = ChangelogItem;
var _ui = require("@sanity/ui");
var _react = _interopRequireDefault(require("react"));
var _PortableTextContent = require("../PortableTextContent");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var LABELS = {
  bugfix: 'Bugfixes',
  feature: 'Feature'
};
function ChangelogItem(props) {
  var changeType = props.changeType,
    description = props.description,
    title = props.title;
  var label = LABELS[changeType];
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    padding: 4,
    shadow: 1,
    radius: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: title ? 4 : 3
  }, label && /*#__PURE__*/_react.default.createElement(_ui.Label, {
    size: 1
  }, label), /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 3
  }, title && /*#__PURE__*/_react.default.createElement(_ui.Heading, {
    size: 1
  }, title), description && /*#__PURE__*/_react.default.createElement(_PortableTextContent.PortableTextContent, {
    value: description
  }))));
}
