// @sanity/base/lib/presence/DocumentPreviewPresence.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocumentPreviewPresence = DocumentPreviewPresence;
var _ui = require("@sanity/ui");
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireWildcard(require("styled-components"));
var _components = require("../components");
var _templateObject, _templateObject2;
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var PRESENCE_MENU_POPOVER_PROPS = {
  portal: true,
  scheme: 'light'
};
var AvatarStackCard = (0, _styledComponents.default)(_ui.Card)(_ref => {
  var theme = _ref.theme,
    $selected = _ref.$selected;
  var color = theme.sanity.color;
  return (0, _styledComponents.css)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n    --card-bg-color: inherit;\n    --card-fg-color: inherit;\n    --card-hairline-hard-color: ", ";\n  "])), $selected ? color.selectable.default.pressed.border : undefined);
});
var TooltipContentBox = (0, _styledComponents.default)(_ui.Box)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  max-width: 150px;\n"])));
var getTooltipText = presence => {
  if (presence.length === 1) {
    return "".concat(presence[0].displayName, " is editing this document");
  }
  if (presence.length > 1) {
    return "".concat(presence.length, " people are editing this document right now");
  }
  return undefined;
};
function DocumentPreviewPresence(props) {
  var presence = props.presence;
  var color = (0, _styledComponents.useTheme)().sanity.color;
  var invertedScheme = color.dark ? 'light' : 'dark';
  var _usePreviewCard = (0, _components.usePreviewCard)(),
    selected = _usePreviewCard.selected;
  var tooltipContent = (0, _react.useMemo)(() => {
    return /*#__PURE__*/_react.default.createElement(TooltipContentBox, {
      padding: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      align: "center",
      size: 1
    }, getTooltipText(presence)));
  }, [presence]);
  return /*#__PURE__*/_react.default.createElement(_ui.Tooltip, _extends({
    content: tooltipContent
  }, PRESENCE_MENU_POPOVER_PROPS), /*#__PURE__*/_react.default.createElement(AvatarStackCard, {
    scheme: selected ? invertedScheme : undefined,
    $selected: selected
  }, /*#__PURE__*/_react.default.createElement(_ui.AvatarStack, {
    maxLength: 2
  }, presence.map(user => /*#__PURE__*/_react.default.createElement(_components.UserAvatar, {
    key: user.id,
    user: user
  })))));
}
