// @sanity/form-builder/lib/inputs/PortableText/BlockActions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BlockActions = BlockActions;
var _portableTextEditor = require("@sanity/portable-text-editor");
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _createBlockActionPatchFn = require("./utils/createBlockActionPatchFn");
var _templateObject;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var Root = _styledComponents.default.div(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  display: flex;\n  pointer-events: 'all';\n"])));
function isClassComponent(component) {
  var _component$prototype;
  return typeof component === 'function' && !!((_component$prototype = component.prototype) !== null && _component$prototype !== void 0 && _component$prototype.isReactComponent);
}
function isFunctionComponent(component) {
  return typeof component === 'function' && String(component).includes('return React.createElement');
}
function BlockActions(props) {
  var editor = (0, _portableTextEditor.usePortableTextEditor)();
  var block = props.block,
    onChange = props.onChange,
    renderBlockActions = props.renderBlockActions;
  var decoratorValues = (0, _react.useMemo)(() => _portableTextEditor.PortableTextEditor.getPortableTextFeatures(editor).decorators.map(d => d.value), [editor]);
  var blockActions = (0, _react.useMemo)(() => {
    if (renderBlockActions) {
      var blockActionProps = {
        block,
        value: _portableTextEditor.PortableTextEditor.getValue(editor),
        set: (0, _createBlockActionPatchFn.createBlockActionPatchFn)('set', block, onChange, decoratorValues),
        unset: (0, _createBlockActionPatchFn.createBlockActionPatchFn)('unset', block, onChange, decoratorValues),
        insert: (0, _createBlockActionPatchFn.createBlockActionPatchFn)('insert', block, onChange, decoratorValues)
      };
      // Support returning a class component for renderBlockActions (to keep backward compatability as it was possible before)
      if (isClassComponent(renderBlockActions) || isFunctionComponent(renderBlockActions)) {
        var RenderComponent = renderBlockActions;
        return /*#__PURE__*/_react.default.createElement(RenderComponent, blockActionProps);
      }
      return renderBlockActions(blockActionProps);
    }
    return undefined;
  }, [renderBlockActions, block, editor, onChange, decoratorValues]);

  // Take focus away from the editor so dealing with block actions doesn't interfere with the editor focus
  var handleClick = (0, _react.useCallback)(() => {
    _portableTextEditor.PortableTextEditor.blur(editor);
  }, [editor]);

  // Don't render anything if the renderBlockActions function returns null.
  // Note that if renderBlockComponent is a React class, this will never be the case.
  if (!blockActions) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement(Root, {
    contentEditable: false,
    onKeyDown: handleClick,
    onMouseDown: handleClick
  }, blockActions);
}
